---
title: MetaNetï¼šå¤šç»„å­¦ç½‘ç»œåˆ†æå·¥å…·ï½œ4.å¸ƒå±€å’Œå¯è§†åŒ–
author: Peng Chen
date: '2025-04-12'
slug: metanet-4
categories:
  - R
tags:
  - network
  - R
description: ä»‹ç»MetaNetä¸­çš„å„ç§å¯è§†åŒ–æ–¹æ³•ï¼Œä»åŸºç¡€ç»˜å›¾åˆ°é«˜çº§å¸ƒå±€æŠ€å·§ã€‚
image: index.en_files/figure-html/unnamed-chunk-13-1.png
math: ~
license: ~
hidden: no
comments: yes
---

```{r include=FALSE}
devtools::load_all("~/Documents/R/pcutils/")
devtools::load_all("~/Documents/R/MetaNet/MetaNet/")
data(otutab, package = "pcutils")
Packages <- c("dplyr", "pcutils", "kableExtra")
pcutils::lib_ps(Packages)
knitr::opts_chunk$set(message = FALSE,warning = FALSE,cache = T,fig.width=8,fig.height=5)
```

ä¹‹å‰å·²ç»ä»‹ç»äº†ç½‘ç»œè®¡ç®—ï¼Œæ„å»ºä»¥åŠå„ç§æ³¨é‡Šäº†ã€‚æœ¬æ–‡å°†è¯¦ç»†ä»‹ç»MetaNetä¸­çš„å„ç§å¯è§†åŒ–æ–¹æ³•ï¼Œä»åŸºç¡€ç»˜å›¾åˆ°é«˜çº§å¸ƒå±€æŠ€å·§ã€‚


- è½¯ä»¶ä¸»é¡µï¼š<https://github.com/Asa12138/MetaNet> **å¤§å®¶å¯ä»¥å¸®å¿™åœ¨githubä¸Šç‚¹ç‚¹starâ­ï¸**ï¼Œè°¢è°¢ğŸ™
- è¯¦ç»†è‹±æ–‡ç‰ˆæ•™ç¨‹ï¼š<https://bookdown.org/Asa12138/metanet_book/>

å¯ä»¥ä» CRAN å®‰è£…ç¨³å®šç‰ˆï¼š`install.packages("MetaNet")`  

æœ€æ–°çš„å¼€å‘ç‰ˆæœ¬å¯ä»¥åœ¨ <https://github.com/Asa12138/MetaNet> ä¸­æ‰¾åˆ°ï¼š

```r
remotes::install_github("Asa12138/MetaNet", dependencies = T)
```

ä¾èµ–åŒ… `pcutils`å’Œ`igraph`ï¼ˆéœ€æå‰å®‰è£…ï¼‰ï¼Œæ¨èé…åˆ `dplyr` è¿›è¡Œæ•°æ®æ“ä½œã€‚

```r
library(MetaNet)
library(igraph)

# ========data manipulation
library(dplyr)
library(pcutils)
```

## ç»˜å›¾è®¾ç½®

åœ¨æ„å»ºç½‘ç»œæ—¶ï¼ŒMetaNetå·²ç»è®¾ç½®äº†ä¸€äº›ä¸å¯è§†åŒ–ç›¸å…³çš„å†…éƒ¨å±æ€§ã€‚æˆ‘ä»¬å¯ä»¥ä½¿ç”¨`c_net_set()`å‡½æ•°è‡ªå®šä¹‰è¿™äº›å±æ€§ä»¥æ»¡è¶³ç ”ç©¶éœ€æ±‚ã€‚

```{r echo=FALSE}
data("multi_test", package = "MetaNet")
data("c_net", package = "MetaNet")

# æ„å»ºå¤šç»„å­¦ç½‘ç»œ
multi1 <- multi_net_build(list(Microbiome = micro, Metabolome = metab, Transcriptome = transc))

# é»˜è®¤ä½¿ç”¨v_groupåˆ†ç»„ï¼Œç»˜åˆ¶ä¸‰ä¸ªä¸åŒçš„è¡¨æ ¼
plot(multi1)

# è®¾ç½®vertex_class
multi1_with_anno <- c_net_set(multi1, micro_g, metab_g, transc_g, 
                             vertex_class = c("Phylum", "kingdom", "type"))

# è®¾ç½®vertex_size
multi1_with_anno <- c_net_set(multi1_with_anno, 
                             data.frame("Abundance1" = colSums(micro)),
                             data.frame("Abundance2" = colSums(metab)),
                             data.frame("Abundance3" = colSums(transc)),
                             vertex_size = paste0("Abundance", 1:3))
plot(multi1_with_anno)
```


å¦‚æœéœ€è¦æ›´çµæ´»åœ°è‡ªå®šä¹‰ç½‘ç»œå›¾ï¼Œå¯ä»¥ä½¿ç”¨`c_net_plot()`å‡½æ•°ï¼Œå®ƒåŒ…å«è®¸å¤šçµæ´»çš„ç»˜å›¾å‚æ•°ï¼š

![](images/arg1.png)

![](images/arg2.png)

![](images/arg3.png)

ç¤ºä¾‹ä»£ç ï¼Œå°è¯•å„ç§è®¾ç½®çœ‹çœ‹ï¼š

```{r}
c_net_plot(multi1_with_anno, 
          labels_num = 5,
          vertex.color = get_cols(11, "col1"),
          vertex_size_range = c(3, 10),
          vertex.label.color = "red",
          edge_width_range = c(0.5, 3),
          edge.color = c("orange", "green4"),
          edge.curved = 0.5,
          legend = T,
          legend_number = T,
          group_legend_order = c("Microbiome", "Metabolome", "Transcriptome"),
          group_legend_title = c("Phylum", "Metabolome", "Transcriptome"),
          edge_legend_title = "Correlation",
          edge_legend_order = c("positive", "negative"),
          size_legend = T,
          size_legend_title = "Abundance",
          width_legend = T,
          width_legend_title = "abs(r)",
          lty_legend = T,
          lty_legend_title = "Omics relationship")
```

### ä½¿ç”¨params_list

`params_list`æ˜¯`c_net_plot()`ä¸­çš„ä¸€ä¸ªç‰¹æ®Šå‚æ•°ï¼Œå®ƒæ˜¯ä¸€ä¸ªåŒ…å«å‚æ•°çš„åˆ—è¡¨ï¼Œå¯ä»¥æ–¹ä¾¿åœ°ç”¨äºç»˜åˆ¶ä¸€ç³»åˆ—å…·æœ‰ç›¸åŒå±æ€§çš„ç½‘ç»œå›¾ï¼š

```{r}
node_colors <- setNames(get_cols(9, "col1"), unique(V(multi1_with_anno)$v_class))
params_list <- list(
  labels_num = 5,
  vertex.color = node_colors,
  vertex_size_range = c(3, 10),
  vertex.label.color = "red",
  edge_width_range = c(0.5, 3),
  edge.color = c("orange", "green4"),
  edge.curved = 0.5,
  legend = T,
  legend_number = T,
  group_legend_order = c("Microbiome", "Metabolome", "Transcriptome"),
  group_legend_title = c("Phylum", "Metabolome", "Transcriptome"),
  edge_legend_title = "Correlation",
  edge_legend_order = c("positive", "negative"),
  size_legend = T,
  size_legend_title = "Abundance",
  width_legend = T,
  width_legend_title = "abs(r)",
  lty_legend = T,
  lty_legend_title = "Omics relationship"
)

c_net_plot(multi1_with_anno, params_list = params_list)

# æ„å»ºå¦ä¸€ä¸ªå¤šç»„å­¦ç½‘ç»œ
multi1_with_anno2 <- c_net_filter(multi1_with_anno, 
                                 v_group %in% c("Microbiome", "Metabolome")) %>% 
  c_net_filter(., e_class == "intra", mode = "e")
c_net_plot(multi1_with_anno2, params_list = params_list)
```

## ç½‘ç»œå¸ƒå±€

å¸ƒå±€æ˜¯ç½‘ç»œå¯è§†åŒ–çš„é‡è¦ç»„æˆéƒ¨åˆ†ï¼Œä¸€ä¸ªå¥½çš„å¸ƒå±€å¯ä»¥æ¸…æ™°åœ°å‘ˆç°ä¿¡æ¯ã€‚

åœ¨MetaNetä¸­ï¼Œæˆ‘ä»¬ä½¿ç”¨`coors`å¯¹è±¡æ¥å­˜å‚¨å¸ƒå±€çš„åæ ‡ã€‚`coors`æ˜¯ä¸€ä¸ªå…·æœ‰"name", "X", "Y"ä¸‰åˆ—çš„dataframeã€‚

### åŸºç¡€å¸ƒå±€

ä½¿ç”¨`c_net_layout()`è·å–ç‰¹å®šå¸ƒå±€æ–¹æ³•çš„åæ ‡ï¼š

```{r}
c_net_layout(co_net2, method = in_circle()) -> coors
c_net_plot(co_net2, coors)
```

å¯ç”¨çš„åŸºç¡€å¸ƒå±€æ–¹æ³•åŒ…æ‹¬ï¼š

1. igraphå¸ƒå±€ï¼š`in_circle()`, `nicely()`, `on_grid()`, `on_sphere()`, `randomly()`, `with_dh()`, `with_fr()`, `with_gem()`, `with_graphopt()`, `with_kk()`, `with_lgl()`, `with_mds()`
2. metanetæ–°å¸ƒå±€æ–¹æ³•ï¼š`as_line()`, `as_arc()`, `as_polygon()`, `as_polyarc()`, `as_polycircle()`, `as_circle_tree()`, `as_multi_layer()`, `as_poly_sector()`
3. ggraphå¸ƒå±€ï¼š"auto", "backbone", "centrality", "circlepack", "dendrogram", "eigen", "focus", "hive", "igraph", "linear", "manual", "matrix", "partition", "pmds", "stress", "treemap", "unrooted"

ç¤ºä¾‹ä»£ç å±•ç¤ºä¸åŒå¸ƒå±€æ•ˆæœï¼š

```{r eval=FALSE}
go <- erdos.renyi.game(30, 0.25)
# get a metanet
go <- c_net_update(go)

layout_methods <- list(
  as_star(), as_tree(), in_circle(), nicely(),
  on_grid(), on_sphere(), randomly(), with_dh(),
  with_fr(), with_gem(), with_graphopt(), with_kk(),
  with_lgl(), with_mds(), as_line(), as_arc(),
  as_polygon(), as_polyarc(), as_polycircle(3), as_circle_tree(),
  as_multi_layer(2), as_poly_sector()
)
names(layout_methods) <- c(
  "as_star ", "as_tree ", "in_circle ", "nicely ",
  "on_grid ", "on_sphere ", "randomly ", "with_dh ",
  "with_fr ", "with_gem ", "with_graphopt ", "with_kk ",
  "with_lgl ", "with_mds", "as_line", "as_arc",
  "as_polygon", "as_polyarc", "as_polycircle", "as_circle_tree",
  "as_multi_layer", "as_poly_sector"
)

par(mfrow = c(6, 4))
for (i in names(layout_methods)) {
  plot(go, layout_methods[[i]], legend = F, main = i, labels_num = 0)
}
```

![](https://bookdown.org/Asa12138/metanet_book/04-visualization_files/figure-html/fig-4-layout-methods-1.png)
å¯¹äºæ¯ç§æ–¹æ³•ï¼Œå¯ä»¥åœ¨å…¶ä¸­é¢å¤–æ·»åŠ ä¸€äº›å‚æ•°ï¼š

```{r eval=FALSE}
# get a metanet
go <- erdos.renyi.game(30, 0.25)
go <- c_net_update(go)

plot(go, coors = with_fr())
plot(go, coors = with_fr(niter = 99, grid = "nogrid"))
```


`as_polygon()`å¾ˆæœ‰è¶£ï¼Œå®ƒå¯ä»¥ç»˜åˆ¶å¤šè¾¹å½¢å½¢çŠ¶çš„ç½‘ç»œï¼Œæ‚¨å¯ä»¥æ›´æ”¹å¤šè¾¹å½¢çš„è¾¹æ•°:

![](https://bookdown.org/Asa12138/metanet_book/images/4-1.multiangle.gif)

### å˜æ¢å¸ƒå±€

ä½¿ç”¨`transform_comors`å¯ä»¥è½¬æ¢å¸ƒå±€ï¼ŒåŒ…æ‹¬ç¼©æ”¾ï¼ŒX/Yæ¯”ï¼Œæ—‹è½¬è§’åº¦ï¼Œé•œåƒï¼Œä¼ª3Dæ•ˆæœç­‰ï¼š

```{r eval=FALSE}
c_net_layout(multi1_with_anno)->coors
c_net_plot(multi1_with_anno,
           transform_coors(coors,
                           scale = 0.8,
                           aspect_ratio = 0.5,
                           rotation = i,
                           mirror_x = T,
                           shear_x = 1)
           )
```


![](https://bookdown.org/Asa12138/metanet_book/images/4-2.trans_coors.gif)

### åˆ†ç»„å¸ƒå±€

é™¤`c_net_layout()`å¤–ï¼Œæˆ‘ä»¬è¿˜ä¸ºå…·æœ‰åˆ†ç»„å˜é‡çš„ç½‘ç»œæä¾›äº†ä¸€ç§é«˜çº§å¸ƒå±€æ–¹æ³•ï¼š`g_layout()`ã€‚
ä½¿ç”¨`g_layout()`å¯ä»¥è½»æ¾æ§åˆ¶æ¯ä¸ªç»„çš„ä½ç½®åŠå…¶å†…éƒ¨å¸ƒå±€ã€‚`g_layout()`è¿”å›çš„ä¹Ÿæ˜¯`coors`å¯¹è±¡ï¼Œè¿™æ„å‘³ç€æˆ‘ä»¬å¯ä»¥ç»§ç»­ç”¨`g_layout()`ç»„åˆï¼Œå®ç°ç–¯ç‹‚å¥—å¨ƒï¼Œé«˜åº¦è‡ªå®šä¹‰çš„å¸ƒå±€ï¼

`g_layout()`åœ¨å¤„ç†å¤šç»„å­¦ç½‘ç»œæˆ–æ¨¡å—ç½‘ç»œï¼‰æ—¶ï¼Œæ˜¯å¸ƒå±€åˆ†ç»„å˜é‡ç½‘ç»œçš„æä½³é€‰æ‹©ã€‚

* é¦–å…ˆï¼ŒæŒ‡å®šåˆ†ç»„å˜é‡group
* è®¾ç½®ç»„é—´å¸ƒå±€`layout1`ï¼Œå¯é€‰ï¼š
  1. æ•°æ®æ¡†æˆ–çŸ©é˜µï¼šè¡Œåä¸ºç»„åï¼Œä¸¤åˆ—åˆ†åˆ«ä¸ºXå’ŒYåæ ‡
  2. å‡½æ•°ï¼š`c_net_layout()`çš„å„ç§å¸ƒå±€æ–¹æ³•ï¼ˆé»˜è®¤ï¼šin_circle()ï¼‰
* è°ƒæ•´`layout1`çš„ç¼©æ”¾æ¯”ä¾‹`zoom1`
* è®¾ç½®ç»„å†…å¸ƒå±€`layout2`ï¼ˆä½¿ç”¨`c_net_layout()`çš„å„ç§å¸ƒå±€æ–¹æ³•ï¼‰ï¼Œ
  ç”¨ä¸€ä¸ªlistï¼Œä¸ºæ¯ä¸ªç»„å•ç‹¬æŒ‡å®šå¸ƒå±€å‡½æ•°æˆ–è€…ç›´æ¥ç»™ä¸€ä¸ªåæ ‡æ•°æ®æ¡†ã€‚
* è°ƒæ•´`layout2`çš„ç¼©æ”¾æ¯”ä¾‹`zoom2`ï¼Œå¯ç”¨å‘é‡åˆ†åˆ«æ§åˆ¶å„ç»„ç¼©æ”¾
* è®¾ç½®`show_big_layout = T`å¯æŸ¥çœ‹`layout1`çš„åˆ†å¸ƒæƒ…å†µ

```{r}
# ä¸ºæ¯ä¸ªç»„è®¾ç½®åœ†å½¢å¸ƒå±€
g_layout(multi1_with_anno, group = "v_group",
        layout1 = in_circle(), zoom1 = 10,
        layout2 = in_circle(), zoom2 = 5) -> g_coors
plot(multi1_with_anno, coors = g_coors)

# ä¸ºæ¯ä¸ªç»„è®¾ç½®ä¸åŒçš„å¸ƒå±€
g_layout(multi1_with_anno, group = "v_group",
        layout1 = in_circle(), zoom1 = 10,
        layout2 = list(in_circle(), with_fr(), as_polygon()),
        zoom2 = 3:5) -> g_coors
plot(multi1_with_anno, coors = g_coors)
```

#### tkplotæ‰‹åŠ¨è°ƒæ•´å¤§å¸ƒå±€

```{r eval=FALSE}
# é¦–å…ˆè·å–ç½‘ç»œéª¨æ¶
get_group_skeleton(co_net, "v_class") %>% clean_igraph() -> s_net

# ä½¿ç”¨tkplotè¿›è¡Œæ‰‹åŠ¨è°ƒæ•´
x <- igraph::tkplot(s_net)
# åœ¨tkplotçª—å£ä¸­ç§»åŠ¨èŠ‚ç‚¹åˆ°ä½ å–œæ¬¢çš„å¸ƒå±€ï¼
da <- igraph::tkplot.getcoords(x)
igraph::tkplot.close(x)

# å°†è°ƒæ•´åçš„åæ ‡ä¼ é€’ç»™layout1
g_layout(co_net, group = "v_class",
        layout1 = da, zoom1 = 20,
        layout2 = in_circle(), zoom2 = c(1, 4, 2, 1, 3, 5)) -> g_coors
plot(co_net, coors = g_coors)
```

![](https://bookdown.org/Asa12138/metanet_book/images/4-3.tkplot1.png)

![](https://bookdown.org/Asa12138/metanet_book/images/4-2.tkplot.png)

MetaNetè¿˜æä¾›äº†ä¸€äº›é¢„è®¾çš„åˆ†ç»„å¸ƒå±€æ–¹æ³•ï¼š

1. `g_layout_circlepack()`
2. `g_layout_treemap()`
3. `g_layout_backbone()`
4. `g_layout_stress()`
5. `g_layout_polyarc()`
6. `g_layout_polygon()`
7. `g_layout_polycircle()`
8. `g_layout_multi_layer()` ä¼ª3Dæ•ˆæœ
9. `g_layout_poly_sector()`

ç¤ºä¾‹ä»£ç ï¼š

```{r eval=FALSE}
E(co_net)$color <- rep("grey", length(E(co_net)))
plot(co_net, coors = g_layout_circlepack(co_net, group = "v_class"),
    legend = F, labels_num = 0, main = "g_layout_circlepack")
plot(co_net, coors = g_layout_polyarc(co_net, group = "v_class"),
    legend = F, labels_num = 0, main = "g_layout_polyarc")
plot(co_net, coors = g_layout_polycircle(co_net, group = "v_class"),
    legend = F, labels_num = 0, main = "g_layout_polycircle")

g1 <- module_net(3)

plot(g1,
  coors = g_layout_multi_layer(g1, group = "v_class", layout = on_grid()),
  legend = F, labels_num = 0, main = "g_layout_multi_layer"
)
```

![](images/g_lay.png)

## `spatstat` layout

```{r fig.width=12,fig.height=10}
if(!require("spatstat"))install.packages("spatstat")
E(co_net)$color <- rep("grey", length(E(co_net)))

par(mfrow = c(2, 2))
poly_x <- c(0, 2, 2, 0)
poly_y <- c(0, 0, 1, 1)
win_poly <- spatstat.geom::owin(poly = list(x = poly_x, y = poly_y))
plot(win_poly)
coors1 <- spatstat_layout(co_net, win_poly, type = "random", mode = "surface")
plot(co_net, coors = coors1)
coors1 <- spatstat_layout(co_net, win_poly, type = "regular", mode = "surface",order_by="v_class")
plot(co_net, coors = coors1)
coors2 <- spatstat_layout(co_net2, win_poly, type = "random", mode = "boundary")
plot(co_net2, coors = coors2)
```


å°è¯•ç”»ä¸ªäº”è§’æ˜Ÿâ­ï¸ï¼š

```{r}
library(spatstat.geom)

create_star_window <- function(r_outer = 1, r_inner = 0.4, center = c(0, 0)) {
  # åˆ›å»ºäº”è§’æ˜Ÿçš„10ä¸ªé¡¶ç‚¹ï¼ˆå¤–ã€å†…äº¤æ›¿ï¼‰
  theta <- seq(0, 2 * pi, length.out = 11)[-11]  # 10ä¸ªç‚¹
  theta_outer <- theta[seq(1, 10, 2)]
  theta_inner <- theta[seq(2, 10, 2)]

  x <- c(r_outer * cos(theta_outer),
         r_inner * cos(theta_inner))
  y <- c(r_outer * sin(theta_outer),
         r_inner * sin(theta_inner))

  # é‡æ–°æ’åºæˆé¦–å°¾ç›¸è¿çš„è·¯å¾„
  order_index <- c(1,6,2,7,3,8,4,9,5,10)
  x <- x[order_index] + center[1]
  y <- y[order_index] + center[2]

  # æ„å»º spatstat çš„ owin çª—å£
  win <- owin(poly = list(x = x, y = y))
  return(win)
}

win_star <- create_star_window()

plot(co_net, coors = spatstat_layout(co_net, win_star, order_by="v_class"))
```

ç”šè‡³å¯ä»¥ç”»æˆåœ°å›¾ï¼š

```{r}
library(rnaturalearth)
library(sf)

# è·å–éæ´²å›½å®¶è¾¹ç•Œï¼ˆ1:50mç²¾åº¦ï¼‰
africa_sf <- ne_countries(continent = "Africa", scale = 50, returnclass = "sf")

# å¯è§†åŒ–
plot(st_geometry(africa_sf), col = "sandybrown", main = "Africa Outline (1:50m)", )

plot(co_net, coors = spatstat_layout(co_net, africa_sf, type = "regular", order_by="v_class"))
```

MetaNetä½¿ç”¨çš„æ˜¯igraphçš„ç»˜å›¾æ–¹å¼ï¼ŒRçš„åŸºç¡€ç»˜å›¾ï¼Œæ‰€ä»¥éœ€è¦ç”¨`pdf`,`png`ç­‰è®¾å¤‡ä¿å­˜å›¾ç‰‡ã€‚ä¸‹ä¸€èŠ‚ä»‹ç»MetaNetå’Œå…¶ä»–ç»˜å›¾æ–¹å¼å¦‚ggplot2ï¼ŒD3ç­‰çš„è½¬æ¢ï¼Œä»¥åŠMetaNeté…åˆGephiï¼ŒCytoscapeç­‰äº¤äº’å¼è½¯ä»¶ä½¿ç”¨ã€‚

## References
1. Koutrouli M, Karatzas E, Paez-Espino D and Pavlopoulos GA (2020) A Guide to Conquer the Biological Network Era Using Graph Theory. Front. Bioeng. Biotechnol. 8:34. doi: 10.3389/fbioe.2020.00034
2. Faust, K., and Raes, J. (2012). Microbial interactions: from networks to models. Nat. Rev. Microbiol. https://doi.org/10.1038/nrmicro2832.
3. Y. Deng, Y. Jiang, Y. Yang, Z. He, et al., Molecular ecological network analyses. BMC bioinformatics (2012), doi:10.1186/1471-2105-13-113.