---
title: æ‰©å¢å­(Amplicon)æ•°æ®åˆ†ææµç¨‹|Qiime2
author: Peng Chen
date: '2024-12-23'
slug: amplicon-workflow
categories:
  - metagenomic
tags:
  - 16S
  - amplicon
description: æ‰©å¢å­ï¼ˆAmpliconï¼‰åˆ†ææ˜¯ç ”ç©¶å¾®ç”Ÿç‰©ç¾¤è½ç»„æˆä¸åŠŸèƒ½çš„æ ¸å¿ƒæ–¹æ³•ä¹‹ä¸€ï¼Œè¿™é‡Œå­¦ä¹ å¹¶æ­å»ºä¸€ä¸‹è‡ªå·±ç”¨çš„æ¯”è¾ƒèˆ’æœçš„æµç¨‹ã€‚
image: images/16s.webp
math: ~
license: ~
hidden: no
comments: yes
---

```{r include=FALSE}
Packages <- c("dplyr","kableExtra","ggplot2")
pcutils::lib_ps(Packages)
knitr::opts_chunk$set(message = FALSE,warning = FALSE,eval = FALSE)
```

## Introduction

æˆ‘å­¦ä¹ å¾®ç”Ÿç‰©ç»„åˆ†æçš„è¿‡ç¨‹æ˜¯ä»shot-gun å®åŸºå› ç»„æ•°æ®åˆ†æå¼€å§‹çš„ï¼Œåè€Œæ²¡æ€ä¹ˆåšè¿‡ç›¸å¯¹æ›´æ—©æ›´æˆç†Ÿçš„æ‰©å¢å­æµ‹åºçš„ä¸Šæ¸¸åˆ†æğŸ˜‚ï¼Œæœ€è¿‘åˆšå¥½æœ‰éœ€æ±‚ï¼Œè¿˜æ˜¯è‡ªå·±å­¦ä¹ å¹¶æ­å»ºä¸€ä¸‹è‡ªå·±ç”¨çš„æ¯”è¾ƒèˆ’æœçš„æµç¨‹å§ã€‚

æ‰©å¢å­ï¼ˆAmpliconï¼‰åˆ†ææ˜¯ç ”ç©¶å¾®ç”Ÿç‰©ç¾¤è½ç»„æˆä¸åŠŸèƒ½çš„æ ¸å¿ƒæ–¹æ³•ä¹‹ä¸€ï¼Œç‰¹åˆ«æ˜¯åœ¨ç”Ÿæ€å­¦ã€åŒ»å­¦å’Œç¯å¢ƒç§‘å­¦é¢†åŸŸã€‚é€šè¿‡å¯¹ç›®æ ‡åŸºå› åŒºåŸŸï¼ˆå¦‚16S rRNAã€18S rRNAã€ITSï¼‰çš„ç‰¹å¼‚æ€§æ‰©å¢ä¸æµ‹åºï¼Œæ‰©å¢å­åˆ†æèƒ½å¤Ÿé«˜æ•ˆæ­ç¤ºæ ·æœ¬ä¸­å¾®ç”Ÿç‰©çš„å¤šæ ·æ€§å’Œç¾¤è½ç»“æ„ã€‚ç›¸æ¯”å®åŸºå› ç»„æµ‹åºï¼Œæ‰©å¢å­åˆ†æå…·æœ‰æˆæœ¬ä½ã€å¤„ç†æµç¨‹ç®€åŒ–ç­‰ä¼˜åŠ¿ï¼Œé€‚ç”¨äºå¤§è§„æ¨¡æ ·æœ¬ç ”ç©¶ã€‚

æ‰©å¢å­åˆ†æçš„æ ¸å¿ƒåŸç†æ˜¯é€šè¿‡èšåˆé…¶é“¾å¼ååº” (PCR) æŠ€æœ¯ï¼Œå¯¹ç›®æ ‡å¾®ç”Ÿç‰©åŸºå› ç‰‡æ®µè¿›è¡Œç‰¹å¼‚æ€§æ‰©å¢å’Œæµ‹åºï¼Œä»¥è§£æå¾®ç”Ÿç‰©ç¾¤è½ç»„æˆåŠå…¶å¤šæ ·æ€§ã€‚é€šå¸¸é€‰æ‹©çš„ç›®æ ‡ç‰‡æ®µä¸ºå…·æœ‰ä¿å®ˆåŒºåŸŸå’Œé«˜å˜å¼‚åŒºåŸŸçš„åŸºå› ï¼Œå¦‚ç»†èŒçš„16S rRNAåŸºå› ã€çœŸèŒçš„ITSåŒºç­‰ã€‚ä¿å®ˆåŒºåŸŸç”¨ä½œå¼•ç‰©è®¾è®¡çš„é¶ç‚¹ï¼Œé«˜å˜å¼‚åŒºåŸŸç”¨äºåŒºåˆ†ä¸åŒç‰©ç§ã€‚

å…·ä½“æ­¥éª¤åŒ…æ‹¬ï¼š
1. DNA æå–ï¼šä»æ ·æœ¬ä¸­æå–æ€» DNAã€‚
2. ç›®æ ‡åŸºå› æ‰©å¢ï¼šä½¿ç”¨ç‰¹å¼‚æ€§å¼•ç‰©æ‰©å¢ç›®æ ‡åŒºåŸŸã€‚
3. é«˜é€šé‡æµ‹åºï¼šæµ‹åºå¾—åˆ°å¤§é‡æ‰©å¢å­çš„åºåˆ—æ•°æ®ã€‚
4. ç”Ÿç‰©ä¿¡æ¯å­¦åˆ†æï¼šå¯¹åºåˆ—è¿›è¡Œè¿‡æ»¤ã€èšç±»ã€æ³¨é‡Šï¼Œè·å¾—ç‰©ç§åˆ†ç±»å’ŒåŠŸèƒ½ä¿¡æ¯ã€‚

è¿™ç§æ–¹æ³•çš„ä¼˜åŠ¿åœ¨äºå…¶é«˜çµæ•åº¦å’Œç‰¹å¼‚æ€§ï¼Œå¯ä»¥å¿«é€Ÿæ­ç¤ºå¾®ç”Ÿç‰©å¤šæ ·æ€§å¹¶è¿›è¡Œå®šé‡åˆ†æã€‚

### ç»†èŒ  
ç»†èŒæ ¸ç³–ä½“RNA (rRNA) åŒ…æ‹¬ 5S rRNA (120bp)ã€16S rRNA (~1540bp) å’Œ 23S rRNA (~2900bp)ã€‚  
- **5S rRNA**: åºåˆ—çŸ­ï¼Œé—ä¼ ä¿¡æ¯é‡æœ‰é™ï¼Œæ— æ³•ç”¨äºåˆ†ç±»é‰´å®šã€‚  
- **23S rRNA**: åºåˆ—è¿‡é•¿ï¼Œçªå˜ç‡é«˜ï¼Œä¸é€‚ç”¨äºè¿œç¼˜ç§ç±»åˆ†æã€‚  
- **16S rRNA**: ç¨³å®šæ€§é«˜ã€æ‹·è´æ•°å¤šï¼Œçº¦å ç»†èŒRNAæ€»é‡çš„80%ï¼ŒåŸºå› ä¸­åŒ…å«10ä¸ªä¿å®ˆåŒºå’Œ9ä¸ªé«˜å˜å¼‚åŒºï¼Œé€‚åˆç»†èŒåˆ†ç±»æ ‡å‡†ã€‚V3-V4 åŒºå› ç‰¹å¼‚æ€§å’Œæ•°æ®åº“æ”¯æŒæœ€ä½³ï¼Œå¸¸ç”¨äºå¤šæ ·æ€§åˆ†æã€‚  

![](images/16s.webp)

### çœŸæ ¸ç”Ÿç‰©  
çœŸæ ¸å¾®ç”Ÿç‰©çš„ rRNA åŒ…æ‹¬ 5.8Sã€18S å’Œ 28S rRNAï¼Œå…¶ä¸­ 18S rRNA å¸¸ç”¨äºå¤šæ ·æ€§åˆ†æã€‚  
- **18S rRNA**: åŸºå› åºåˆ—åŒ…æ‹¬ä¿å®ˆåŒºå’Œå¯å˜åŒº (V1-V9, æ—  V6)ã€‚å¯å˜åŒºåæ˜ ç‰©ç§å·®å¼‚ï¼ŒV4 åŒºå› åˆ†ç±»æ•ˆæœä½³ã€æ•°æ®åº“æ”¯æŒä¸°å¯Œï¼Œæ˜¯åˆ†æå’Œæ³¨é‡Šçš„æœ€ä½³é€‰æ‹©ã€‚  

### çœŸèŒ  
**ITS (å†…æºè½¬å½•é—´éš”åŒº)** ä½äºçœŸèŒ 18Sã€5.8S å’Œ 28S rRNA åŸºå› ä¹‹é—´ï¼Œç”± ITS1 å’Œ ITS2 ç»„æˆï¼Œåˆ†åˆ«é•¿çº¦ 350 bp å’Œ 400 bpã€‚  
- **ä¿å®ˆæ€§**: åŸºå› ç‰‡æ®µåœ¨ç§å†…ç¨³å®šï¼Œç§é—´å·®å¼‚æ˜¾è‘—ï¼Œé€‚ç”¨äºç»†ç²’åº¦åˆ†ç±»ã€‚  
- **çµæ´»æ€§**: å®¹æ˜“åˆ†æå’Œæ‰©å¢ï¼Œå¹¿æ³›ç”¨äºçœŸèŒç§å±å’ŒèŒæ ªç³»ç»Ÿå‘è‚²ç ”ç©¶ã€‚  

### å¤èŒ  
å¤èŒç”Ÿæ´»äºæç«¯ç¯å¢ƒï¼Œæ˜¯ç”Ÿå‘½æé™çš„ä»£è¡¨ã€‚é€šè¿‡å¼•ç‰©è®¾è®¡ï¼Œ16S rRNA çš„ V4V5 åŒºæ˜¯å¤èŒå¤šæ ·æ€§ç ”ç©¶çš„æ ¸å¿ƒåŒºåŸŸã€‚  
- **Illumina MiSeq**: 519F/915R é€‚ç”¨äº V4V5 æ‰©å¢ã€‚  
- **Roche 454**: 344F/915R è¡¨ç°æ›´ä½³ã€‚  
å¤èŒç ”ç©¶æ­ç¤ºäº†å…¶ç‹¬ç‰¹çš„ç”Ÿç‰©ç‰¹å¾ï¼Œä¿ƒè¿›äº†æç«¯ç”Ÿæ€å­¦å’Œç”Ÿç‰©ç³»ç»Ÿå­¦çš„å‘å±•ã€‚

## Qiime2

**QIIME 2** (Quantitative Insights Into Microbial Ecology 2) æ˜¯ä¸€ä¸ªåŠŸèƒ½å¼ºå¤§çš„å¾®ç”Ÿç‰©ç»„æ•°æ®åˆ†æå¹³å°ï¼Œä¸“ä¸ºä»åŸå§‹ DNA åºåˆ—åˆ°å¯è§£é‡Šçš„ç”Ÿç‰©å­¦ç»“è®ºçš„å…¨è¿‡ç¨‹è€Œè®¾è®¡ã€‚å®ƒæ˜¯ QIIME çš„å‡çº§ç‰ˆæœ¬ï¼Œå…·æœ‰æ›´é«˜çš„çµæ´»æ€§ã€å¯æ‰©å±•æ€§å’Œå¯é‡å¤æ€§ã€‚

ç½‘ç«™ï¼š<https://qiime2.org/>

è¯¦ç»†æ•™ç¨‹ï¼š<https://docs.qiime2.org/>

**ä¸»è¦ç‰¹ç‚¹**ï¼š
1. **æ¨¡å—åŒ–æ¶æ„**ï¼š
   - QIIME 2 ä½¿ç”¨æ’ä»¶ç³»ç»Ÿï¼Œæ”¯æŒä¸åŒçš„åˆ†æä»»åŠ¡ï¼ˆå¦‚æ•°æ®å¯¼å…¥ã€è´¨é‡æ§åˆ¶ã€OTU/ASV åˆ†ç±»ã€å¯è§†åŒ–ç­‰ï¼‰ã€‚
   - å¸¸ç”¨æ’ä»¶åŒ…æ‹¬ `dada2`ï¼ˆå»å™ªï¼‰ã€`feature-classifier`ï¼ˆåˆ†ç±»ï¼‰ã€`diversity`ï¼ˆå¤šæ ·æ€§åˆ†æï¼‰ã€‚

2. **å¯é‡å¤æ€§**ï¼š
   - åˆ†æå·¥ä½œæµè¢«è®°å½•åœ¨ `.qza` å’Œ `.qzv` æ–‡ä»¶ä¸­ï¼ŒåŒ…å«å®Œæ•´çš„å…ƒæ•°æ®å’Œå‚æ•°ä¿¡æ¯ã€‚
   - æ˜“äºåˆ†äº«å’Œå¤ç°ç ”ç©¶ç»“æœã€‚

3. **äº¤äº’å¼å¯è§†åŒ–**ï¼š
   - æ”¯æŒäº¤äº’å¼çš„ç»“æœå¯è§†åŒ–ï¼ˆå¦‚å¤šæ ·æ€§åˆ†æã€ç‰¹å¾åˆ†ç±»ï¼‰ã€‚
   - å¯åœ¨ QIIME 2 View ç½‘ç«™ä¸Šç›´æ¥æŸ¥çœ‹ `.qzv` æ–‡ä»¶ã€‚

4. **å¹¿æ³›æ”¯æŒçš„æ•°æ®ç±»å‹**ï¼š
   - æ”¯æŒå¤šç§æ•°æ®è¾“å…¥æ ¼å¼ï¼Œå¦‚ FASTQã€BIOMã€‚
   - å¯å¤„ç†å•ç«¯æˆ–åŒç«¯åºåˆ—ï¼Œä»¥åŠæ‰©å¢å­ã€å®åŸºå› ç»„ã€å®è½¬å½•ç»„æ•°æ®ã€‚

5. **å¼ºå¤§çš„ç¤¾åŒºå’Œç”Ÿæ€ç³»ç»Ÿ**ï¼š
   - QIIME 2 æ‹¥æœ‰æ´»è·ƒçš„ç”¨æˆ·ç¤¾åŒºå’Œä¸°å¯Œçš„æ–‡æ¡£æ”¯æŒã€‚
   - æ˜“äºä¸å…¶ä»–å·¥å…·ï¼ˆå¦‚ Rã€Pythonã€QIIME 1ã€PICRUStï¼‰é›†æˆã€‚

QIIME 2 çš„åˆ†æè¿‡ç¨‹é€šå¸¸åŒ…æ‹¬ä»¥ä¸‹æ­¥éª¤ï¼š
1. **æ•°æ®å¯¼å…¥**ï¼š
   - å°†åŸå§‹æ•°æ®ï¼ˆå¦‚ FASTQ æ–‡ä»¶ï¼‰å¯¼å…¥ä¸º QIIME 2 çš„ `.qza` æ ¼å¼ã€‚

2. **è´¨é‡æ§åˆ¶ä¸å»å™ª**ï¼š
   - ä½¿ç”¨æ’ä»¶å¦‚ `DADA2` æˆ– `Deblur` è¿›è¡Œåºåˆ—ä¿®å‰ªã€å»å™ªå’Œå»å†—ä½™ã€‚

3. **ç‰¹å¾è¡¨æ„å»º**ï¼š
   - ç”Ÿæˆ ASV è¡¨ï¼ˆæˆ– OTU è¡¨ï¼‰å’Œä»£è¡¨æ€§åºåˆ—ã€‚

4. **åºåˆ—åˆ†ç±»**ï¼š
   - åˆ©ç”¨åˆ†ç±»å™¨ï¼ˆå¦‚ SILVAã€Greengenesã€UNITEï¼‰å¯¹ç‰¹å¾åºåˆ—è¿›è¡Œç‰©ç§æ³¨é‡Šã€‚

5. **å¤šæ ·æ€§åˆ†æ**ï¼š
   - è®¡ç®— Î± å¤šæ ·æ€§ï¼ˆç‰©ç§ä¸°åº¦ï¼‰å’Œ Î² å¤šæ ·æ€§ï¼ˆæ ·æœ¬é—´å·®å¼‚ï¼‰ï¼Œå¹¶è¿›è¡Œç»Ÿè®¡åˆ†æã€‚

6. **å¯è§†åŒ–**ï¼š
   - ç”Ÿæˆå›¾å½¢åŒ–ç»“æœï¼Œå¦‚ PCA å›¾ã€ä¸°åº¦çƒ­å›¾ã€åˆ†ç±»é¥¼å›¾ç­‰ã€‚

**ä¼˜ç‚¹**ï¼š
- é«˜åº¦å¯æ‰©å±•ï¼Œæ”¯æŒå¤šç§å¾®ç”Ÿç‰©ç»„åˆ†æã€‚
- å·¥ä½œæµé€æ˜ï¼Œæ–¹ä¾¿é‡å¤åˆ†æå’Œç»“æœå…±äº«ã€‚
- æ’ä»¶ç”Ÿæ€ä¸°å¯Œï¼Œå¯æ ¹æ®éœ€è¦é€‰æ‹©ä¸åŒçš„å·¥å…·ã€‚

**ç¼ºç‚¹**ï¼š
- å­¦ä¹ æ›²çº¿è¾ƒé™¡ï¼Œéœ€ç†Ÿæ‚‰å‘½ä»¤è¡Œæ“ä½œã€‚
- å¯¹è®¡ç®—èµ„æºéœ€æ±‚è¾ƒé«˜ï¼Œå¤„ç†å¤§è§„æ¨¡æ•°æ®å¯èƒ½è¾ƒæ…¢ã€‚

QIIME 2 æ˜¯ç°ä»£å¾®ç”Ÿç‰©ç»„ç ”ç©¶çš„æ ¸å¿ƒå·¥å…·ä¹‹ä¸€ï¼Œå°¤å…¶é€‚åˆå¤§è§„æ¨¡æ‰©å¢å­æ•°æ®çš„å¤„ç†å’Œåˆ†æã€‚å®‰è£…çš„è¯å°½é‡ç”¨condaç›´æ¥è£…ä¸‹æ•´ä¸ªç¯å¢ƒï¼Œç„¶åæ•´ä¸ªQiime2åˆ†æå¹³å°å¤ªå¤šå¤ªå¤æ‚äº†ï¼Œåƒå¤šæ ·æ€§åˆ†æï¼Œå¯è§†åŒ–ç­‰æ­¥éª¤éƒ½å¯ä»¥åç»­åœ¨Rè¯­è¨€é‡Œæ›´çµæ´»åœ°å®ç°ï¼Œæ²¡å¿…è¦ä¸€å®šç”¨Qiime2ã€‚æˆ‘ä»¬æ˜ç¡®ä¸€ä¸‹æˆ‘ä»¬çš„è¾“å…¥å’Œè¾“å‡ºç›®æ ‡ï¼ˆè·å¾—feature-tableï¼Œfeatureçš„ç‰©ç§æ³¨é‡Šç»“æœï¼Œfeatureçš„ç³»ç»Ÿå‘è‚²æ ‘åŸºæœ¬å°±å¤Ÿäº†ï¼‰ï¼Œè¿™é‡Œåªç”¨ä¸‹é¢æœ€å…³é”®çš„å‡ ä¸ªæ­¥éª¤å°±è¡Œï¼š

## Input data

æµ‹åºå…¬å¸ä¸€èˆ¬è¿”å›çš„æ˜¯æŒ‰ç…§æ¯ä¸ªæ ·æœ¬æ‹†åˆ†å¥½çš„åŒç«¯fqæ–‡ä»¶ï¼Œæˆ–è€…å•ç«¯fqæ–‡ä»¶ï¼Œæˆ–è€…å·²ç»åšäº†ç®€å•è´¨æ§å’Œæ‹¼æ¥åçš„å•ä¸ªfqæ–‡ä»¶ï¼ˆå¯ä»¥å½“ä½œå•ç«¯fqæ–‡ä»¶ï¼‰ï¼Œæˆ‘ä»¬ä»è¿™é‡Œå¼€å§‹è¿›è¡Œåˆ†æã€‚

metadata.tsvçš„è¯æ˜¯æˆ‘ä»¬çš„å®éªŒè®¾è®¡ï¼ˆç¬¬ä¸€åˆ—SampleIDï¼Œå¯ä»¥æœ‰ä¸€äº›å®éªŒåˆ†ç»„çš„åˆ—ï¼‰ï¼Œæƒ³æçš„è¯å¯ä»¥æä¸€ä¸ªï¼Œåç»­åŸºäºQiimeå¯è§†åŒ–çš„è¯å¯ä»¥ç”¨ä¸Šï¼Œæ²¡æœ‰çš„è¯ä¹Ÿæ²¡å…³ç³»ã€‚

### åŒç«¯fqæ–‡ä»¶

éœ€è¦å…ˆå‡†å¤‡ä¸€ä¸ªä¸€ä¸ªåˆ¶è¡¨ç¬¦åˆ†éš”çš„ï¼ˆå³ .tsvï¼‰æ–‡æœ¬æ–‡ä»¶manifestã€‚ç¬¬ä¸€åˆ—å®šä¹‰æ ·æœ¬ IDï¼Œè€Œç¬¬äºŒåˆ—ï¼ˆå’Œå¯é€‰çš„ç¬¬ä¸‰åˆ—ï¼‰å®šä¹‰æ­£å‘ï¼ˆå’Œå¯é€‰åå‘ï¼‰è¯»å–çš„ç»å¯¹æ–‡ä»¶è·¯å¾„ï¼š

```
sample-id     forward-absolute-filepath       reverse-absolute-filepath
sample-1      $PWD/some/filepath/sample1_R1.fastq.gz  $PWD/some/filepath/sample1_R2.fastq.gz
sample-2      $PWD/some/filepath/sample2_R1.fastq.gz  $PWD/some/filepath/sample2_R2.fastq.gz
sample-3      $PWD/some/filepath/sample3_R1.fastq.gz  $PWD/some/filepath/sample3_R2.fastq.gz
sample-4      $PWD/some/filepath/sample4_R1.fastq.gz  $PWD/some/filepath/sample4_R2.fastq.gz
```

ç„¶åæ•°æ®å¯¼å…¥qiime2ï¼Œè¿™é‡Œå‡è®¾readsæ ¼å¼ä¸ºåŒç«¯33æ ¼å¼ï¼ˆPairedEndFastqManifestPhred33V2ï¼‰

```{bash}
qiime tools import \
  --type 'SampleData[PairedEndSequencesWithQuality]' \
  --input-path manifest \
  --output-path demux.qza \
  --input-format PairedEndFastqManifestPhred33V2
```

### å•ç«¯fqæ–‡ä»¶

åŒæ ·éœ€è¦å‡†å¤‡manifestæ–‡ä»¶ï¼š
```
sample-id     absolute-filepath
sample-1      $PWD/some/filepath/sample1_R1.fastq
sample-2      $PWD/some/filepath/sample2_R1.fastq
```

ç„¶åæ•°æ®å¯¼å…¥qiime2ï¼Œè¿™é‡Œå‡è®¾readsæ ¼å¼ä¸ºå•ç«¯33æ ¼å¼ï¼ˆSingleEndFastqManifestPhred33V2ï¼‰

```{bash}
qiime tools import \
  --type 'SampleData[PairedEndSequencesWithQuality]' \
  --input-path manifest \
  --output-path demux.qza \
  --input-format SingleEndFastqManifestPhred33V2
```

### å…¶ä»–ç±»å‹

å…¶ä»–ç±»å‹æ•°æ®çš„å¯¼å…¥æ–¹å¼å¯ä»¥å‚è€ƒ<https://docs.qiime2.org/2024.10/tutorials/importing/>ï¼Œä¸»è¦æ˜¯æœªæ‹†åˆ†çš„fqæ–‡ä»¶ï¼Œå°±æ˜¯è¯´æ‰€æœ‰æ ·æœ¬éƒ½åœ¨ä¸€ä¸ªfqæ–‡ä»¶é‡Œï¼Œä½†æ˜¯ä¸åŒæ ·æœ¬å…·æœ‰ä¸åŒçš„barcodeå¯ä»¥ç”¨æ¥åŒºåˆ†ã€‚

è¿™ç§çš„è¯éœ€è¦æ–°å»ºä¸€ä¸ªmuxed-se-barcode-in-seqæ–‡ä»¶å¤¹ï¼ŒæŠŠå¤šæ ·æœ¬çš„ä¸€ä¸ªfqæ–‡ä»¶å’Œå¯¹åº”metadataæ–‡ä»¶æ”¾åœ¨æ–‡ä»¶å¤¹é‡Œï¼Œmetadataå…¶ä¸­åŒ…å«ä¸€åˆ—ç”¨äº FASTQ è§£å¤ç”¨ï¼ˆdemultiplexingï¼‰çš„æ¯ä¸ªæ ·æœ¬barcodeï¼ˆæˆ–ä¸¤åˆ—åŒç´¢å¼•æ¡å½¢ç ï¼‰

```{bash}
qiime tools import \
  --type MultiplexedPairedEndBarcodeInSequence \
  --input-path muxed-pe-barcode-in-seq \
  --output-path multiplexed-seqs.qza
```


## ASV/OTU

**ASV**ï¼ˆAmplicon Sequence Variantï¼‰å’Œ **OTU**ï¼ˆOperational Taxonomic Unitï¼‰æ˜¯å¾®ç”Ÿç‰©ç»„ç ”ç©¶ä¸­ç”¨äºè¡¨å¾æ ·æœ¬ä¸­å¾®ç”Ÿç‰©åˆ†ç±»å•å…ƒçš„ä¸¤ä¸ªæ ¸å¿ƒæ¦‚å¿µã€‚å…·ä½“å¯ä»¥å‚è€ƒå¸ˆå§ä»¥å‰å†™çš„ä»‹ç»<https://www.jianshu.com/p/f31581bbfb80>ã€‚

**1. OTU**
- **å®šä¹‰**ï¼šOTU æ˜¯é€šè¿‡å¯¹æ‰©å¢å­åºåˆ—è¿›è¡Œèšç±»åå®šä¹‰çš„åˆ†ç±»å•å…ƒã€‚é€šå¸¸åŸºäºåºåˆ—é—´çš„ç›¸ä¼¼æ€§ï¼ˆå¦‚ 97% æˆ– 99% ç›¸ä¼¼æ€§ï¼‰å°†åºåˆ—å½’ä¸ºåŒä¸€ OTUã€‚
- **ç‰¹ç‚¹**ï¼š
  - å¸¸ç”¨èšç±»æ–¹æ³•å¦‚ UCLUSTã€VSEARCHã€USEARCHã€‚
  - éœ€è¦å‚è€ƒæ•°æ®åº“æˆ– **de novo** èšç±»ï¼ˆæ— å‚è€ƒï¼‰ã€‚
  - OTU è¡¨å¯èƒ½å­˜åœ¨å†—ä½™ï¼ˆå› èšç±»è¯¯å·®å¯¼è‡´éƒ¨åˆ†ä¸ä¸€è‡´ï¼‰ã€‚

**2. ASV**
- **å®šä¹‰**ï¼šASV æ˜¯é€šè¿‡ä¸¥æ ¼çš„å»å™ªç®—æ³•ï¼ˆå¦‚ DADA2 æˆ– Deblurï¼‰ç²¾ç¡®å®šä¹‰çš„å•ä¸ªæ ¸è‹·é…¸çº§åˆ«å˜å¼‚çš„åºåˆ—å•å…ƒã€‚
- **ç‰¹ç‚¹**ï¼š
  - ä¸ä¾èµ–ç›¸ä¼¼æ€§é˜ˆå€¼ï¼Œç›´æ¥åŸºäºåºåˆ—ç²¾ç¡®å˜å¼‚ã€‚
  - ç”Ÿæˆçš„åºåˆ—è¡¨æ›´åŠ çœŸå®å’Œç²¾ç»†ï¼Œå‡å°‘äº†å› èšç±»å¯¼è‡´çš„ä¿¡æ¯ä¸¢å¤±ã€‚
  - å»é™¤äº†æµ‹åºè¯¯å·®ï¼Œä¿ç•™äº†çœŸå®ç”Ÿç‰©å­¦å˜å¼‚ã€‚


| **ç‰¹æ€§**               | **ASV**                           | **OTU**                           |
|------------------------|------------------------------------|------------------------------------|
| **å®šä¹‰**              | ç²¾ç¡®çš„åºåˆ—å˜å¼‚å•ä½                | ç›¸ä¼¼æ€§èšç±»çš„å•ä½                  |
| **åˆ†ææ–¹æ³•**          | å»å™ªç®—æ³•ï¼ˆDADA2ã€Deblurï¼‰         | èšç±»ç®—æ³•ï¼ˆVSEARCHã€UCLUSTï¼‰       |
| **åˆ†è¾¨ç‡**            | æ ¸è‹·é…¸çº§åˆ«                       | ä¾èµ–è®¾å®šçš„ç›¸ä¼¼æ€§é˜ˆå€¼ï¼ˆå¦‚ 97%ï¼‰   |
| **è¯¯å·®å¤„ç†**          | å»å™ªæ¨¡å‹ç²¾å‡†å»é™¤æµ‹åºè¯¯å·®          | èšç±»å¯èƒ½ä¿ç•™éƒ¨åˆ†æµ‹åºè¯¯å·®          |
| **å¯¹æ•°æ®è´¨é‡è¦æ±‚**    | è¾ƒé«˜                              | è¾ƒä½                              |

ASV æ˜¯ç°ä»£æ‰©å¢å­åˆ†æçš„ä¸»æµæ–¹æ³•ï¼Œå…¶ç²¾ç¡®æ€§å’Œå¯æ¯”è¾ƒæ€§æ˜¾è‘—ä¼˜äº OTUã€‚OTU åœ¨ä¸€äº›ç‰¹å®šåœºæ™¯ä»æœ‰ä½¿ç”¨ä»·å€¼ï¼Œä½†é€æ¸è¢« ASV æ‰€å–ä»£ã€‚

### ASV

åœ¨ QIIME 2 ä¸­ï¼Œå»å™ªï¼ˆdenoisingï¼‰æ˜¯é‡è¦æ­¥éª¤ï¼Œç”¨äºå»é™¤æµ‹åºè¯¯å·®å¹¶ç”Ÿæˆé«˜åˆ†è¾¨ç‡çš„ **ASVï¼ˆAmplicon Sequence Variantsï¼‰**ã€‚ç›®å‰æ”¯æŒä¸¤ç§ä¸»æµçš„å»å™ªæ–¹æ³•ï¼š**DADA2** å’Œ **Deblur**ã€‚**DADA2** ä½¿ç”¨åŸºäºé”™è¯¯æ¨¡å‹çš„ç®—æ³•ï¼Œé€šè¿‡å­¦ä¹ æµ‹åºè¯¯å·®åˆ†å¸ƒï¼Œç²¾ç¡®æ ¡æ­£åºåˆ—é”™è¯¯å¹¶å»é™¤åµŒåˆä½“ï¼›å®ƒé€‚åˆäºå•ç«¯å’ŒåŒç«¯æ•°æ®ï¼Œèƒ½å¤Ÿç›´æ¥è¾“å‡ºæ— é”™è¯¯çš„ ASV è¡¨ã€‚**Deblur** åˆ™æ˜¯åŸºäºå‚è€ƒåºåˆ—çš„å»å™ªæ–¹æ³•ï¼Œé€šè¿‡å¯¹åºåˆ—è¿›è¡Œæˆªæ–­å’Œå½’ä¸€åŒ–æ“ä½œï¼Œå¿«é€Ÿå»é™¤ä½è´¨é‡åºåˆ—åŠå™ªéŸ³ï¼Œé€‚åˆè´¨é‡å·²ä¼˜åŒ–çš„å•ç«¯æ•°æ®ã€‚ä¸¤è€…å‡èƒ½é«˜æ•ˆç”Ÿæˆé«˜åˆ†è¾¨ç‡çš„ ASVï¼Œä½† DADA2 æ›´æ³¨é‡é”™è¯¯æ ¡æ­£ï¼Œè€Œ Deblur åœ¨é€Ÿåº¦ä¸Šæ›´å…·ä¼˜åŠ¿ã€‚

```{bash}
# é€‰æ‹©1ï¼ŒDADA2ï¼Œä¸ä¸€å®šè¦trimå’Œtruncï¼Œçœ‹è‡ªå·±æ•°æ®ï¼Œè¿™é‡Œæ˜¯å•ç«¯æ•°æ®ï¼ŒåŒç«¯ç¨å¾®æ”¹ä¸€ä¸‹å‚æ•°
qiime dada2 denoise-single \
  --i-demultiplexed-seqs demux.qza \
  --p-trim-left 0 \
  --p-trunc-len 120 \
  --o-representative-sequences rep-seqs-dada2.qza \
  --o-table table-dada2.qza \
  --o-denoising-stats stats-dada2.qza

# æ”¹åï¼Œæ”¾å…¥ä¸»æµç¨‹
mv rep-seqs-dada2.qza rep-seqs.qza
mv table-dada2.qza table.qza
```

```{bash}
# é€‰æ‹©2ï¼Œdeblurï¼Œä¸ä¸€å®šè¦trimå’Œtruncï¼Œçœ‹è‡ªå·±æ•°æ®ï¼Œè¿™é‡Œæ˜¯å•ç«¯æ•°æ®ï¼ŒåŒç«¯ç¨å¾®æ”¹ä¸€ä¸‹å‚æ•°
qiime quality-filter q-score \
  --i-demux demux.qza \
  --o-filtered-sequences demux-filtered.qza \
  --o-filter-stats demux-filter-stats.qza
 
qiime deblur denoise-16S \
  --i-demultiplexed-seqs demux-filtered.qza \
  --p-trim-length 120 \
  --o-representative-sequences rep-seqs-deblur.qza \
  --o-table table-deblur.qza \
  --p-sample-stats \
  --o-stats deblur-stats.qza

# æ”¹åï¼Œæ”¾å…¥ä¸»æµç¨‹
mv rep-seqs-deblur.qza rep-seqs.qza
mv table-deblur.qza table.qza
```

### OTU

åœ¨ QIIME 2 ä¸­ï¼Œåºåˆ—èšç±»ç”¨äºæ ¹æ®ç›¸ä¼¼æ€§å¯¹æ‰©å¢å­åºåˆ—è¿›è¡Œåˆ†ç»„ï¼Œç”Ÿæˆ **OTUï¼ˆOperational Taxonomic Unitsï¼‰** è¡¨ã€‚ç›®å‰æ”¯æŒä¸‰ç§ä¸»è¦èšç±»ç­–ç•¥ï¼š**De novo èšç±»** å®Œå…¨åŸºäºæ ·æœ¬åºåˆ—é—´çš„ç›¸ä¼¼æ€§ï¼Œæ— éœ€å‚è€ƒæ•°æ®åº“ï¼Œé€‚åˆç ”ç©¶æœªè¢«å……åˆ†è¡¨å¾çš„å¾®ç”Ÿç‰©ç¾¤è½ï¼›**å°é—­å¼•ç”¨èšç±»** ä»…å°†åºåˆ—ä¸å‚è€ƒæ•°æ®åº“è¿›è¡Œæ¯”å¯¹ï¼ŒæœªåŒ¹é…çš„åºåˆ—ä¼šè¢«ä¸¢å¼ƒï¼Œé€‚åˆæ ‡å‡†åŒ–åˆ†æå’Œä¸å·²æœ‰æ•°æ®çš„æ¯”è¾ƒï¼›**å¼€æ”¾å¼•ç”¨èšç±»** åˆ™ç»“åˆä¸¤è€…ä¼˜åŠ¿ï¼Œå…ˆå¯¹åºåˆ—ä¸å‚è€ƒæ•°æ®åº“æ¯”å¯¹ï¼Œå†å¯¹æœªåŒ¹é…åºåˆ—è¿›è¡Œ de novo èšç±»ï¼Œé€‚åˆå…¼é¡¾æ¢ç´¢æ€§ç ”ç©¶å’Œæ•°æ®åº“å‚è€ƒåˆ†æçš„åœºæ™¯ã€‚**æ³¨æ„ï¼Œä¼˜å…ˆä½¿ç”¨ASVå§ã€‚**

```{bash}
# å¦‚æœinputæ•°æ®æ˜¯fqæ–‡ä»¶ï¼Œéœ€è¦å…ˆæƒ³åŠæ³•è´¨æ§åè½¬åŒ–ä¸ºfastaæ–‡ä»¶ï¼Œ
# å¯ä»¥å‚è€ƒhttps://docs.qiime2.org/2019.1/tutorials/read-joining/
# å› ä¸ºOTUèšç±»éœ€è¦è¾“å…¥fastaæ–‡ä»¶
qiime tools import \
  --input-path seqs.fna \
  --output-path seqs.qza \
  --type 'SampleData[Sequences]'
  
# å¯¼å…¥æ•°æ®åï¼Œç¬¬ä¸€æ­¥å¯ä»¥ä½¿ç”¨ dereplicate-sequences å‘½ä»¤å»é‡å¤ã€‚
qiime vsearch dereplicate-sequences \
  --i-sequences seqs.qza \
  --o-dereplicated-table table.qza \
  --o-dereplicated-sequences rep-seqs.qza

# é€‰æ‹©1ï¼ŒDe novo clustering
qiime vsearch cluster-features-de-novo \
  --i-table table.qza \
  --i-sequences rep-seqs.qza \
  --p-perc-identity 0.99 \
  --o-clustered-table table-dn-99.qza \
  --o-clustered-sequences rep-seqs-dn-99.qza

# é€‰æ‹©2ï¼ŒClosed-reference clustering
qiime vsearch cluster-features-closed-reference \
  --i-table table.qza \
  --i-sequences rep-seqs.qza \
  --i-reference-sequences 85_otus.qza \
  --p-perc-identity 0.85 \
  --o-clustered-table table-cr-85.qza \
  --o-clustered-sequences rep-seqs-cr-85.qza \
  --o-unmatched-sequences unmatched-cr-85.qza

# é€‰æ‹©3ï¼ŒOpen-reference clustering
qiime vsearch cluster-features-open-reference \
  --i-table table.qza \
  --i-sequences rep-seqs.qza \
  --i-reference-sequences 85_otus.qza \
  --p-perc-identity 0.85 \
  --o-clustered-table table-or-85.qza \
  --o-clustered-sequences rep-seqs-or-85.qza \
  --o-new-reference-sequences new-ref-seqs-or-85.qza
```


### å»ºæ ‘

```{bash}
### æ„å»ºè¿›åŒ–æ ‘ï¼Œå¯ä»¥ç”¨æ¥å±•ç¤ºç‰©ç§ï¼Œä¹Ÿå¯ä»¥è®¡ç®—åç»­çš„åŸºäºå‘è‚²è·ç¦»çš„alphaå’Œbetaå¤šæ ·æ€§
qiime phylogeny align-to-tree-mafft-fasttree \
  --i-sequences rep-seqs.qza \
  --o-alignment aligned-rep-seqs.qza \
  --o-masked-alignment masked-aligned-rep-seqs.qza \
  --o-tree unrooted-tree.qza \
  --o-rooted-tree rooted-tree.qza
```

ç„¶åè¿˜å¯ä»¥åšä¸€ä¸ªç®€å•çš„summaryå’Œå¯è§†åŒ–ï¼š

```{bash}
qiime feature-table summarize \
  --i-table table.qza \
  --o-visualization table.qzv \
  --m-sample-metadata-file metadata.tsv
qiime feature-table tabulate-seqs \
  --i-data rep-seqs.qza \
  --o-visualization rep-seqs.qzv
```

è¿™äº›qzvæ–‡ä»¶éƒ½å¯ä»¥ç›´æ¥æ‹–åŠ¨åˆ°<https://view.qiime2.org/>ç½‘ç«™æ¥æŸ¥çœ‹å¯è§†åŒ–ç»“æœã€‚

## ç‰©ç§æ³¨é‡Š

### æ•°æ®åº“æ³¨é‡Š

å¸¸ç”¨çš„[æ‰©å¢å­æ•°æ®åº“](../amplicon)ä¸Šæ¬¡ä»‹ç»è¿‡äº†ï¼Œæ¶‰åŠåˆ°16S rRNAåŸºå› çš„åºåˆ—æ•°æ®åº“æ—¶ï¼Œæœ‰ä¸‰ä¸ªä¸»è¦çš„æ•°æ®åº“æ˜¯å¸¸ç”¨çš„ï¼š**Greengenes**ã€**SILVA** å’Œ **RDP**ã€‚**UNITEæ•°æ®åº“**æ˜¯ç”¨äºçœŸèŒé‰´å®šå’Œå¤šæ ·æ€§æ£€æµ‹çš„ä¸»è¦markeråŸºå› æ•°æ®åº“ã€‚å…·ä½“ä¿¡æ¯æ¯ä¸ªæ•°æ®åº“ä¸»é¡µéƒ½æœ‰å†™ï¼Œæˆ‘ä»¬æ‹¿æ¥ç”¨çš„è¯å…³é”®å°±æ˜¯ä¸¤ä¸ªæ–‡ä»¶ï¼Œä¸€ä¸ªæ˜¯æ•°æ®åº“æä¾›çš„åºåˆ—fastaæ–‡ä»¶ï¼Œå¦ä¸€ä¸ªæ˜¯è¿™äº›åºåˆ—å¯¹åº”çš„taxonomyæ–‡ä»¶ã€‚

æ‹¿ä¸€ä¸ªUNITEæ•°æ®åº“åšä¾‹å­ï¼ˆç”¨æ¥çœŸèŒITSæ¯”è¾ƒå¤šçš„ï¼‰ï¼Œå…ˆå»å®˜ç½‘<https://unite.ut.ee/repository.php>æ‰¾åˆ°å¯¹åº”çš„èµ„æºä¸‹è½½ï¼ˆæ³¨æ„ç‰ˆæœ¬å·ç­‰ä¿¡æ¯ï¼‰ã€‚

è¯»å…¥æ•°æ®åº“ï¼š
```{bash}
# å¯¼å…¥å‚è€ƒåºåˆ—
qiime tools import \
  --type FeatureData[Sequence] \
  --input-path sh_refs_qiime_ver9_99_s_all_25.07.2023.fasta \
  --output-path unite-ver9-seqs_99_25.07.2023.qza
# å¯¼å…¥ç‰©ç§åˆ†ç±»ä¿¡æ¯
qiime tools import \
  --type FeatureData[Taxonomy] \
  --input-path sh_taxonomy_qiime_ver9_99_s_all_25.07.2023.txt \
  --output-path unite-ver9-taxonomy_99_25.07.2023.qza \
  --input-format HeaderlessTSVTaxonomyFormat
```

æ³¨é‡Šçš„è¯ï¼Œåœ¨ QIIME 2 ä¸­ï¼Œ`feature-classifier` æ’ä»¶æä¾›å¤šç§æ–¹æ³•ç”¨äºæ‰©å¢å­åºåˆ—çš„åˆ†ç±»æ³¨é‡Šï¼Œå¸®åŠ©è¯†åˆ«å¾®ç”Ÿç‰©ç‰©ç§æˆ–æ›´é«˜åˆ†ç±»çº§åˆ«ã€‚å¸¸ç”¨æ–¹æ³•åŒ…æ‹¬ï¼š**æœ´ç´ è´å¶æ–¯ï¼ˆNaive bayesï¼‰åˆ†ç±»å™¨**ï¼Œåˆ©ç”¨é¢„è®­ç»ƒçš„å‚è€ƒæ•°æ®åº“ï¼ˆå¦‚ Silva æˆ– Greengenesï¼‰å¯¹åºåˆ—è¿›è¡Œæ¦‚ç‡æ€§åˆ†ç±»ï¼Œæ˜¯æœ€å¸¸ç”¨çš„åˆ†ç±»æ–¹æ³•ï¼›**vsearch-based consensus åˆ†ç±»**ï¼Œé€šè¿‡ä¸å‚è€ƒæ•°æ®åº“çš„åºåˆ—æ¯”å¯¹ï¼ŒåŸºäºç›¸ä¼¼æ€§å¾—å‡ºå…±è¯†åˆ†ç±»ï¼Œé€‚åˆç²¾ç¡®åŒ¹é…éœ€æ±‚ï¼›**BLAST+ åˆ†ç±»**ï¼Œä½¿ç”¨ BLAST ç®—æ³•ä¸å‚è€ƒåºåˆ—æ¯”å¯¹ï¼Œèƒ½å¤Ÿæä¾›çµæ´»çš„åŒ¹é…æ§åˆ¶ã€‚æ­¤å¤–ï¼Œè¿˜æ”¯æŒè®­ç»ƒè‡ªå®šä¹‰åˆ†ç±»å™¨ï¼Œé€‚ç”¨äºç‰¹å®šç ”ç©¶éœ€æ±‚ã€‚å„æ–¹æ³•åœ¨é€Ÿåº¦ã€çµæ•åº¦å’Œå¯¹å‚è€ƒæ•°æ®åº“çš„ä¾èµ–ä¸Šå„æœ‰ä¼˜åŠ£ï¼Œç”¨æˆ·å¯æ ¹æ®æ•°æ®ç‰¹ç‚¹å’Œç ”ç©¶ç›®æ ‡çµæ´»é€‰æ‹©ã€‚

```{bash}
# é€‰æ‹©1ï¼Œæœ´ç´ è´å¶æ–¯åˆ†ç±»å™¨
# æ³¨æ„ï¼Œé¦–å…ˆè¦æœ‰è®­ç»ƒå¥½çš„åˆ†ç±»å™¨qzaæ–‡ä»¶æ‰èƒ½ç”¨ï¼Œå¯ä»¥å‚è€ƒä¸‹ä¸€èŠ‚è®­ç»ƒåˆ†ç±»å™¨
qiime feature-classifier classify-sklearn \
  --i-classifier unite-ver9-99-classifier-25.07.2023.qza \
  --i-reads rep-seqs.qza \
  --o-classification taxonomy.qza
  
# é€‰æ‹©2ï¼Œvsearch
qiime feature-classifier classify-consensus-vsearch \
  --i-query rep-seqs.qza \
  --i-reference-reads unite-ver9-seqs_99_25.07.2023.qza \
  --i-reference-taxonomy unite-ver9-taxonomy_99_25.07.2023.qza \
  --o-classification taxonomy_vsearch.qza
  
# é€‰æ‹©3ï¼Œblast
qiime feature-classifier classify-consensus-blast \
  --i-query rep-seqs.qza \
  --i-reference-reads unite-ver9-seqs_99_25.07.2023.qza \
  --i-reference-taxonomy unite-ver9-taxonomy_99_25.07.2023.qza \
  --o-classification taxonomy_blast.qza
```

ç®€å•ç”»ä¸ªç‰©ç§å †ç§¯æŸ±å½¢å›¾ï¼Œåç»­å¯¼å‡ºæ•°æ®å†Ré‡Œç”»ä¹Ÿè¡Œã€‚ä¸Šé¢å‡ ç§æ–¹æ³•çš„ç‰©ç§æ³¨é‡Šç»“æœæœ‰å·®å¼‚ï¼Œåœ¨åˆé€‚æƒ…å†µä¸‹è¿˜æ˜¯é€‰æœ€å¸¸ç”¨çš„ç¬¬ä¸€ä¸ªå§ã€‚
```{bash}
# å¯è§†åŒ–ç‰©ç§æ³¨é‡Š
qiime metadata tabulate \
  --m-input-file taxonomy.qza \
  --o-visualization taxonomy.qzv
  
# å †å æŸ±çŠ¶å›¾å±•ç¤º
qiime taxa barplot \
  --i-table table.qza \
  --i-taxonomy taxonomy.qza \
  --m-metadata-file metadata.txt \
  --o-visualization taxa-bar-plots.qzv
```


### è®­ç»ƒåˆ†ç±»å™¨

æ³¨æ„ï¼Œè¿™ä¸ªæ­¥éª¤ä¸€èˆ¬éœ€è¦è¾ƒå¤§å†…å­˜ï¼ˆçœ‹æ•°æ®åº“ï¼ŒUNITEè¿™ä¸ªæˆ‘æ¶ˆè€—äº†100Gå·¦å³å†…å­˜ï¼‰ï¼Œè¿è¡Œæ—¶é—´ä¹Ÿæ¯”è¾ƒä¹…ï¼Œ10å¤šä¸ªå°æ—¶
```{bash}
qiime feature-classifier fit-classifier-naive-bayes \
  --i-reference-reads unite-ver9-seqs_99_25.07.2023.qza \
  --i-reference-taxonomy unite-ver9-taxonomy_99_25.07.2023.qza \
  --o-classifier unite-ver9-99-classifier-25.07.2023.qza
```

æ‰€ä»¥å°½å¯èƒ½å¯ä»¥å»æ‰¾åˆ«äººè®­ç»ƒå¥½çš„ç»“æœï¼ˆå› ä¸ºåºåˆ—æ˜¯ä¸€æ ·çš„ï¼Œæ‹¿æ¥ç”¨å°±è¡Œï¼Œä½†æ˜¯è¦æ³¨æ„æ˜¯åœ¨è·Ÿä½ åŒç‰ˆæœ¬çš„Qiime2ç¯å¢ƒä¸‹è®­ç»ƒçš„ï¼Œä¸ç„¶å¯èƒ½ç”¨ä¸äº†ï¼‰

è¿™é‡Œæ•´ç†äº†ä¸€äº›ä¸‹è½½è·¯å¾„ï¼Œå¯ä»¥å°è¯•ç”¨ç”¨ï¼Œæ¯•ç«Ÿä¸ä¸€å®šæ¯ä¸ªäººéƒ½æœ‰è¶³å¤Ÿçš„å†…å­˜å’Œæ—¶é—´æ¥è·‘è¿™ä¸ªåˆ†ç±»å™¨ï¼š

- UNITE (ITS)ï¼š<https://github.com/colinbrislawn/unite-train/releases>
- Greengenes (16S rRNA)ï¼š<http://ftp.microbio.me/greengenes_release/2022.10/sklearn-1.4.2-compatible-nb-classifiers/>
- Silva (16S/18S rRNA): <https://anw-sh.github.io/silva-138_classifiers/>
- RDP (16S/28S rRNA): <https://sourceforge.net/projects/rdp-classifier/>

## Export data

qzaï¼Œqzvæ–‡ä»¶æ˜¯Qiime2çš„æ ‡å‡†æ ¼å¼ï¼Œè¿˜æ˜¯å¯¼å‡ºæˆæ™®é€šçš„æ–‡æœ¬æ–‡ä»¶ï¼ˆè¡¨æ ¼ç­‰ï¼‰ç”¨æ¥åç»­åˆ«çš„å¹³å°åˆ†æå§ï¼š

```{bash}
# å¯¼å‡ºfeatureä¸°åº¦è¡¨
qiime tools export \
  --input-path feature-table.qza \
  --output-path exported-feature-table
# å¯¼å‡ºåºåˆ—ï¼Œåœ¨rep-seqsæ–‡ä»¶å¤¹ä¸‹
qiime tools export \
  --input-path rep-seqs.qza \
  --output-path rep-seqs
# å¯¼å‡ºç³»ç»Ÿå‘è‚²æ ‘
qiime tools export \
  --input-path unrooted-tree.qza \
  --output-path exported-tree
```

## åŠŸèƒ½é¢„æµ‹

16Sæµ‹åºçš„ä¸€å¤§ç¼ºç‚¹å°±æ˜¯åªèƒ½çš„åˆ°ç‰©ç§ä¿¡æ¯ï¼Œç¼ºä¹åŠŸèƒ½ä¿¡æ¯ï¼Œä½¿ç”¨PICRUStè½¯ä»¶è¿›è¡Œ16SåŠŸèƒ½é¢„æµ‹åˆ†æï¼ˆé¢„æµ‹ï¼Œç¨å¾®å¼¥è¡¥è¿™ä¸€ç¼ºç‚¹ï¼‰ã€‚
2018å¹´æ¨å‡ºäº†å…¨æ–°ç‰ˆæœ¬çš„PICRUStï¼Œå³PICRUSt2<https://github.com/picrust/picrust2>ã€‚

PICRUSt2 (Phylogenetic Investigation of Communities by Reconstruction of Unobserved States)æ˜¯ä¸€æ¬¾åŸºäºæ ‡è®°åŸºå› åºåˆ—æ¥é¢„æµ‹åŠŸèƒ½ä¸°åº¦çš„è½¯ä»¶ã€‚

â€œåŠŸèƒ½â€é€šå¸¸æŒ‡çš„æ˜¯åŸºå› å®¶æ—ï¼Œå¦‚KEGGåŒæºåŸºå› ï¼Œé¢„æµ‹é€šå¸¸åŸºäº16S rRNAåŸºå› æµ‹åºæ•°æ®ï¼Œä½†ä¹Ÿå¯ä»¥ä½¿ç”¨å…¶ä»–æ ‡è®°åŸºå› ã€‚

```{bash}
# é€‰æ‹©1. ä½¿ç”¨ bioconda å®‰è£… PICRUSt2 ç¯å¢ƒ
conda create -n picrust2 -c bioconda -c conda-forge picrust2

#inputå°±æ˜¯æˆ‘ä»¬ä¸Šé¢å¯¼å‡ºæ¥çš„otu.fastaå’Œotu_table.txt
picrust2_pipeline.py -s otu.fasta -i otu_table.txt -o picrust2_result -p 4
```


```{bash}
# é€‰æ‹©2. å®‰è£…q2 picrust2æ’ä»¶ï¼ˆå‰é¢æåˆ°qiime2çš„ä¸€ä¸ªç‰¹ç‚¹æ˜¯æ’ä»¶åŒ–ï¼Œè¿™é‡Œåˆšå¥½å¯ä»¥è¯•è¯•ï¼Œæ³¨æ„å®‰è£…å¯¹äºQiime2ç‰ˆæœ¬çš„ï¼‰
conda install q2-picrust2 \
  -c conda-forge \
  -c bioconda \
  -c picrust 

# è·‘æ•´ä¸ªæµç¨‹
qiime picrust2 full-pipeline \
   --i-table feature-table.qza \
   --i-seq rep-seqs.qza \
   --output-dir q2-picrust2_output \
   --p-placement-tool sepp \
   --p-threads 1 \
   --p-hsp-method pic \
   --p-max-nsti 2 \
   --verbose
```

q2-picrust2_output ä¸­çš„è¿™äº›è¾“å‡ºæ–‡ä»¶æ˜¯ï¼š
- ec_metagenome.qza - EC å®åŸºå› ç»„é¢„æµ‹ï¼ˆè¡Œæ˜¯ EC ç¼–å·ï¼Œåˆ—æ˜¯æ ·æœ¬ï¼‰ã€‚
- ko_metagenome.qza - KO å®åŸºå› ç»„é¢„æµ‹ï¼ˆè¡Œæ˜¯ KOï¼Œåˆ—æ˜¯æ ·æœ¬ï¼‰ã€‚
- Pathway_abundance.qza - MetaCyc é€”å¾„ä¸°åº¦é¢„æµ‹ï¼ˆè¡Œæ˜¯Pathwayï¼Œåˆ—æ˜¯æ ·æœ¬ï¼‰ã€‚

```{bash}
# å¯¼å‡ºåŠŸèƒ½ä¸°åº¦è¡¨
qiime tools export \
   --input-path q2-picrust2_output/pathway_abundance.qza \
   --output-path pathabun_exported
```


è¿™æ ·ï¼Œæ•´ä¸ªæ‰©å¢å­ä¸Šæ¸¸å°±å·®ä¸å¤šäº†ï¼Œå¾—åˆ°äº†featureä¸°åº¦è¡¨ï¼Œfeatureçš„ç‰©ç§æ³¨é‡Šä¿¡æ¯ï¼Œé¢„æµ‹çš„åŠŸèƒ½ä¸°åº¦è¡¨ï¼Œåç»­åˆ†æå°±ç±»ä¼¼å®åŸºå› ç»„çš„éƒ¨åˆ†æ“ä½œäº†ã€‚

## References
1. https://docs.qiime2.org/2024.10/
2. https://www.jianshu.com/p/f31581bbfb80
3. Evan Bolyen, Jai Ram Rideout, et al. 2019. Reproducible, interactive, scalable and extensible microbiome data science using QIIME 2. **Nature Biotechnology** 37: 852-857. https://doi.org/10.1038/s41587-019-0209-9
4. Douglas, G.M., Maffei, V.J., Zaneveld, J.R. et al. PICRUSt2 for prediction of metagenome functions. Nat Biotechnol 38, 685â€“688 (2020). https://doi.org/10.1038/s41587-020-0548-6
5. https://github.com/YongxinLiu/EasyAmplicon
