---
title: "ä½¿ç”¨ReporterScoreåŒ…è¿›è¡Œå¯Œé›†åˆ†æ"
author: "Peng Chen"
date: "2023-07-04"
categories:
- R
- metagenomic
tags:
- R
- package
- åŠŸèƒ½åŸºå› 
description: Reporter Scoreæ˜¯ä¸€ç§æ”¹è¿›å¾®ç”Ÿç‰©å¯Œé›†åˆ†æçš„æ–°æ–¹æ³•ï¼Œè¿™é‡Œæˆ‘ä»¬åˆ†äº«ä¸€ä¸‹ä½¿ç”¨ReporterScoreåŒ…è¿›è¡Œå¯Œé›†åˆ†æçš„å…·ä½“æ–¹æ³•ã€‚
image: "index.en_files/figure-html/unnamed-chunk-7-1.png"
math: yes
license: null
hidden: no
comments: yes
bibliography: ../../bib/My Library.bib
link-citations: yes
csl: ../../bib/science.csl
slug: "reporterscore-package"
---

```{r include=FALSE}
Packages <- c("dplyr","kableExtra","ggplot2","ReporterScore")
pcutils::lib_ps(Packages)
knitr::opts_chunk$set(message = FALSE,warning = FALSE,eval = T,cache = T)
```

## Introduction

ä¸Šæ¬¡å·²ç»åœ¨[ä¸€ç¯‡æ¨æ–‡](../reporter-score)ä¸­ä»‹ç»è¿‡äº†å¾®ç”Ÿç‰©ç»„åˆ†æå¸¸ç”¨çš„åŠŸèƒ½å¯Œé›†åˆ†ææ–¹æ³•ä»¥åŠreporter scoreæ–¹æ³•çš„åŸç†ï¼Œå¹¶ä¸”ä¹Ÿä»‹ç»äº†æˆ‘å¼€å‘çš„RåŒ…`ReporterScore`ã€‚

ä½†æ—¶é‚£ä¸ªæ—¶å€™RåŒ…å†™çš„è¿˜æ¯”è¾ƒç²—ç³™ï¼ŒåŠŸèƒ½ä¹Ÿä¸å¤šï¼Œæœ€è¿‘è¿›ä¸€æ­¥ä¼˜åŒ–äº†è¿™ä¸ªåŒ…çš„å„ç§åŠŸèƒ½ï¼Œæ”¯æŒä¸¤ç§æ¨¡å¼ï¼Œå¤šç§ç»Ÿè®¡æ£€éªŒæ–¹æ³•ï¼Œæ”¯æŒä¸¤ç»„ç”šè‡³æ›´å¤šç»„çš„å®éªŒè®¾è®¡ï¼ˆè¿™ä¸ªæŒºå¥½ç”¨çš„ï¼‰ï¼ŒKEGGæ•°æ®åº“çš„åŒæ­¥åšçš„ä¹Ÿæ¯”è¾ƒå¥½äº†ï¼Œè¿˜å¢åŠ äº†ä¸€äº›å¯è§†åŒ–å½¢å¼ã€‚

ä»¥ä¸‹æ˜¯æˆ‘ç»™è¿™ä¸ªRåŒ…githubä¸»é¡µï¼ˆ<https://github.com/Asa12138/ReporterScore>ï¼‰ä¸‹å†™çš„ä»‹ç»å’Œç”¨æ³•ï¼Œè¿™é‡Œç®€å•ç¿»è¯‘ä¸€ä¸‹è´´è¿‡æ¥äº†ã€‚ä½†å…¶å®é‡Œé¢è¿˜æœ‰ä¸å°‘åŠŸèƒ½æ²¡åœ¨ä¸‹é¢å†™å‡ºï¼Œå¯ä»¥åœ¨RåŒ…é‡Œæ¢ç´¢ä¸€ä¸‹ï¼Œå“ˆå“ˆã€‚


## Citation

è¿™ä¸ªåŒ…æš‚æ—¶è¿˜æ²¡æœ‰åœ¨åˆŠç‰©ä¸Šå‘è¡¨ï¼Œè¦åœ¨å‡ºç‰ˆç‰©ä¸­å¼•ç”¨ `ReporterScore`ï¼Œè¯·ä½¿ç”¨ï¼š

Chen Peng, Chao Jiang. ReporterScore: an R package for Reporter Score-based Microbial Enrichment Analysis. R package (2023), <https://github.com/Asa12138/ReporterScore>

ğŸ¤—ä¹Ÿæ¬¢è¿åœ¨GitHubä¸Šç‚¹ä¸ªstarâ­ï¸

## Install

```{r eval=FALSE}
if(!require("devtools"))install.packages("devtools")
devtools::install_github('Asa12138/pcutils',dependencies=T)
devtools::install_github('Asa12138/ReporterScore',dependencies=T)
library(ReporterScore)
```

## Usage

### 1. Inputdata (KO abundance table and metadata)

é€šå¸¸ï¼Œæˆ‘ä»¬å¯ä»¥ä½¿ç”¨[KEGGæ•°æ®åº“](https://www.kegg.jp/kegg/)æ¥æ³¨é‡Šæˆ‘ä»¬çš„å¾®ç”Ÿç‰©ç»„æµ‹åºæ•°æ®ï¼Œç‰¹åˆ«æ˜¯ç¯å¢ƒå¾®ç”Ÿç‰©ç»„ï¼Œå› ä¸ºKEGGç›¸å¯¹æ¥è¯´æ›´å…¨é¢ï¼ˆå½“ç„¶å¤§éƒ¨åˆ†è¿˜æ˜¯unknownï¼‰ã€‚

å…·ä½“æ–¹æ³•åŒ…æ‹¬ç›´æ¥ä½¿ç”¨blastå¯¹KEGGåºåˆ—æ•°æ®åº“è¿›è¡Œæ¯”å¯¹ï¼Œä½¿ç”¨KEGGå®˜æ–¹mapperè½¯ä»¶ï¼Œä½¿ç”¨[EggNOGæ•°æ®åº“](http://eggnog5.embl.de/#/app/home)å¹¶å°†ç»“æœè½¬åŒ–ä¸ºKEGGæ³¨é‡Šã€‚

è¿™æ ·æˆ‘ä»¬å°±å¯ä»¥å¾—åˆ°ä¸€ä¸ªKOä¸°åº¦è¡¨ï¼ˆè¡Œæ˜¯KOï¼Œåˆ—æ˜¯æ ·æœ¬ï¼‰ç”¨äºæˆ‘ä»¬çš„å¯Œé›†åˆ†æï¼š

```{r}
data("KO_abundance_test")
head(KO_abundance[,1:6])
```

è¿˜éœ€è¦æä¾›å®éªŒè®¾è®¡çš„å…ƒæ•°æ®metadataï¼ˆè¡Œæ˜¯æ ·æœ¬ï¼Œåˆ—æ˜¯ç»„ï¼‰ã€‚

```{r}
head(metadata)
```

### 2. Pathway database

`ReporterScore`å†…ç½®äº†KEGG é€šè·¯å’Œæ¨¡å—æ•°æ®åº“ï¼ˆ2023-08 ç‰ˆï¼‰ç”¨äºKO ä¸°åº¦è¡¨åˆ†æã€‚

ä½ å¯ä»¥ä½¿ç”¨ `load_KOlist()` æŸ¥çœ‹å¹¶ä½¿ç”¨ `update_KO_file()` æ›´æ–°è¿™äº›æ•°æ®åº“ï¼ˆé€šè¿‡ KEGG APIï¼‰ï¼Œå› ä¸ºä½¿ç”¨æœ€æ–°çš„æ•°æ®åº“éå¸¸é‡è¦ã€‚

æˆ–è€…ä½ å¯ä»¥ä½¿ç”¨`custom_modulelist()`è‡ªå®šä¹‰ä½ è‡ªå·±çš„é€šè·¯æ•°æ®åº“ï¼ˆæ„Ÿå…´è¶£çš„åŸºå› é›†ï¼‰ã€‚

```{r eval=FALSE}
load_KOlist()
head(KOlist$pathway)
```

### 3. One step enrichment

ä½¿ç”¨å‡½æ•°`reporter_score`å¯ä»¥ä¸€æ­¥å¾—åˆ°reporter scoreç»“æœã€‚

æœ‰ä¸€äº›é‡è¦çš„å‚æ•°å¯ä¾›è°ƒèŠ‚ï¼š

-   **mode**: "mixed" æˆ– "directed"ï¼ˆä»…é€‚ç”¨äºä¸¤ç»„å·®å¼‚åˆ†ææˆ–å¤šç»„ç›¸å…³åˆ†æï¼‰ï¼Œè¯¦ç»†ä¿¡æ¯å‚è§`pvalue2zs`ã€‚
-   **æ–¹æ³•**ï¼šç»Ÿè®¡æ£€éªŒç±»å‹ã€‚ é»˜è®¤ä¸º"wilcox.test"ï¼š
    -   `t.test` ï¼ˆå‚æ•°ï¼‰å’Œ `wilcox.test` ï¼ˆéå‚æ•°ï¼‰ã€‚ å¯¹ä¸¤ç»„æ ·å“è¿›è¡Œæ¯”è¾ƒã€‚ å¦‚æœåˆ†ç»„å˜é‡åŒ…å«ä¸¤ä¸ªä»¥ä¸Šæ°´å¹³ï¼Œåˆ™æ‰§è¡Œæˆå¯¹æ¯”è¾ƒã€‚                 -   `anova`ï¼ˆå‚æ•°ï¼‰å’Œ `kruskal.test`ï¼ˆéå‚æ•°ï¼‰ã€‚ æ‰§è¡Œæ¯”è¾ƒå¤šä¸ªç»„çš„å•å‘æ–¹å·®åˆ†ææµ‹è¯•ã€‚
    -   "pearson"ã€"kendall"æˆ–"spearman"ï¼ˆç›¸å…³åˆ†æï¼‰ï¼Œè¯·å‚è§`cor`ã€‚
-   **p.adjust.method**ï¼šç”¨äºæµ‹è¯•ç»“æœçš„p.adjust.methodï¼Œå‚è§`p.adjust`ã€‚
-   **type**/**modulelist**ï¼šé€‰æ‹©é€šè·¯æ•°æ®åº“ï¼Œé»˜è®¤æ•°æ®åº“ä¸º"pathway"æˆ–"module"ï¼Œæˆ–ä½¿ç”¨è‡ªå®šä¹‰çš„æ¨¡å—åˆ—è¡¨ã€‚

groupä½œä¸ºå› å­å˜é‡ï¼Œç¬¬ä¸€ä¸ªæ°´å¹³å°†è¢«è®¾ç½®ä¸º**å¯¹ç…§ç»„**ï¼Œä½ å¯ä»¥æ›´æ”¹å› å­æ°´å¹³æ¥æ”¹å˜ä½ çš„æ¯”è¾ƒã€‚

ä¾‹å¦‚ï¼Œæˆ‘ä»¬æƒ³è¦æ¯”è¾ƒä¸¤ç»„"WT-OE"ï¼Œå¹¶ä½¿ç”¨"directed"æ¨¡å¼ï¼Œå› ä¸ºæˆ‘ä»¬åªæƒ³çŸ¥é“ **OE ç»„** ä¸­å“ªäº›é€šè·¯è¢«å¯Œé›†æˆ–è€—å°½ï¼š

```{r eval=TRUE,collapse=TRUE}
cat("Comparison: ",levels(factor(metadata$Group)))

reporter_score_res=reporter_score(KO_abundance,"Group",metadata,mode="directed")
```

ç»“æœæ˜¯ä¸€ä¸ª"reporter_score"å¯¹è±¡ï¼š

| elements     | description                        |
|--------------|------------------------------------|
| `kodf`       | ä½ çš„è¾“å…¥ KO_abundance è¡¨           |
| `ko_pvalue`  | ko ç»Ÿè®¡ç»“æœåŒ…å« p.value            |
| `ko_stat`    | koç»Ÿè®¡ç»“æœåŒ…å«p.valueå’Œz_score     |
| `reporter_s` | åœ¨æ¯ä¸ªé€”å¾„ä¸­çš„reporter score       |
| `modulelist` | é»˜è®¤ KOlist æˆ–è‡ªå®šä¹‰æ¨¡å—åˆ—è¡¨æ•°æ®æ¡† |
| `group`      | ä½ çš„æ•°æ®ä¸­çš„æ¯”è¾ƒç»„                 |
| `metadata`   | ç¤ºä¾‹ä¿¡æ¯æ•°æ®æ¡†åŒ…å«ç»„               |

### 4. Visualization

ç»˜åˆ¶æœ€æ˜¾è‘—å¯Œé›†çš„é€šè·¯ï¼š

```{r fig.height=7,fig.width=10}
#View(reporter_score_res$reporter_s)
plot_report(reporter_score_res,rs_threshold = c(-2,2))
```

å½“æˆ‘ä»¬èšç„¦äºä¸€æ¡é€šè·¯æ—¶ï¼Œä¾‹å¦‚ â€œmap00780â€ï¼š

```{r}
plot_KOs_in_pathway(reporter_score_res,map_id = "map00780")
```

æˆ–è€…æ˜¾ç¤ºä¸ºç½‘ç»œï¼š

```{r}
plot_KOs_network(reporter_score_res,map_id = "map00780",main="")
```

æˆ‘ä»¬ä¹Ÿå¯ä»¥æŸ¥çœ‹é€šè·¯ä¸­æ¯ä¸ª KO çš„ä¸°åº¦ï¼š

```{r fig.height=7,fig.width=8}
plot_KOs_box(reporter_score_res,map_id = "map00780",only_sig = TRUE)
```

æˆ–è€…çƒ­å›¾å½¢å¼å‘ˆç°ï¼š

```{r }
plot_KOs_heatmap(reporter_score_res,map_id = "map00780",only_sig = TRUE,heatmap_param = list(cutree_rows=2))
```

### example for "mixed"

å¦‚æœæˆ‘ä»¬çš„å®éªŒè®¾è®¡è¶…è¿‡ä¸¤ç»„ï¼Œæˆ‘ä»¬å¯ä»¥é€‰æ‹©å¤šç»„æ¯”è¾ƒå’Œâ€œmixedâ€æ¨¡å¼ï¼š

```{r collapse=TRUE}
cat("Comparison: ",levels(factor(metadata$Group2)))

reporter_score_res2=reporter_score(KO_abundance,"Group2",metadata,mode="mixed",
      method = "kruskal.test",p.adjust.method1 = "none")

plot_KOs_in_pathway(reporter_score_res2,map_id = "map00541")
```


## Details

### Step by step

ä¸€æ­¥å‡½æ•° `reporter_score` ç”±ä¸‰éƒ¨åˆ†ç»„æˆï¼š

1. `ko.test`ï¼šæ­¤å‡½æ•°æœ‰åŠ©äºé€šè¿‡å„ç§å†…ç½®æ–¹æ³•è®¡ç®— KO_abundance çš„ *p-value*ï¼Œä¾‹å¦‚å·®å¼‚åˆ†æï¼ˆt.testã€wilcox.testã€kruskal.testã€anovaï¼‰æˆ–ç›¸å…³åˆ†æï¼ˆpearson ã€spearmanã€kendallï¼‰ã€‚ **ä½ è¿˜å¯ä»¥é€šè¿‡å…¶ä»–æ–¹æ³•è®¡ç®— KO_abundance çš„ p-value**ï¼Œä¾‹å¦‚â€œDESeq2â€ã€â€œEdgerâ€ã€â€œLimmaâ€ã€â€œALDEXâ€ã€â€œANCOMâ€ï¼Œå¹¶è‡ªè¡Œè¿›è¡Œ på€¼çŸ«æ­£ï¼Œç„¶åè·³è¿‡` ko.test` æ­¥éª¤è½¬åˆ°æ­¥éª¤2...
2. `pvalue2zs`ï¼šè¯¥å‡½æ•°å°† KO çš„ p-value è½¬æ¢ä¸º Z-scoreï¼ˆé€‰æ‹©æ¨¡å¼: â€œmixedâ€ æˆ– â€œdirectedâ€ï¼‰ã€‚
3. `get_reporter_score` è¯¥å‡½æ•°è®¡ç®—ç‰¹å®šæ•°æ®åº“ä¸­æ¯ä¸ªé€šè·¯çš„reporter scoreã€‚ ä½ å¯ä»¥åœ¨æ­¤å¤„ä½¿ç”¨è‡ªå®šä¹‰æ•°æ®åº“ã€‚

è¿™æ ·ä½ å°±å¯ä»¥ä¸€æ­¥ä¸€æ­¥å¾—åˆ°reporter scoreã€‚

### Other enrichment methods

`ReporterScore` è¿˜æä¾›äº†å…¶ä»–ä¸°å¯Œæ–¹æ³•ï¼Œå¦‚ `KO_fisher`(fisher.test)ã€`KO_enrich`(fisher.test, from `clusterProfiler`)ã€`KO_gsea` (GSEA, from `clusterProfiler`)ï¼Œè¾“å…¥æ•°æ®æ¥è‡ª `reporter_score `ï¼Œå¹¶ä¸”ä¹Ÿæ”¯æŒè‡ªå®šä¹‰æ•°æ®åº“ï¼Œå› æ­¤ä½ å¯ä»¥è½»æ¾æ¯”è¾ƒå„ç§å¯Œé›†æ–¹æ³•çš„ç»“æœå¹¶è¿›è¡Œå…¨é¢åˆ†æï¼š

```{r message=FALSE}
data("KO_abundance_test")
reporter_score_res2=reporter_score(KO_abundance,"Group",metadata,mode="mixed")
#View(reporter_score_res2$reporter_s)
#reporter_score
reporter_score_res2$reporter_s$p.adjust=p.adjust(reporter_score_res2$reporter_s$p.value,"BH")
filter(reporter_score_res2$reporter_s,(ReporterScore)>1.64,p.adjust<0.05)%>%pull(ID)->RS
#fisher
ko_pvalue=reporter_score_res2$ko_pvalue
fisher_res=KO_fisher(ko_pvalue)
filter(fisher_res,p.adjust<0.05)%>%pull(ID)->Fisher
#enricher
enrich_res=KO_enrich(ko_pvalue)
filter(enrich_res,p.adjust<0.05)%>%pull(ID)->clusterProfiler
#GESA
set.seed(1234)
gsea_res=KO_gsea(ko_pvalue)
filter(gsea_res@result,p.adjust<0.05)%>%pull(ID)->GSEA

venn_res=list(RS=RS,Fisher=Fisher,CP=clusterProfiler,GSEA=GSEA)
library(pcutils)
venn(venn_res)
venn(venn_res,"network",vertex.label.cex=c(rep(1,4),rep(0.5,22)))
```

### KO levels

[KEGG BRITE](https://www.genome.jp/kegg/brite.html) æ˜¯ä¸€ä¸ªå±‚æ¬¡åˆ†ç±»ç³»ç»Ÿçš„é›†åˆï¼Œæ•è·å„ç§ç”Ÿç‰©å¯¹è±¡çš„åŠŸèƒ½å±‚æ¬¡ç»“æ„ï¼Œç‰¹åˆ«æ˜¯é‚£äº›è¡¨ç¤ºä¸º KEGG å¯¹è±¡çš„åŠŸèƒ½å±‚æ¬¡ç»“æ„ã€‚

æˆ‘ä»¬æ”¶é›†äº† k00001 KEGG Orthology (KO) è¡¨ï¼Œä»¥ä¾¿ä½ å¯ä»¥æ€»ç»“æ¯ä¸ªçº§åˆ«çš„ä¸°åº¦ã€‚ ä½¿ç”¨ `load_KO_htable` è·å– KO_htable å¹¶ä½¿ç”¨ `update_KO_htable` è¿›è¡Œæ›´æ–°ã€‚ ä½¿ç”¨ `up_level_KO` å¯ä»¥å‡çº§åˆ°â€œpathwayâ€ã€â€œmoduleâ€ã€â€œlevel1â€ã€â€œlevel2â€ã€â€œlevel3â€ã€â€œmodule1â€ã€â€œmodule2â€ã€â€œmodule3â€ä¹‹ä¸€ä¸­çš„ç‰¹å®šçº§åˆ«ã€‚

```{r fig.height=7,fig.width=10}
load_KO_htable()
head(KO_htable)
plot_KO_htable()
```

```{r collapse=TRUE}
KO_level1=up_level_KO(KO_abundance,level = "level1",show_name = TRUE)
pcutils::stackplot(KO_level1[-which(rownames(KO_level1)=="Unknown"),])
```

## Reference

1.  Patil, K. R.
    & Nielsen, J. Uncovering transcriptional regulation of metabolism by using metabolic network topology.
    Proc Natl Acad Sci U S A 102, 2685--2689 (2005).

2.  L. Liu, R. Zhu, D. Wu, Misuse of reporter score in microbial enrichment analysis. iMeta. 2, e95 (2023).

3.  <https://github.com/wangpeng407/ReporterScore>