---
title: ä½¿ç”¨Krakenè¿›è¡Œ16S/ITSç‰©ç§æ³¨é‡Šï¼ˆè¶…å¿«ï¼‰
author: Peng Chen
date: '2025-01-14'
slug: kraken-16s-its
categories:
  - metagenomic
tags:
  - amplicon
description: Qiime2ç‰©ç§æ³¨é‡Šå¯çœŸæ…¢å‘€ğŸ˜‚ï¼Œèµ¶ç´§æ‰¾äº†Krakençš„æ›¿ä»£æ–¹æ³•ï¼ŒKrakenç”¨äº16Sæ˜¯ç°æˆçš„ã€‚ä½†æ˜¯ITSçš„Uniteæ•°æ®åº“éœ€è¦è‡ªå·±æ„å»ºï¼Œè‡ªå·±æŠ˜è…¾äº†ä¸€ä¸‹è§£å†³äº†ã€‚
image: images/k16s1.png
math: ~
license: ~
hidden: no
comments: yes
---

```{r include=FALSE}
knitr::opts_chunk$set(message = FALSE,warning = FALSE,eval = FALSE)
```

## Introduction

ä¸Šæ¬¡è®°å½•äº†ä¸€ä¸‹qiime2çš„[æ‰©å¢å­åˆ†ææµç¨‹](../amplicon-workflow)ã€‚ä½†æ˜¯å®é™…ä½¿ç”¨æ—¶ï¼Œåœ¨æœ€åä¸€æ­¥ç‰©ç§æ³¨é‡Šæ—¶å¡ä½äº†ï¼Œå¦‚æœä½¿ç”¨å¸¸ç”¨çš„`feature-classifier`æ–¹æ³•ï¼Œæˆ‘å°è¯•ä½¿ç”¨uniteæ•°æ®åº“é‰´å®šäº†100æ¡ITSåºåˆ—ï¼Œå±…ç„¶ç”¨äº†åŠä¸ªå°æ—¶ï¼å°´å°¬çš„æ˜¯ï¼Œæˆ‘æœ‰æ•´æ•´10ä¸‡æ¡ASVåºåˆ—ï¼Œè¿™å¾—è·‘åˆ°çŒ´å¹´é©¬æœˆã€‚æˆ‘æƒ³åˆ°åšå®åŸºå› ç»„æ—¶ï¼Œåƒä¸‡çº§æ•°é‡çš„readsç”¨krakenåšç‰©ç§æ³¨é‡Šä¹Ÿä¸éœ€è¦è¿™ä¹ˆä¹…ï¼Œè‚¯å®šæœ‰å¿«çš„æ–¹æ³•çš„ã€‚

ç„¶åå‘ç°Krakenç”¨äº16Så·²ç»æœ‰æ–‡ç« äº†ï¼š
Lu, J., Salzberg, S.L. Ultrafast and accurate 16S rRNA microbial community analysis using Kraken 2. Microbiome 8, 124 (2020). <https://doi.org/10.1186/s40168-020-00900-2>

Kraken 2ç°åœ¨æ”¯æŒ16S rRNAæ•°æ®åº“ï¼Œå¯ä»¥ç›´æ¥ä¸QIIMEå’Œç±»ä¼¼ç³»ç»Ÿè¿›è¡Œæ¯”è¾ƒã€‚Kraken 2é€šè¿‡å…¶ç‹¬ç‰¹çš„æ— å¯¹é½ç®—æ³•ï¼Œèƒ½å¤Ÿåœ¨æçŸ­çš„æ—¶é—´å†…å®Œæˆå¤§é‡åºåˆ—çš„åˆ†ç±»ï¼Œå¹¶ä¸”æ”¯æŒå¤šçº¿ç¨‹å¤„ç†ï¼Œæ˜¾è‘—æå‡äº†åˆ†ææ•ˆç‡ã€‚æ­¤å¤–ï¼ŒKraken 2è¿˜ç»“åˆäº†Brackenå·¥å…·ï¼Œèƒ½å¤Ÿæ›´ç²¾ç¡®åœ°ä¼°è®¡ç‰©ç§çš„ç›¸å¯¹ä¸°åº¦ï¼Œå°¤å…¶æ˜¯åœ¨å¤„ç†å¤æ‚å¾®ç”Ÿç‰©ç¾¤è½æ—¶è¡¨ç°å‡ºè‰²ã€‚

ä½†æ˜¯Kraken2è¦ç”¨åœ¨ITSä¸Šçš„è¯ç›®å‰æ˜¯å®˜æ–¹æ²¡æœ‰æ”¯æŒçš„ï¼Œéœ€è¦æ‰‹åŠ¨æ„å»ºç›¸åº”æ•°æ®åº“ï¼Œè¿™ä¸ªæŠ˜è…¾äº†ä¸€ä¼šï¼Œå…·ä½“æ–¹æ³•æ”¾åœ¨åé¢äº†ã€‚

## Krakenæ•ˆæœ

### é€Ÿåº¦å¿«

![](images/k16s1.png)

Kraken 2åœ¨é€Ÿåº¦ä¸Šçš„è¡¨ç°éå¸¸çªå‡ºã€‚æ ¹æ®Luå’ŒSalzbergçš„ç ”ç©¶ï¼ŒKraken 2åœ¨ç”Ÿæˆ16S rRNAæ•°æ®åº“æ—¶ï¼Œæ¯”QIIME 2å¿«å¾—å¤šã€‚ä¾‹å¦‚ï¼Œä½¿ç”¨å•çº¿ç¨‹æ—¶ï¼ŒKraken 2ç”ŸæˆGreengenesæ•°æ®åº“ä»…éœ€9åˆ†é’Ÿï¼Œè€ŒQIIME 2åˆ™éœ€è¦78åˆ†é’Ÿã€‚å¯¹äºæ›´å¤§çš„SILVAæ•°æ®åº“ï¼ŒKraken 2ä»…éœ€34åˆ†é’Ÿï¼Œè€ŒQIIME 2åˆ™éœ€è¦è¶…è¿‡58å°æ—¶ã€‚åœ¨åˆ†ç±»é€Ÿåº¦ä¸Šï¼ŒKraken 2åŒæ ·è¡¨ç°å‡ºè‰²ï¼Œå•çº¿ç¨‹ä¸‹ä»…éœ€1åˆ†é’Ÿå³å¯å®Œæˆä¸€ä¸ªæ ·æœ¬çš„åˆ†ç±»ï¼Œè€ŒQIIME 2åˆ™éœ€è¦35åˆ†é’Ÿï¼ˆä½¿ç”¨16çº¿ç¨‹æ—¶ï¼‰ã€‚å¯¹äºå¤§è§„æ¨¡æ•°æ®åˆ†æï¼ŒKraken 2çš„é€Ÿåº¦ä¼˜åŠ¿å°¤ä¸ºæ˜æ˜¾ï¼Œèƒ½å¤Ÿæ˜¾è‘—ç¼©çŸ­åˆ†ææ—¶é—´ã€‚

### æ›´ç²¾ç¡®

![](images/k16s2.png)

Kraken 2ä¸ä»…åœ¨é€Ÿåº¦ä¸Šå ä¼˜ï¼Œå…¶åˆ†ç±»ç²¾åº¦ä¹Ÿä¼˜äºQIIME 2ã€‚ç ”ç©¶æ˜¾ç¤ºï¼ŒKraken 2åœ¨åˆ†ç±»16S rRNAåºåˆ—æ—¶ï¼Œèƒ½å¤Ÿæ›´å‡†ç¡®åœ°è¯†åˆ«ç‰©ç§ï¼Œå°¤å…¶æ˜¯åœ¨å¤„ç†å¤æ‚å¾®ç”Ÿç‰©ç¾¤è½æ—¶ã€‚Brackenå·¥å…·è¿›ä¸€æ­¥æå‡äº†Kraken 2çš„ç²¾åº¦ï¼Œèƒ½å¤Ÿæ›´å‡†ç¡®åœ°ä¼°è®¡ç‰©ç§çš„ç›¸å¯¹ä¸°åº¦ã€‚é€šè¿‡é‡æ–°åˆ†é…Kraken 2åœ¨å±æˆ–æ›´é«˜åˆ†ç±»æ°´å¹³ä¸Šçš„åˆ†ç±»ç»“æœï¼ŒBrackenèƒ½å¤Ÿå°†åˆ†ç±»ç»“æœç»†åŒ–åˆ°ç§æ°´å¹³ï¼Œä»è€Œæä¾›æ›´ç²¾ç¡®çš„ç‰©ç§ä¸°åº¦ä¼°è®¡ã€‚æ­¤å¤–ï¼ŒKraken 2çš„æ¯è¯»é•¿åˆ†ç±»åŠŸèƒ½ä½¿å…¶èƒ½å¤Ÿä¸ºæ¯ä¸ªè¯»é•¿æä¾›è¯¦ç»†çš„åˆ†ç±»ä¿¡æ¯ï¼Œè¿›ä¸€æ­¥æé«˜äº†åˆ†æçš„å‡†ç¡®æ€§ã€‚

![](images/k16s3.png)

æ€»çš„æ¥è¯´ï¼ŒKraken 2å’ŒBrackençš„ç»„åˆä¸º16S rRNAæ•°æ®åˆ†ææä¾›äº†ä¸€ä¸ªå¿«é€Ÿã€é«˜æ•ˆä¸”ç²¾ç¡®çš„è§£å†³æ–¹æ¡ˆï¼Œç‰¹åˆ«é€‚åˆå¤„ç†å¤§è§„æ¨¡å¾®ç”Ÿç‰©ç¾¤è½æ•°æ®ã€‚

## Kraken-16S

Kraken2æ˜¯ç›´æ¥æ”¯æŒ16Så¸¸ç”¨çš„ä¸‰å¤§æ•°æ®åº“çš„(greengenesï¼Œrdpï¼Œsilva)ï¼Œç”¨èµ·æ¥ä¹Ÿéå¸¸ç®€å•ï¼š

```{bash}
#ç›´æ¥æ„å»ºæ•°æ®åº“ï¼ŒTYPEå¯é€‰greengenesï¼Œrdpï¼Œsilva
kraken2-build --db $DBNAME --special TYPE

#æ„å»ºå®Œæˆåå°±å¯ä»¥é‰´å®šäº†ï¼Œå†…å­˜å¤Ÿçš„è¯ï¼Œ1minå†…éƒ½æœ‰ç»“æœ
kraken2 rep-seqs/dna-sequences.fasta --db $DBNAME --out kraken_output --threads 4 --report kraken_report.txt -c 0.05

#å¯ä»¥å€ŸåŠ©taxonkitå¾—åˆ°ç±»ä¼¼qiime2æ ‡å‡†æ³¨é‡Šæ–‡ä»¶
cut -f3 kraken_output >tmpid
taxonkit lineage tmpid --data-dir $DBNAME/taxonomy/|taxonkit reformat --data-dir $DBNAME/taxonomy/ -P >tmpres

paste kraken_output tmpres |cut -f2,8>kraken_taxonomy
rm tmpid tmpres
```

## Kraken-ITS

ITSä¸€èˆ¬ç”¨Uniteæ•°æ®åº“æ³¨é‡Šï¼Œä½†è¿™ä¸ªkraken2ä¸æ”¯æŒï¼Œæˆ‘å°è¯•å°†Uniteæ•°æ®åº“æ•´ç†ä¸ºgreengenesçš„æ ¼å¼ï¼Œç„¶åç”¨`build_gg_taxonomy.pl`ç”Ÿæˆnames.dmp, nodes.dmp, seqid2taxidå¹¶å»ºç«‹krakenæ•°æ®åº“ï¼Œæœ€åè¿›è¡ŒITSåºåˆ—é‰´å®šæ˜¯å¯è¡Œçš„ã€‚

å…·ä½“æ­¥éª¤å¦‚ä¸‹ï¼š
1. Uniteå®˜ç½‘ä¸‹è½½fastaæ–‡ä»¶ï¼Œ<https://unite.ut.ee/repository.php> ï¼Œç‚¹å‡»General FASTA releaseä¸‹è½½è§£å‹ï¼Œæˆ‘ç”¨çš„æ˜¯sh_general_release_dynamic_all_04.04.2024.fastaæ–‡ä»¶ã€‚

2. è¿è¡Œä»¥ä¸‹å‘½ä»¤å»ºç«‹æ•°æ®åº“ï¼Œæ³¨æ„`split_fasta.py`å’Œ`build_gg_taxonomy.pl`åœ¨åé¢ï¼Œè®°å¾—æ‹·è´è¿‡å»ç”¨ã€‚

```{bash}
# æ‰‹åŠ¨æ„å»ºç›®å½•ç»“æ„
mkdir ~/db/kraken_unite/
pushd ~/db/kraken_unite/
mkdir -p data taxonomy library
pushd data

# ç”¨split_fasta.pyæŠŠuniteæ•°æ®åº“æä¾›çš„fastaæ ¼å¼æ•´ç†ä¸ºgreengenesçš„
GG_VERSION=unite_all_04.04.2024
./split_fasta.py -i ../sh_general_release_dynamic_all_04.04.2024.fasta -o ${GG_VERSION}

# ç”¨build_gg_taxonomy.plè¿›ä¸€æ­¥ç”Ÿæˆnames.dmp, nodes.dmp, seqid2taxid
perl ../build_gg_taxonomy.pl "${GG_VERSION}_taxonomy.txt"
popd
mv data/names.dmp data/nodes.dmp taxonomy/
mv data/seqid2taxid.map .
mv "data/${GG_VERSION}.fa" library/unite.fna
popd

# ç”¨kraken2-buildæ„å»ºç”¨äºæ³¨é‡Šçš„æ•°æ®åº“
kraken2-build --db ~/db/kraken_unite --build --threads 4
```

3. é‰´å®šè·å¾—outputæ–‡ä»¶

```{bash}
#1ç§’å‡ºç»“æœ
kraken2 rep-seqs/dna-sequences.fasta --db ~/db/kraken_unite/ --out kraken_output --threads 4 --report kraken_report.txt -c 0.05
```

4. å€ŸåŠ©taxonkitè·å–æ ‡å‡†æ³¨é‡Šæ–‡ä»¶ï¼Œtaxonkitè¿™è½¯ä»¶ä¹Ÿå¾ˆå¥½ç”¨ã€‚
```{bash}
cut -f3 kraken_output >tmpid
taxonkit lineage tmpid --data-dir ~/db/kraken_unite/taxonomy/|taxonkit reformat  --data-dir ~/db/kraken_unite/taxonomy/ -P >tmpres

paste kraken_output tmpres |cut -f2,8>kraken_taxonomy
rm tmpid tmpres
```

æœ€åçš„ç»“æœå°±æ˜¯ç±»ä¼¼ä¸‹é¢çš„å½¢å¼ï¼š

```
0229a5c7f887dad960b40a0505cf9776	k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetales_fam_Incertae_sedis;g__Candida;s__Candida Candida_albicans
f88d9001e10f6d37672514972e6a199e	k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetales_fam_Incertae_sedis;g__Candida;s__Candida Candida_albicans
c3519270e269b92d3577ed450032bbc4	k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Penicillium;s__
```

è¡¥å……è„šæœ¬ä¿¡æ¯ï¼š

1.split_fasta.pyï¼š
```{python}
#!/share/home/jianglab/pengchen/miniconda3/bin/python3.9
import argparse

def process_fasta(input_file, unite_fa, unite_taxonomy):
    with open(input_file, 'r') as infile, open(unite_fa, 'w') as fa_out, open(unite_taxonomy, 'w') as tax_out:
        for line in infile:
            if line.startswith('>'):
                # Extract information from the header
                header = line.strip()
                parts = header.split('|')
                accession = parts[1]
                taxonomy = parts[-1]
                formatted_taxonomy = taxonomy.replace(';', '; ').replace(' ', '').replace(';', '; ')

                # Write to unite.fa
                fa_out.write(f">{accession}\n")

                # Write to unite_taxonomy.txt
                tax_out.write(f"{accession}\t{formatted_taxonomy}\n")
            else:
                # Write sequence lines to unite.fa
                fa_out.write(line)

def main():
    parser = argparse.ArgumentParser(description="Process a fasta file to generate unite.fa and unite_taxonomy.txt.")
    parser.add_argument("-i", required=True, help="Input fasta file")
    parser.add_argument("-o", required=True, help="Output prefix for unite.fa and unite_taxonomy.txt")
    args = parser.parse_args()

    input_file = args.i
    output_prefix = args.o

    unite_fa = f"{output_prefix}.fa"
    unite_taxonomy = f"{output_prefix}_taxonomy.txt"

    process_fasta(input_file, unite_fa, unite_taxonomy)

if __name__ == "__main__":
    main()
```

2.build_gg_taxonomy.pl : <https://github.com/DerrickWood/kraken2/blob/master/scripts/build_gg_taxonomy.pl>

3.taxonkit: https://github.com/shenwei356/taxonkit

## References
1. Lu, J., Salzberg, S.L. Ultrafast and accurate 16S rRNA microbial community analysis using Kraken 2. Microbiome 8, 124 (2020). https://doi.org/10.1186/s40168-020-00900-2
2. https://github.com/DerrickWood/kraken2/wiki/Manual#custom-databases
3. https://github.com/DerrickWood/kraken2/issues/97