---
title: Metagenomic workflow
author: Peng Chen
date: '2023-03-26'
slug: metagenomic-workflow
categories:
  - metagenomic
tags:
  - shell
  - ç¼–ç¨‹
description: å®åŸºå› ç»„ï¼ˆMetagenomeï¼‰æ˜¯æŒ‡å¯¹ä¸€ä¸ªç”Ÿæ€ç³»ç»Ÿä¸­çš„æ‰€æœ‰å¾®ç”Ÿç‰©è¿›è¡ŒDNAåˆ†æçš„è¿‡ç¨‹ï¼Œå¯ä»¥å¸®åŠ©ç ”ç©¶äººå‘˜äº†è§£å¾®ç”Ÿç‰©çš„å¤šæ ·æ€§ã€åŠŸèƒ½å’Œäº’ä½œå…³ç³»ã€‚è¿™é‡Œä»‹ç»å¸¸ç”¨åˆ†ææµç¨‹ã€‚
image: images/workflow.png
math: ~
license: ~
hidden: no
comments: yes
bibliography: [../../bib/My Library.bib]
link-citations: yes
csl: ../../bib/science.csl
---
```{r include=FALSE}
Packages <- c("dplyr","kableExtra","ggplot2")
pcutils::lib_ps(Packages)
knitr::opts_chunk$set(message = FALSE,warning = FALSE,eval = FALSE)
```

## Introduction

å®åŸºå› ç»„ï¼ˆMetagenomeï¼‰æ˜¯æŒ‡å¯¹ä¸€ä¸ªç”Ÿæ€ç³»ç»Ÿä¸­çš„æ‰€æœ‰å¾®ç”Ÿç‰©è¿›è¡ŒDNAåˆ†æçš„è¿‡ç¨‹ï¼Œå¯ä»¥å¸®åŠ©ç ”ç©¶äººå‘˜äº†è§£å¾®ç”Ÿç‰©çš„å¤šæ ·æ€§ã€åŠŸèƒ½å’Œäº’ä½œå…³ç³»ã€‚

å®åŸºå› ç»„çš„åº”ç”¨éå¸¸å¹¿æ³›ï¼ŒåŒ…æ‹¬ï¼š

-   ç”Ÿç‰©å¤šæ ·æ€§ç ”ç©¶ï¼šé€šè¿‡å¯¹å®åŸºå› ç»„è¿›è¡Œåˆ†æï¼Œå¯ä»¥äº†è§£ä¸åŒç”Ÿæ€ç³»ç»Ÿä¸­å¾®ç”Ÿç‰©çš„å¤šæ ·æ€§å’Œåˆ†å¸ƒæƒ…å†µã€‚

-   ç”Ÿæ€å­¦ç ”ç©¶ï¼šå®åŸºå› ç»„å¯ä»¥å¸®åŠ©ç ”ç©¶äººå‘˜äº†è§£å¾®ç”Ÿç‰©åœ¨ç”Ÿæ€ç³»ç»Ÿä¸­çš„åŠŸèƒ½ã€äº’ä½œå…³ç³»å’Œç”Ÿæ€ä½ç­‰ã€‚

-   ç”Ÿç‰©æŠ€æœ¯ï¼šå®åŸºå› ç»„å¯ä»¥ç”¨äºç­›é€‰å…·æœ‰ç‰¹å®šåŠŸèƒ½çš„å¾®ç”Ÿç‰©ï¼Œä¾‹å¦‚ï¼Œå¯»æ‰¾èƒ½å¤Ÿé™è§£æœ‰å®³ç‰©è´¨çš„å¾®ç”Ÿç‰©ã€‚

å®åŸºå› ç»„çš„åˆ†æä¸€èˆ¬åŒ…æ‹¬ä»¥ä¸‹æ­¥éª¤ï¼š

1.  DNAæå–ä¸å»ºåº“ã€‚

2.  é«˜é€šé‡æµ‹åºï¼šä½¿ç”¨é«˜é€šé‡æµ‹åºæŠ€æœ¯å¯¹æ‰©å¢åçš„DNAè¿›è¡Œæµ‹åºï¼Œå¾—åˆ°åŸå§‹åºåˆ—æ•°æ®ã€‚

3.  æ•°æ®æ¸…æ´—å’Œç»„è£…ï¼šå¯¹åŸå§‹æ•°æ®è¿›è¡Œè´¨é‡æ§åˆ¶ã€å»é™¤ä½è´¨é‡åºåˆ—å’Œå†—ä½™åºåˆ—ï¼Œå°†åºåˆ—æ‹¼æ¥æˆè¾ƒé•¿çš„è¿ç»­åºåˆ—ï¼ˆcontigsï¼‰ã€‚

4.  åŸºå› æ³¨é‡Šï¼šå°†contigsä¸­çš„åŸºå› è¿›è¡Œæ³¨é‡Šï¼Œå¾—åˆ°åŸºå› åŠŸèƒ½ä¿¡æ¯ã€‚

5.  æ•°æ®åˆ†æï¼šäº†è§£å¾®ç”Ÿç‰©å¤šæ ·æ€§ã€ç¾¤è½ç»“æ„ã€åŠŸèƒ½ç‰¹å¾ç­‰ä¿¡æ¯ï¼ˆæ›´å¤šæ˜¯æŒ‡è·å–äº†ç‰©ç§ä¸°åº¦è¡¨æˆ–åŠŸèƒ½ä¸°åº¦è¡¨ä¹‹åçš„è¿›ä¸€æ­¥åˆ†æï¼‰ã€‚

6.  MAGs binningï¼Œ è¿›åŒ–åŠ¨æ€ç­‰è¿›ä¸€æ­¥åˆ†æ

è¿™æ˜¯æˆ‘å¸¸ç”¨çš„ä¸€å¥—åŸºæœ¬æµç¨‹(Figure \@ref(fig:1-work))ï¼Œå½“ç„¶åœ¨é¢å¯¹ä¸åŒé¡¹ç›®æ—¶åº”è¯¥æœ‰ä¸åŒçš„ä¾§é‡ç‚¹å’Œé€‚ç”¨çš„åˆ†ææ–¹æ³•ï¼Œå¯ä»¥åœ¨æ­¤åŸºç¡€ä¸Šæ·»åŠ æˆ–ä¿®æ”¹ã€‚

æœ€æ—©è¿™æ–¹é¢çš„åˆ†ææˆ‘éƒ½æ˜¯å‚è€ƒåˆ˜æ°¸é‘«è€å¸ˆçš„[**EasyMetagenome**](https://github.com/YongxinLiu/EasyMetagenome),ç°åœ¨è¿™å¥—æµç¨‹ä¹Ÿå‘æ–‡ç« äº† @liuPracticalGuideAmplicon2021ï¼Œå€¼å¾—å‚è€ƒï¼Œå¯¹ä¸Šæ‰‹16Sæµ‹åºæ•°æ®æˆ–å®åŸºå› ç»„æ•°æ®éƒ½å¾ˆæœ‰å¸®åŠ©ã€‚

```{r 1-work,echo=FALSE,eval=TRUE,fig.cap="Basic workflow",out.width="80%"}
knitr::include_graphics("images/workflow.png")
```

## preprocess

ç»å¤§å¤šæ•°è¿™é‡Œä»‹ç»çš„è½¯ä»¶éƒ½æ˜¯ä»…æ”¯æŒlinuxå¹³å°çš„ï¼Œæˆ‘ä»¬åšæµ‹åºæ–‡ä»¶çš„ä¸Šæ¸¸åˆ†æä¹Ÿè‚¯å®šæ˜¯åœ¨æœåŠ¡å™¨ä¸Šåšï¼Œä¸ªäººPCä¸€èˆ¬å¾ˆéš¾æ»¡è¶³éœ€æ±‚ï¼Œæ‰€ä»¥åœ¨åšè¿™äº›åˆ†æå‰å¿…é¡»å…ˆå­¦ä¹ linuxåŸºç¡€çŸ¥è¯†å¦‚æ–‡ä»¶ç³»ç»Ÿï¼Œshellè„šæœ¬ç¼–å†™ï¼Œè½¯ä»¶å®‰è£…ç­‰ã€‚

å®‰è£…è½¯ä»¶å»ºè®®ä½¿ç”¨condaæˆ–mambaï¼ˆæ–°å»ºç¯å¢ƒå’Œç®¡ç†ï¼‰ï¼Œæœ‰å¾ˆå¤šå‚è€ƒæ–¹æ³•ã€‚

æˆ‘ä»¬æœåŠ¡å™¨ä½¿ç”¨çš„æ˜¯slurmä½œä¸šç®¡ç†ç³»ç»Ÿï¼Œå°½é‡å…ˆå­¦ä¹ ä¸€ä¸‹[slurmçš„ä½¿ç”¨](https://docs.hpc.sjtu.edu.cn/job/slurm.html)å†å°è¯•æäº¤ä½œä¸šã€‚

ä¸€èˆ¬æŠŠæ‰€æœ‰æ ·æœ¬çš„æµ‹åºåŒç«¯æ–‡ä»¶æ”¾åœ¨ä¸€ä¸ªæ–‡ä»¶å¤¹ä¸‹

### è´¨æ§ï¼šfastp

```{bash}
#!/bin/bash
#SBATCH --job-name=fastp
#SBATCH --output=/share/home/jianglab/pengchen/work/asthma/fastp/log/%x_%a.out
#SBATCH --error=/share/home/jianglab/pengchen/work/asthma/fastp/log/%x_%a.err
#SBATCH --array=1-33
#SBATCH --partition=short
#SBATCH --cpus-per-task=8


echo start: `date +'%Y-%m-%d %T'`
start=`date +%s`
echo "SLURM_ARRAY_TASK_ID: " $SLURM_ARRAY_TASK_ID
sample=$(head -n $SLURM_ARRAY_TASK_ID ~/work/asthma/data/namelist | tail -1)
#sample=$(head -n 1 namelist | tail -1)
echo handling: $sample
####################

fastp -w 8 -i ~/work/asthma/data/$sample/$sample'_f1.fastq' -o ${sample}_1 \
-I ~/work/asthma/data/$sample/$sample'_r2.fastq' -O ${sample}_2 -j ~/work/asthma/fastp/${i}.json

####################
echo end: `date +'%Y-%m-%d %T'`
end=`date +%s`
echo TIME:`expr $end - $start`s

```

åé¢æ¥ä¸€ä¸ªpythonè„šæœ¬å°±å¯ä»¥ç»Ÿè®¡å¸¸ç”¨æŒ‡æ ‡äº†ã€‚

æŠŠæ‰€æœ‰çš„.jsonæ–‡ä»¶ç§»åˆ°ä¸€ä¸ªæ–‡ä»¶å¤¹é‡Œï¼Œreport/ä¸‹ï¼Œå°±å¯ä»¥ç»Ÿè®¡äº†ã€‚

### å»å®¿ä¸»ï¼šbowtie2

å…¶å®å°±æ˜¯å°†åºåˆ—æ¯”å¯¹åˆ°äººç±»åŸºå› ç»„ä¸Šï¼Œæ²¡æœ‰æ¯”å¯¹åˆ°çš„åºåˆ—æ•´åˆæˆæ–°æ–‡ä»¶å°±æ˜¯å»å®¿ä¸»åçš„äº†ã€‚

```{bash}
#!/bin/bash
#SBATCH --job-name=rm_human
#SBATCH --output=/share/home/jianglab/pengchen/work/meta/%x_%a.out
#SBATCH --error=/share/home/jianglab/pengchen/work/meta/%x_%a.err
#SBATCH --cpus-per-task=32
#SBATCH --partition=short

echo start: `date +'%Y-%m-%d %T'`
start=`date +%s`
#############
for i in C1 C2
do
bowtie2 -p 32 -x ~/db/humangenome/hg38 -1 seq/${i}_1.fq.gz \
 -2 seq/${i}_2.fq.gz -S ${i}.sam --un-conc ${i}.fq --very-sensitive
done
##############
echo end: `date +'%Y-%m-%d %T'`
end=`date +%s`
echo TIME:`expr $end - $start`s
```

### åŸºæœ¬ä¿¡æ¯ç»Ÿè®¡

å¯ä»¥ç”¨FastqCountï¼š

```{bash}
~/biosoft/FastqCount-master/FastqCount_v0.5 xx.fastq.gz

Total Reads     Total Bases     N Bases Q20     Q30     GC
11568822 (11.57 M)      1702829127 (1.70 G)     0.00%   98.00%  94.00%  54.00%
```

## reads-based

### ç‰©ç§æ³¨é‡Šï¼škraken2

Kraken 2æ˜¯ä¸€ä¸ªç”¨äºå¯¹é«˜é€šé‡æµ‹åºæ•°æ®è¿›è¡Œåˆ†ç±»å’Œæ ‡è¯†ç‰©ç§çš„è½¯ä»¶ã€‚å®ƒä½¿ç”¨å‚è€ƒæ•°æ®åº“ä¸­çš„åŸºå› ç»„åºåˆ—æ¥è¿›è¡Œåˆ†ç±»ï¼Œå¹¶ä½¿ç”¨k-meræ–¹æ³•æ¥å®ç°å¿«é€Ÿå’Œå‡†ç¡®çš„åˆ†ç±»ã€‚

ä½¿ç”¨Kraken 2è¿›è¡ŒåŸºæœ¬åˆ†ç±»çš„ç®€å•æ­¥éª¤ï¼š

1.  å‡†å¤‡å‚è€ƒæ•°æ®åº“ï¼šKraken 2éœ€è¦ä¸€ä¸ªå‚è€ƒæ•°æ®åº“ï¼Œä»¥ä¾¿å¯¹æµ‹åºæ•°æ®è¿›è¡Œåˆ†ç±»ã€‚å¯ä»¥ä»NCBIã€Ensemblæˆ–å…¶ä»–æ•°æ®åº“ä¸‹è½½ç›¸åº”çš„åŸºå› ç»„åºåˆ—ï¼Œå¹¶ä½¿ç”¨Kraken 2å†…ç½®çš„å·¥å…·æ¥æ„å»ºæ•°æ®åº“ã€‚

2.  å®‰è£…Kraken 2ï¼šå¯ä»¥ä»Kraken 2å®˜æ–¹ç½‘ç«™ä¸‹è½½å¹¶å®‰è£…Kraken 2è½¯ä»¶ã€‚

3.  è¿è¡ŒKraken 2ï¼šä½¿ç”¨Kraken 2å¯¹æµ‹åºæ•°æ®è¿›è¡Œåˆ†ç±»éœ€è¦ä½¿ç”¨ä»¥ä¸‹å‘½ä»¤ï¼š

`kraken2 \--db \<path_to_database\> \<input_file\> \--output \<output_file\>`

è¿™é‡Œï¼Œ**`<path_to_database>`**æ˜¯å‚è€ƒæ•°æ®åº“çš„è·¯å¾„ï¼Œ**`<input_file>`**æ˜¯éœ€è¦è¿›è¡Œåˆ†ç±»çš„è¾“å…¥æ–‡ä»¶ï¼Œ**`<output_file>`**æ˜¯è¾“å‡ºæ–‡ä»¶çš„åç§°ã€‚Kraken 2å°†è¾“å‡ºä¸€ä¸ªåˆ†ç±»æŠ¥å‘Šæ–‡ä»¶å’Œä¸€ä¸ªåºåˆ—æ–‡ä»¶ã€‚

`kraken2-build --standard --threads 24 --db ./`

\--standardæ ‡å‡†æ¨¡å¼ä¸‹åªä¸‹è½½5ç§æ•°æ®åº“ï¼šå¤èŒarchaeaã€ç»†èŒbacteriaã€äººç±»humanã€è½½ä½“UniVec_Coreã€ç—…æ¯’viralã€‚ä¹Ÿå¯é€‰ç›´æ¥ä¸‹è½½ä½œè€…æ„å»ºçš„ç´¢å¼•ï¼Œè¿˜åŒ…æ‹¬brackençš„ç´¢å¼•ã€‚

è¿™ä¸ªkrakenæ•°æ®åº“æ˜¯å¯ä»¥è‡ªå·±æ„å»ºçš„ï¼Œæ‰€ä»¥é€‚ç”¨äºå„ç§é¡¹ç›®çš„ç‰©ç§æ³¨é‡Šï¼Œæˆ‘åšçš„æ¯”è¾ƒå¤šçš„æ˜¯ç¯å¢ƒæ ·æœ¬çš„å®åŸºå› ç»„ï¼Œå°±å¯èƒ½éœ€è¦æ›´å…¨é¢çš„ç‰©ç§æ•°æ®åº“ï¼ˆç”šè‡³é™¤äº†å„ç§å¾®ç”Ÿç‰©ï¼Œè¿˜è¦åŠ¨æ¤ç‰©æ•°æ®ç­‰ï¼‰ï¼Œå®éªŒå®¤çš„WXå¸ˆå§æ”¶é›†æ„å»ºäº†ä¸€ä¸ªè¶…å¤§çš„ç‰©ç§åº“ã€‚

éœ€è¦æ³¨æ„çš„æ˜¯krakenè¿è¡Œè‡³å°‘è¦æä¾›æ•°æ®åº“å¤§å°çš„å†…å­˜å¤§å°ï¼ˆè¿è¡Œå†…å­˜ï¼‰ï¼Œå› ä¸ºå®ƒä¼šæŠŠæ•´ä¸ªæ•°æ®åº“è½½å…¥å†…å­˜åè¿›è¡Œåºåˆ—çš„æ³¨é‡Šï¼Œæ‰€ä»¥å¦‚æœå‘ç°æ— æ³•è½½å…¥æ•°æ®åº“çš„æŠ¥é”™ï¼Œå¯ä»¥å°è¯•è°ƒå¤§å†…å­˜èµ„æºã€‚

krakenè½¯ä»¶è¿è¡Œæ—¶è½½å…¥æ•°æ®åº“æ˜¯ä¸€ä¸ªååˆ†è€—æ—¶çš„æ­¥éª¤ï¼Œè€Œæ¯æ¡åºåˆ—çš„é‰´å®šæ—¶é—´å·®ä¸å¤šï¼Œæ‰€ä»¥æˆ‘ä»¬å¯ä»¥å°†å¾ˆå¤šæ ·æœ¬çš„fastqæ–‡ä»¶åˆå¹¶æˆä¸€ä¸ªå¤§æ–‡ä»¶åè¾“å…¥krakenæ³¨é‡Šï¼Œä¹‹åå†æŒ‰ç…§åºåˆ—çš„æ•°é‡æ‹†åˆ†ç»“æœæ–‡ä»¶ï¼Œè¿™æ ·å¤šä¸ªæ ·æœ¬ä¹Ÿåªéœ€è¦è½½å…¥ä¸€æ¬¡æ•°æ®åº“ï¼ŒèŠ‚çœæ—¶é—´ã€‚

```{bash}
#!/bin/bash
#SBATCH --job-name=kraken2M
#SBATCH --output=/share/home/jianglab/pengchen/work/asthma/kraken/%x_%a.out
#SBATCH --error=/share/home/jianglab/pengchen/work/asthma/kraken/%x_%a.err
#SBATCH --time=14-00:00:00
#SBATCH --partition=mem
#SBATCH --cpus-per-task=32
#SBATCH --mem-per-cpu=100G

fqp=~/work/asthma/data/CRR205159/
python /share/home/jianglab/shared/krakenDB/K2ols/kraken2M.py -t 32 \
    -i ${fqp} \
    -c 0.05 \
    -s _f1.fastq,_r2.fastq \
    -o ~/work/asthma/kraken/ \
    -d /share/home/jianglab/shared/krakenDB/mydb2 \
    -k ~/miniconda3/envs/waste/bin/kraken2 \
    -kt /share/home/jianglab/shared/krakenDB/K2ols/KrakenTools 
```

#### è¾“å‡ºæ–‡ä»¶æ ¼å¼

Krakenæ ‡å‡†è¾“å‡ºæ ¼å¼

äº”åˆ—è¡¨ output

-   C/Uä»£è¡¨åˆ†ç±»classifiedæˆ–éåˆ†ç±»unclassifed

-   åºåˆ—ID

-   ç‰©ç§æ³¨é‡Š

-   æ¯”åºåˆ—æ³¨é‡Šçš„åŒºåŸŸï¼Œå¦‚98\|94ä»£è¡¨å·¦ç«¯98bpï¼Œå³ç«¯94bpæ¯”å¯¹è‡³æ•°æ®åº“

-   LCAæ¯”å¯¹ç»“æœï¼Œå¦‚"562:13 561:4"ä»£è¡¨13 k-meræ¯”å¯¹è‡³ç‰©ç§#562ï¼Œ4 k-meræ¯”å¯¹è‡³#561ç‰©ç§

æŠ¥å‘Šè¾“å‡ºæ ¼å¼ report

åŒ…æ‹¬6åˆ—ï¼Œæ–¹ä¾¿æ•´ç†ä¸‹æ¸¸åˆ†æã€‚

1.  ç™¾åˆ†æ¯”

2.  count

3.  countæœ€ä¼˜

4.  (U)nclassified, (R)oot, (D)omain, (K)ingdom, (P)hylum, (C)lass, (O)rder, (F)amily, (G)enus, or (S)pecies. "G2"ä»£è¡¨ä½äºå±ä¸€ç§é—´

5.  NCBIç‰©ç§ID

6.  ç§‘å­¦ç‰©ç§å

å¸¸ç”¨çš„ç‰©ç§ä¸°åº¦è¡¨æ ¼å¼é™¤äº†kraken reportï¼Œè¿˜æœ‰mpaï¼Œspfï¼Œkronaç­‰æ ¼å¼ï¼Œå…³äºkrakenç»“æœçš„æ•´ç†ä»¥åŠæ ¼å¼è½¬æ¢æ–¹å¼ï¼Œæœ‰ä¸€äº›ç°æˆçš„è„šæœ¬æˆ–è€…è‡ªå·±å†™ã€‚

[KrakenTools (jhu.edu)](https://ccb.jhu.edu/software/krakentools/) å°±æ˜¯ä¸€å¥—å¾ˆå¥½ç”¨çš„krakenå·¥å…·åŒ…ï¼Œå…¶ä¸­å¸¸ç”¨çš„æœ‰ï¼š

1.  extract_kraken_reads.py

æ­¤ç¨‹åºæå–è¯»å–åœ¨ä»»ä½•ç”¨æˆ·æŒ‡å®šçš„åˆ†ç±»idå¤„åˆ†ç±»çš„å†…å®¹ã€‚ç”¨æˆ·å¿…é¡»æŒ‡å®šKrakenè¾“å‡ºæ–‡ä»¶ã€åºåˆ—æ–‡ä»¶å’Œè‡³å°‘ä¸€ä¸ªåˆ†ç±»æ³•IDã€‚ä¸‹é¢æŒ‡å®šäº†å…¶ä»–é€‰é¡¹ã€‚æˆªè‡³2021å¹´4æœˆ19æ—¥ï¼Œæ­¤è„šæœ¬ä¸KrakenUniq/Kraken2UniqæŠ¥å‘Šå…¼å®¹ã€‚

2.  combine_kreports.py

This script combines multiple Kraken reports into a combined report file.

`python combine_kreports.py`

-   -r 1.KREPORT 2.KREPORT\...\...\...\...\...\...\...\...Kraken-style reports to combine

-   -o COMBINED.KREPORT\...\...\...\...\...\...\...\...\...Output file

3.  kreport2krona.py

This program takes a Kraken report file and prints out a krona-compatible TEXT file

æ¢æˆkronaæ–‡ä»¶å¥½ç”»å›¾ã€‚å˜¿å˜¿

kronaè£…äº†ä¸€ä¸ªexcelçš„æ’ä»¶å¯ä»¥å¾ˆå®¹æ˜“ç”»å›¾\
`python kreport2krona.py`

-   -r/\--report MYFILE.KREPORT\...\.....Kraken report file

-   -o/\--output MYFILE.KRONA\...\...\....Output Krona text file

then, `ktImportText MYSAMPLE.krona -o MYSAMPLE.krona.html`

å¥½çœ‹çš„ç½‘é¡µå°±å‡ºæ¥äº†ã€‚

4.  kreport2mpa.py

This program takes a Kraken report file and prints out a mpa (MetaPhlAn) -style TEXT file

`python kreport2mpa.py`

-   -r/\--report MYFILE.KREPORT\...\.....Kraken report file

-   -o/\--output MYFILE.MPA.TXT\...\.....Output MPA-STYLE text file

5.  combine_mpa.py

`python combine_mpa.py`

-   -i/\--input MYFILE1.MPA MYFILE2.MPA\...\....Multiple MPA-STYLE text files (separated by spaces)

-   -o/\--output MYFILE.COMBINED.MPA\...\...\....Output MPA-STYLE text file

This program combines multiple outputs from [kreport2mpa.py](#kreport2mpapy). Files to be combined must have been generated using the same kreport2mpa.py options.

`python combine_mpa.py -i -o --intermediate-ranks`

### HUMAnN

HUMAnN2ï¼ˆThe HMP Unified Metabolic Analysis Network 2ï¼‰æ˜¯ä¸€æ¬¾ç”¨äºåˆ†æäººç±»å¾®ç”Ÿç‰©ç»„çš„åŠŸèƒ½å’Œä»£è°¢èƒ½åŠ›çš„å·¥å…·ã€‚å®ƒé€šè¿‡å°†å®åŸºå› ç»„åºåˆ—ä¸å‚è€ƒåŸºå› ç»„æ•°æ®åº“æ¯”å¯¹ï¼Œåˆ©ç”¨MetaCycä»£è°¢é€šè·¯æ•°æ®åº“å’ŒUniRefè›‹ç™½è´¨åºåˆ—æ•°æ®åº“ï¼Œåˆ†æå¾®ç”Ÿç‰©ç»„åœ¨åŠŸèƒ½å’Œä»£è°¢é€šè·¯æ°´å¹³ä¸Šçš„ç»„æˆå’Œæ´»æ€§ã€‚HUMAnN2è¿˜æä¾›äº†å¤šæ ·æ€§åˆ†æã€å…³è”åˆ†æå’Œå¯è§†åŒ–å·¥å…·ï¼Œå¯ç”¨äºæ·±å…¥ç ”ç©¶äººç±»å¾®ç”Ÿç‰©ç»„å¯¹å®¿ä¸»å¥åº·çš„å½±å“å’Œæ²»ç–—ç­–ç•¥çš„åˆ¶å®šç­‰æ–¹é¢ã€‚

HUMAnN2æ˜¯ç”±ç¾å›½å›½å®¶äººç±»å¾®ç”Ÿç‰©ç»„è®¡åˆ’ï¼ˆHMPï¼‰å¼€å‘çš„ï¼Œç›®å‰æœ€æ–°ç‰ˆæœ¬ä¸º[HUMAnN3](https://github.com/biobakery/humann)ï¼Œäº2020å¹´å‘å¸ƒã€‚ä¸HUMAnN2ç›¸æ¯”ï¼ŒHUMAnN3æ”¹è¿›äº†åŸºå› å®¶æ—æ³¨é‡Šçš„æ–¹æ³•ï¼Œæé«˜äº†æ³¨é‡Šç²¾åº¦å’Œé€Ÿåº¦ï¼Œå¹¶æä¾›äº†æ–°çš„åŠŸèƒ½å’Œå·¥å…·ï¼Œå¦‚åŠŸèƒ½éŸ§åº¦åˆ†æã€ä»£è°¢æŒ‡çº¹è¯†åˆ«å’Œå¤šæ ·æ€§åˆ†æç­‰ã€‚

ä½†æ˜¯HUMAnN2çš„æ•°æ®åº“åŸºæœ¬éƒ½æ˜¯ä¸äººç›¸å…³çš„å¾®ç”Ÿç‰©ï¼Œæ¯”è¾ƒé€‚åˆåšå„ç§äººä½“å¾®ç”Ÿç‰©ç»„ï¼ˆè‚ é“ï¼Œè‚ºéƒ¨ï¼Œå£è…”ï¼Œçš®è‚¤ç­‰ç­‰ï¼‰ï¼Œå¯¹äºç¯å¢ƒæ ·æœ¬å¯èƒ½unclassifiedæ¯”è¾ƒå¤šã€‚

**HUMAnN2è¦æ±‚åŒç«¯åºåˆ—åˆå¹¶çš„æ–‡ä»¶ä½œä¸ºè¾“å…¥**ï¼Œforå¾ªç¯æ ¹æ®å®éªŒè®¾è®¡æ ·æœ¬åæ‰¹é‡åŒç«¯åºåˆ—åˆå¹¶ã€‚

-   **ç‰©ç§ç»„æˆè°ƒç”¨MetaPhlAn2, bowtie2æ¯”å¯¹è‡³æ ¸é…¸åºåˆ—**ï¼Œè§£å†³æœ‰å“ªäº›å¾®ç”Ÿç‰©å­˜åœ¨çš„é—®é¢˜ï¼›

-   **åŠŸèƒ½ç»„æˆä¸ºhumann2è°ƒç”¨diamondæ¯”å¯¹è‡³è›‹ç™½åº“11Gb**ï¼Œè§£å†³è¿™äº›å¾®ç”Ÿç‰©å‚ä¸å“ªäº›åŠŸèƒ½é€šè·¯çš„é—®é¢˜ï¼›

```{bash}
cd alldata
for i in `cat ~/work/asthma/data/namelist`
do
echo $i
cat ${i}_f1.fastq ${i}_r2.fastq >${i}_paired.fastq
done

#!/bin/bash
#SBATCH --job-name=humann2
#SBATCH --output=/share/home/jianglab/pengchen/work/asthma/humann/%x_%a.out
#SBATCH --error=/share/home/jianglab/pengchen/work/asthma/humann/%x_%a.err
#SBATCH --array=1-32
#SBATCH --cpus-per-task=24
#SBATCH --partition=cpu

echo start: `date +'%Y-%m-%d %T'`
start=`date +%s`
##############
myarray=(`cat ~/work/asthma/data/namelist`)
echo $SLURM_ARRAY_TASK_ID
#this is your single file name
sample=${myarray[${SLURM_ARRAY_TASK_ID}]}
echo handling: $sample
humann2 --input data/alldata/${sample}_paired.fastq  \
  --output temp/humann2/ --threads 24
  
## é“¾æ¥é‡è¦æ–‡ä»¶è‡³humann2ç›®å½•
ln temp/humann2/${sample}_paired_humann2_temp/${sample}_paired_metaphlan_bugs_list.tsv temp/humann2/
## åˆ é™¤ä¸´æ—¶æ–‡ä»¶
rm -rf temp/humann2/${sample}_paired_humann2_temp
##############
echo end: `date +'%Y-%m-%d %T'`
end=`date +%s`
echo TIME:`expr $end - $start`s


## åˆå¹¶ã€ä¿®æ­£æ ·æœ¬åã€é¢„è§ˆ
merge_metaphlan_tables2.py \
  temp/humann2/*_metaphlan_bugs_list.tsv | \
  sed 's/_metaphlan_bugs_list//g' \
  > metaphlan2/taxonomy.tsv
```

## contigs-based

### ç»„è£…ï¼šmegahit

MegaHitæ˜¯ä¸€ä¸ªç”¨äºå¯¹é«˜é€šé‡æµ‹åºæ•°æ®è¿›è¡Œde novoç»„è£…çš„è½¯ä»¶ã€‚å®ƒä½¿ç”¨äº†ä¸€ç§åŸºäºçŸ­è¯»æ¯”å¯¹å’Œå›¾å½¢æ„å»ºçš„ç®—æ³•æ¥ç»„è£…åŸºå› ç»„ï¼Œå¯ä»¥é«˜æ•ˆåœ°å¤„ç†å¤§è§„æ¨¡çš„æ•°æ®é›†ã€‚ä»¥ä¸‹æ˜¯MegaHitçš„ä¸€äº›ä¼˜ç‚¹å’Œé€‚ç”¨æƒ…å†µï¼š

1.  é€Ÿåº¦å¿«ï¼šMegaHitçš„ç®—æ³•éå¸¸é«˜æ•ˆï¼Œå¯ä»¥å¤„ç†å¤§è§„æ¨¡çš„æ•°æ®é›†ï¼Œé€šå¸¸æ¯”å…¶ä»–de novoç»„è£…å·¥å…·æ›´å¿«ã€‚

2.  é«˜è´¨é‡çš„ç»„è£…ï¼šMegaHitåœ¨ç»„è£…ç»“æœçš„è¿é€šæ€§å’Œå‡†ç¡®æ€§æ–¹é¢è¡¨ç°ä¼˜å¼‚ï¼Œå°¤å…¶åœ¨å¤„ç†é«˜GCå«é‡åŸºå› ç»„æ—¶æ•ˆæœæ˜¾è‘—ã€‚

3.  é€‚ç”¨äºä¸åŒç±»å‹çš„æµ‹åºæ•°æ®ï¼šMegaHitæ”¯æŒå¤šç§ä¸åŒç±»å‹çš„æµ‹åºæ•°æ®ï¼ŒåŒ…æ‹¬ Illumina HiSeq/MiSeqã€IonTorrentå’ŒPacBioç­‰å¹³å°ã€‚

4.  æ˜“äºä½¿ç”¨ï¼šMegaHitå…·æœ‰ç®€å•çš„å‘½ä»¤è¡Œè¯­æ³•ï¼Œæ–¹ä¾¿ç”¨æˆ·è¿›è¡Œç»„è£…æ“ä½œï¼Œä¸”å…·æœ‰ä¸­æ–­ç‚¹ï¼Œé¿å…å¤±è´¥åå…¨éƒ¨é‡è·‘ã€‚

```{bash}
#!/bin/bash
#SBATCH --job-name=asthma_megahit
#SBATCH --output=/share/home/jianglab/pengchen/work/asthma/megahit/log/%x_%a.out
#SBATCH --error=/share/home/jianglab/pengchen/work/asthma/megahit/log/%x_%a.err
#SBATCH --array=1-33
#SBATCH --partition=cpu
#SBATCH --cpus-per-task=32

echo start: `date +'%Y-%m-%d %T'`
start=`date +%s`
echo "SLURM_ARRAY_TASK_ID: " $SLURM_ARRAY_TASK_ID
sample=$(head -n $SLURM_ARRAY_TASK_ID ~/work/asthma/data/namelist | tail -1)
#sample=$(head -n 1 namelist | tail -1)
echo handling: $sample	
####################
megahit -t 32 -1 ~/work/asthma/data/$sample/$sample'_f1.fastq' \
-2 ~/work/asthma/data/$sample/$sample'_r2.fastq' -o ~/work/asthma/megahit/$sample --out-prefix $sample
####################
echo end: `date +'%Y-%m-%d %T'`
end=`date +%s`
echo TIME:`expr $end - $start`s
```

### ç»„è£…è¯„ä¼°ï¼šQUAST

**QUAST**ä»£è¡¨è´¨é‡è¯„ä¼°å·¥å…·ã€‚ QUASTå¯ä»¥ä½¿ç”¨å‚è€ƒåŸºå› ç»„ä»¥åŠä¸ä½¿ç”¨å‚è€ƒåŸºå› ç»„æ¥è¯„ä¼°è£…é…ã€‚ QUASTç”Ÿæˆè¯¦ç»†çš„æŠ¥å‘Šï¼Œè¡¨æ ¼å’Œå›¾è§£ï¼Œä»¥æ˜¾ç¤ºè£…é…çš„ä¸åŒæ–¹é¢ã€‚

### åŸºå› é¢„æµ‹ï¼šProdigal

è¾“å…¥æ–‡ä»¶ï¼šæ‹¼è£…å¥½çš„åºåˆ—æ–‡ä»¶ megahit/final.contigs.fa

è¾“å‡ºæ–‡ä»¶ï¼šprodigalé¢„æµ‹çš„åŸºå› åºåˆ— prodigal/gene.fa

prodigalä¸æ”¯æŒå¤šçº¿ç¨‹è¿è¡Œï¼Œæ‰€ä»¥æˆ‘ä»¬å¯ä»¥è‡ªè¡Œåˆ†å‰²åºåˆ—æ–‡ä»¶è°ƒç”¨å¤šä¸ªprodigalç¨‹åºåˆ†åˆ«è·‘å®ç°ä¼ªå¤šçº¿ç¨‹ã€‚

```{bash}
#!/bin/bash
#SBATCH --job-name=prodigal
#SBATCH --output=/share/home/jianglab/pengchen/work/asthma/prodigal/log/%x_%a.out
#SBATCH --error=/share/home/jianglab/pengchen/work/asthma/prodigal/log/%x_%a.err
#SBATCH --array=1-33
#SBATCH --partition=cpu
#SBATCH --cpus-per-task=1

echo start: `date +'%Y-%m-%d %T'`
start=`date +%s`
echo "SLURM_ARRAY_TASK_ID: " $SLURM_ARRAY_TASK_ID
sample=$(head -n $SLURM_ARRAY_TASK_ID ~/work/asthma/data/namelist | tail -1)
#sample=$(head -n 1 namelist | tail -1)
echo handling: $sample
####################
prodigal -i ~/work/asthma/megahit/contigs/$sample.fa \
    -d ~/work/asthma/prodigal/$sample.gene.fa \
    -o ~/work/asthma/prodigal/$sample.gene.gff \
    -p meta -f gff 
    
grep 'partial=00' ~/work/asthma/prodigal/$sample.gene.fa | cut -f1 -d ' '| sed 's/>//' > ~/work/asthma/prodigal/$sample.fullid
seqkit grep -f ~/work/asthma/prodigal/$sample.fullid ~/work/asthma/prodigal/$sample.gene.fa > ~/work/asthma/prodigal/fullgene/$sample.gene.fa

####################
echo end: `date +'%Y-%m-%d %T'`
end=`date +%s`
echo TIME:`expr $end - $start`s

============================================================================
## ç»Ÿè®¡åŸºå› æ•°é‡
grep -c '>' temp/prodigal/gene.fa 
## ç»Ÿè®¡å®Œæ•´åŸºå› æ•°é‡ï¼Œæ•°æ®é‡å¤§å¯åªç”¨å®Œæ•´åŸºå› éƒ¨åˆ†
grep -c 'partial=00' temp/prodigal/gene.fa 
## æå–å®Œæ•´åŸºå› (å®Œæ•´ç‰‡æ®µè·å¾—çš„åŸºå› å…¨ä¸ºå®Œæ•´ï¼Œå¦‚æˆç¯çš„ç»†èŒåŸºå› ç»„)
grep 'partial=00' temp/prodigal/gene.fa | cut -f1 -d ' '| sed 's/>//' > temp/prodigal/full_length.id
seqkit grep -f temp/prodigal/full_length.id temp/prodigal/gene.fa > temp/prodigal/full_length.fa
seqkit stat temp/prodigal/full_length.fa
```

### å»å†—ä½™

ä¸Šé¢äº§ç”Ÿäº†nä¸ªæ ·æœ¬çš„åŸºå› é¢„æµ‹ç»“æœæ–‡ä»¶ï¼Œgene.faæ–‡ä»¶è¦æƒ³åŠæ³•æ•´åˆä¸ºä¸€ä¸ªæ–‡ä»¶å†å»å»å†—ä½™ã€‚

```{bash}
#!/bin/bash
#ä¿®æ”¹æ¯æ¡åºåˆ—çš„åç§°ï¼ŒåŠ ä¸Šæ ·æœ¬å
for i in `cat ~/work/asthma/data/namelist`
do
echo $i
sed -i "/>/s/>/>${i}_/" $i.gene.fa 
done
echo 'start merge'
cat *.gene.fa>all.fullgene.fa
echo 'done'
```

#### Cd-hit
```{bash}
#!/bin/bash
#SBATCH --job-name=cdhit
#SBATCH --output=/share/home/jianglab/pengchen/work/asthma/%x_%a.out
#SBATCH --error=/share/home/jianglab/pengchen/work/asthma/%x_%a.err
#SBATCH --cpus-per-task=32
#SBATCH --partition=short

echo start: `date +'%Y-%m-%d %T'`
start=`date +%s`
##############
## è¾“å…¥æ–‡ä»¶ï¼šprodigalé¢„æµ‹çš„åŸºå› åºåˆ— all.fullgene.fa
## è¾“å‡ºæ–‡ä»¶ï¼šå»å†—ä½™åçš„åŸºå› å’Œè›‹ç™½åºåˆ—ï¼šNR/nucleotide.fa;NR/protein.fa
mkdir NR
## aSè¦†ç›–åº¦ï¼Œcç›¸ä¼¼åº¦ï¼ŒGå±€éƒ¨æ¯”å¯¹ï¼Œgæœ€ä¼˜è§£ï¼ŒTå¤šçº¿ç¨‹ï¼ŒMå†…å­˜0ä¸é™åˆ¶
## 2ä¸‡åŸºå› 2mï¼Œ2åƒä¸‡éœ€è¦2000hï¼Œå¤šçº¿ç¨‹å¯åŠ é€Ÿ	
cd-hit-est -i prodigal/fullgene/all.fullgene.fa \
    -o NR/nucleotide.fa \
    -aS 0.9 -c 0.9 -G 0 -g 0 -T 0 -M 0
## ç»Ÿè®¡éå†—ä½™åŸºå› æ•°é‡ï¼Œå•æ¬¡æ‹¼æ¥ç»“æœæ•°é‡ä¸‹é™ä¸å¤§ï¼Œå¤šæ‰¹æ‹¼æ¥å†—ä½™åº¦é«˜

echo 'after remove, the number of genes: '
grep -c '>' NR/nucleotide.fa
## ç¿»è¯‘æ ¸é…¸ä¸ºå¯¹åº”è›‹ç™½åºåˆ—ï¼Œemboss
## emboss transeqå·¥å…·ï¼Œ93.9 MB
conda install emboss -y

transeq -sequence NR/nucleotide.fa \
  -outseq NR/protein.fa -trim Y 
## åºåˆ—åè‡ªåŠ¨æ·»åŠ äº†_1ï¼Œä¸ºä¸æ ¸é…¸å¯¹åº”è¦å»é™¤
sed -i 's/_1 / /' NR/protein.fa
##############
echo end: `date +'%Y-%m-%d %T'`
end=`date +%s`
echo TIME:`expr $end - $start`s
```
#### mmseq2

è¿™ä¸ªè¦æ¯”cd-hitå¿«éå¸¸å¤š
```{bash}
#mmseqs2
##min-seq-id:identityï¼Œcè¦†ç›–åº¦ï¼Œrescore-mode3: global alignment

#èšç±»
input_fa=tmp_com.gene
mmseqs easy-linclust $input_fa lin_res tmp \
    --min-seq-id 0.9 -c 0.9 --cov-mode 1  --threads 8
#æˆ–è€…
mmseqs easy-cluster $input_fa lin_res tmp \
    --min-seq-id 0.9 -c 0.9 --cov-mode 1  --threads 8

##ä»¥ä¸‹æ˜¯ä¸ºäº†ç†è§£æ­¥éª¤åšçš„ï¼Œä½¿ç”¨ä¸Šé¢ğŸ‘†çš„easyæ¨¡å¼å³å¯
#å»ºåº“
input_fa=../prodigal/C1.gene.fa
DB=C1.geneDB
DB_clu=mmseq_out
mmseqs createdb $input_fa $DB
#èšç±»
mmseqs cluster $DB $DB_clu tmp \
    --min-seq-id 0.9 -c 0.9 --cov-mode 1  --threads 8  --rescore-mode 3
#è¾“å‡ºè½¬æ¢
#ç¬¬ä¸€åˆ—æ˜¯ä»£è¡¨åºåˆ—idï¼Œç¬¬äºŒåˆ—æ˜¯æˆå‘˜åºåˆ—id
#mmseqs createtsv $DB $DB mmseq_out mmseq_out.tsv

#mmseqs createseqfiledb $DB $DB_clu ${DB_clu}_seq
#mmseqs result2flat $DB $DB ${DB_clu}_seq ${DB_clu}_seq.fasta

#è·å–represent sequence
mmseqs createsubdb $DB_clu $DB ${DB_clu}_rep
mmseqs convert2fasta ${DB_clu}_rep ${DB_clu}_rep.fasta   
```


### åŸºå› å®šé‡ï¼šsalmon

1.  å»ºç«‹ç´¢å¼•

```{bash}
#!/bin/bash
#SBATCH --job-name=salmon-index
#SBATCH --output=/share/home/jianglab/pengchen/work/asthma/%x_%a.out
#SBATCH --error=/share/home/jianglab/pengchen/work/asthma/%x_%a.err
#SBATCH --cpus-per-task=32

#SBATCH --partition=short

echo start: `date +'%Y-%m-%d %T'`
start=`date +%s`
##############
mkdir -p temp/salmon

## å»ºç´¢å¼•, -tåºåˆ—, -i ç´¢å¼•ï¼Œ10s
salmon index \
  -t NR/nucleotide.fa \
  -p 32 \
  -i temp/salmon/index 
##############
echo end: `date +'%Y-%m-%d %T'`
end=`date +%s`
echo TIME:`expr $end - $start`s
```

2.  å¯¹æ¯ä¸ªæ ·æœ¬å®šé‡

```{bash}
#!/bin/bash
#SBATCH --job-name=salmon
#SBATCH --output=/share/home/jianglab/pengchen/work/asthma/salmon/log/%x_%a.out
#SBATCH --error=/share/home/jianglab/pengchen/work/asthma/salmon/log/%x_%a.err
#SBATCH --array=1-33
#SBATCH --partition=cpu
#SBATCH --cpus-per-task=32

echo start: `date +'%Y-%m-%d %T'`
start=`date +%s`
echo "SLURM_ARRAY_TASK_ID: " $SLURM_ARRAY_TASK_ID
sample=$(head -n $SLURM_ARRAY_TASK_ID ~/work/asthma/data/namelist | tail -1)
#sample=$(head -n 1 namelist | tail -1)
echo handling: $sample
####################
## è¾“å…¥æ–‡ä»¶ï¼šå»å†—ä½™åçš„åŸºå› å’Œè›‹ç™½åºåˆ—ï¼šNR/nucleotide.fa
## è¾“å‡ºæ–‡ä»¶ï¼šSalmonå®šé‡åçš„ç»“æœï¼šsalmon/gene.count;salmon/gene.TPM
## å®šé‡ï¼Œlæ–‡åº“ç±»å‹è‡ªåŠ¨é€‰æ‹©ï¼Œpçº¿ç¨‹ï¼Œ--metaå®åŸºå› ç»„æ¨¡å¼
salmon quant \
    -i temp/salmon/index -l A -p 32 --meta \
    -1 data/alldata/${sample}_f1.fastq \
    -2 data/alldata/${sample}_r2.fastq \
    -o temp/salmon/${sample}.quant
####################
echo end: `date +'%Y-%m-%d %T'`
end=`date +%s`
echo TIME:`expr $end - $start`s
```

3.  åˆå¹¶å„æ ·æœ¬ç»“æœ

```{bash}
## åˆå¹¶
mkdir -p salmon
salmon quantmerge \
    --quants temp/salmon/*.quant \
    -o salmon/gene.TPM
salmon quantmerge \
    --quants temp/salmon/*.quant \
    --column NumReads -o salmon/gene.count
sed -i '1 s/.quant//g' salmon/gene.*

## é¢„è§ˆç»“æœè¡¨æ ¼
head -n3 salmon/gene.*
```

### åŠŸèƒ½åŸºå› æ³¨é‡Š

ä¸Šä¸€æ­¥å·²ç»æœ‰äº†æ‰€æœ‰çš„åŸºå› å’Œæ¯ä¸ªæ ·æœ¬æ‰€æœ‰åŸºå› çš„read countå®šé‡ç»“æœï¼Œæˆ‘ä»¬åªéœ€è¦å¯¹ä¸Šä¸€æ­¥çš„åŸºå› åºåˆ—ï¼ˆæˆ–è›‹ç™½è´¨åºåˆ—ï¼‰è¿›è¡Œä¸åŒæ•°æ®åº“çš„æ³¨é‡Šï¼ˆå¾ˆå¤šè½¯ä»¶éƒ½æ˜¯ç”¨diamondæ¯”å¯¹ï¼Œå¦‚æœæ²¡æœ‰ä¸“ç”¨è½¯ä»¶çš„æ•°æ®åº“æˆ‘ä»¬ä¹Ÿå¯ä»¥è‡ªå·±ç”¨diamondæ¯”å¯¹ï¼‰ï¼Œåˆå¹¶æ³¨é‡Šç»“æœå¾—åˆ°çš„å°±æ˜¯åŠŸèƒ½ä¸°åº¦è¡¨ã€‚

diamondé€‰æ‹©\--outfmt 6çš„è¾“å‡ºç»“æœå’Œblastpä¸€æ ·ï¼š

|      |          |                                              |
|------|----------|----------------------------------------------|
| 1\.  | qseqid   | query sequence id                            |
| 2\.  | sseqid   | subject (e.g., reference genome) sequence id |
| 3\.  | pident   | percentage of identical matches              |
| 4\.  | length   | alignment length                             |
| 5\.  | mismatch | number of mismatches                         |
| 6\.  | gapopen  | number of gap openings                       |
| 7\.  | qstart   | start of alignment in query                  |
| 8\.  | qend     | end of alignment in query                    |
| 9\.  | sstart   | start of alignment in subject                |
| 10\. | send     | end of alignment in subject                  |
| 11\. | evalue   | expect value                                 |
| 12\. | bitscore | bit score                                    |

#### 1 eggNOG(COG/KEGG/CAZy)

[EggNOGæ•°æ®åº“](http://eggnogdb.embl.de/)æ”¶é›†äº†COGï¼ˆClusters of Orthologous Groups of proteinsï¼Œç›´ç³»åŒæºè›‹ç™½ç°‡ï¼‰,æ„æˆæ¯ä¸ªCOGçš„è›‹ç™½éƒ½æ˜¯è¢«å‡å®šä¸ºæ¥è‡ªäºä¸€ä¸ªç¥–å…ˆè›‹ç™½ï¼Œå› æ­¤æ˜¯orthologsæˆ–è€…æ˜¯paralogsã€‚é€šè¿‡æŠŠæ‰€æœ‰å®Œæ•´åŸºå› ç»„çš„ç¼–ç è›‹ç™½ä¸€ä¸ªä¸€ä¸ªçš„äº’ç›¸æ¯”è¾ƒç¡®å®šçš„ã€‚åœ¨è€ƒè™‘æ¥è‡ªä¸€ä¸ªç»™å®šåŸºå› ç»„çš„è›‹ç™½æ—¶ï¼Œè¿™ç§æ¯”è¾ƒå°†ç»™å‡ºæ¯ä¸ªå…¶ä»–åŸºå› ç»„çš„ä¸€ä¸ªæœ€ç›¸ä¼¼çš„è›‹ç™½ï¼ˆå› æ­¤éœ€è¦ç”¨å®Œæ•´çš„åŸºå› ç»„æ¥å®šä¹‰COGï¼‰ï¼Œè¿™äº›åŸºå› çš„æ¯ä¸€ä¸ªéƒ½è½®ç•ªåœ°è¢«è€ƒè™‘ã€‚å¦‚æœåœ¨è¿™äº›è›‹ç™½ï¼ˆæˆ–å­é›†ï¼‰ä¹‹é—´ä¸€ä¸ªç›¸äº’çš„æœ€ä½³åŒ¹é…å…³ç³»è¢«å‘ç°ï¼Œé‚£ä¹ˆé‚£äº›ç›¸äº’çš„æœ€ä½³åŒ¹é…å°†å½¢æˆä¸€ä¸ªCOGã€‚è¿™æ ·ï¼Œä¸€ä¸ªCOGä¸­çš„æˆå‘˜å°†ä¸è¿™ä¸ªCOGä¸­çš„å…¶ä»–æˆå‘˜æ¯”èµ·è¢«æ¯”è¾ƒçš„åŸºå› ç»„ä¸­çš„å…¶ä»–è›‹ç™½æ›´ç›¸åƒã€‚

[EggNOG](http://eggnogdb.embl.de/)é‡Œé¢åŒ…å«äº†GOï¼ŒKEGGï¼ŒCAZyç­‰ã€‚

```{bash}
## ä¸‹è½½å¸¸ç”¨æ•°æ®åº“ï¼Œæ³¨æ„è®¾ç½®ä¸‹è½½ä½ç½®
mkdir -p ${db}/eggnog5 && cd ${db}/eggnog5
## -yé»˜è®¤åŒæ„ï¼Œ-få¼ºåˆ¶ä¸‹è½½ï¼Œeggnog.db.gz 7.9G+4.9G
download_eggnog_data.py -y -f --data_dir ./

## ä¸‹è½½æ–¹å¼2(å¯é€‰)ï¼šé“¾æ¥ç›´æ¥ä¸‹è½½
wget -c http://eggnog5.embl.de/download/emapperdb-5.0.0/eggnog.db.gz ## 7.9G
wget -c http://eggnog5.embl.de/download/emapperdb-5.0.0/eggnog_proteins.dmnd.gz ## 4.9G
gunzip *.gz
```

```{bash}
#!/bin/bash
#SBATCH --job-name=eggo
#SBATCH --output=/share/home/jianglab/pengchen/work/asthma/%x_%j.out
#SBATCH --error=/share/home/jianglab/pengchen/work/asthma/%x_%j.err
#SBATCH --cpus-per-task=32
#SBATCH --partition=cpu

echo start: `date +'%Y-%m-%d %T'`
start=`date +%s`
##############
#åˆ‡æ¢ç¯å¢ƒ
## diamondæ¯”å¯¹åŸºå› è‡³eggNOG 5.0æ•°æ®åº“, 1~9hï¼Œé»˜è®¤diamond 1e-3
mkdir -p temp/eggnog
emapper.py --no_annot --no_file_comments --override \
  --data_dir ~/db/eggnog5 \
  -i NR/protein.fa \
  --cpu 32 -m diamond \
  -o temp/eggnog/protein
## æ¯”å¯¹ç»“æœåŠŸèƒ½æ³¨é‡Š, 1h
emapper.py \
  --annotate_hits_table temp/eggnog/protein.emapper.seed_orthologs \
  --data_dir ~/db/eggnog5 \
  --cpu 32 --no_file_comments --override \
  -o temp/eggnog/output

## æ·»è¡¨å¤´, 1åˆ—ä¸ºIDï¼Œ9åˆ—KOï¼Œ16åˆ—CAZyï¼Œ21åˆ—COGï¼Œ22åˆ—æè¿°
sed '1 i Name\tortholog\tevalue\tscore\ttaxonomic\tprotein\tGO\tEC\tKO\tPathway\tModule\tReaction\trclass\tBRITE\tTC\tCAZy\tBiGG\ttax_scope\tOG\tbestOG\tCOG\tdescription' \
  temp/eggnog/output.emapper.annotations \
  > temp/eggnog/output
##############
echo end: `date +'%Y-%m-%d %T'`
end=`date +%s`
echo TIME:`expr $end - $start`s
```

#### 2 ç¢³æ°´åŒ–åˆç‰©dbCAN2

```{bash}
## dbCAN2 http://bcb.unl.edu/dbCAN2
## åˆ›å»ºæ•°æ®åº“å­˜æ”¾ç›®å½•å¹¶è¿›å…¥
mkdir -p ${db}/dbCAN2 && cd ${db}/dbCAN2
## ä¸‹è½½åºåˆ—å’Œæè¿°
wget -c http://bcb.unl.edu/dbCAN2/download/CAZyDB.07312020.fa
wget -c http://bcb.unl.edu/dbCAN2/download/Databases/CAZyDB.07302020.fam-activities.txt
## å¤‡ç”¨æ•°æ®åº“ä¸‹è½½åœ°å€å¹¶è§£å‹ 
#wget -c http://210.75.224.110/db/dbcan2/CAZyDB.07312020.fa.gz
#gunzip CAZyDB.07312020.fa.gz
## diamondå»ºç´¢å¼•ï¼Œ800Mï¼Œ1m
diamond --version ## 0.8.22/2.0.5
time diamond makedb \
  --in CAZyDB.07312020.fa \
  --db CAZyDB.07312020

```

```{bash}
## æ¯”å¯¹CAZyæ•°æ®åº“, ç”¨æ—¶2~18m; åŠ --sensitiveæ›´å…¨ä½†æ…¢è‡³1h
mkdir -p temp/dbcan2
diamond blastp   \
	--db ~/db/dbcan2/CAZyDB.07312020  \
	--query NR/protein.fa   \
	--threads 64 -e 1e-5 --outfmt 6 \
	--max-target-seqs 1 --quiet \
	--out temp/dbcan2/gene_diamond.f6
```

#### 3  ARGsï¼ˆCARDï¼‰

\## Github: <https://github.com/arpcard/rgi>

```{bash}
#!/bin/bash
#SBATCH --job-name=rgi
#SBATCH --output=/share/home/jianglab/pengchen/work/asthma/%x_%a.out
#SBATCH --error=/share/home/jianglab/pengchen/work/asthma/%x_%a.err
#SBATCH --cpus-per-task=32
#SBATCH --partition=cpu

echo start: `date +'%Y-%m-%d %T'`
start=`date +%s`
##############
mkdir -p card

rgi main --input_sequence ~/work/asthma/temp/protein.fa \
--output_file card/protein \
--input_type protein --num_threads 32 \
--clean --alignment_tool DIAMOND

##############
echo end: `date +'%Y-%m-%d %T'`
end=`date +%s`
echo TIME:`expr $end - $start`s
```

#### 4 æ¯’åŠ›å› å­VFDB

å®˜ç½‘åœ°å€ï¼š<http://www.mgc.ac.cn/VFs/> åœ¨å®˜ç½‘ä¸‹è½½æ•°æ®åº“æ—¶ï¼Œå¸¦æœ‰setA çš„åº“ä¸ºVFDBæ•°æ®åº“æ ¸å¿ƒåº“(set A)ï¼Œè€ŒsetBä¸ºå…¨åº“(setB), å…¶ä¸­setAä»…åŒ…å«ç»å®éªŒéªŒè¯è¿‡çš„æ¯’åŠ›åŸºå› ï¼Œè€ŒsetBåˆ™åœ¨setAçš„åŸºç¡€ä¸Šå¢åŠ äº†é¢„æµ‹çš„æ¯’åŠ›åŸºå› ï¼Œé€‰æ‹©å¥½æ•°æ®åº“åï¼Œç›´æ¥ç”¨blast/diamondå³å¯å®Œæˆæ³¨é‡Šã€‚

```{bash}
mkdir -p temp/vfdb
diamond blastp   \
	--db ~/db/VFDB/VFDB_setB_pro  \
	--query NR/protein.fa   \
	--threads 32 -e 1e-5 --outfmt 6 \
	--max-target-seqs 1 --quiet \
	--out temp/vfdb/gene_diamond.f6
```


#### 5 å…¶ä»–å„ç§æ•°æ®åº“

### åŠŸèƒ½æ³¨é‡Šåˆå¹¶ 
å†™ä¸€ä¸ªpythonè„šæœ¬ï¼Œå°†è¡¨1ï¼ˆåŸºå› -åŠŸèƒ½çš„å¯¹åº”è¡¨ï¼‰ä¸è¡¨2ï¼ˆåŸºå› ä¸°åº¦è¡¨ï¼‰åˆå¹¶ï¼Œå³ä¸åŒåŸºå› å¯èƒ½æ³¨é‡Šåˆ°ç›¸åŒåŠŸèƒ½ï¼ŒæŠŠå®ƒä»¬çš„ä¸°åº¦åŠ åœ¨ä¸€èµ·å¾—åˆ°æ–°è¡¨3ï¼ˆåŠŸèƒ½ä¸°åº¦è¡¨ï¼‰

## binning

å®åŸºå› ç»„binningæ˜¯æŒ‡å°†ä¸åŒçš„åºåˆ—é›†åˆï¼ˆå¦‚metagenomeåºåˆ—é›†åˆï¼‰æ ¹æ®å®ƒä»¬çš„ç‰©ç§å½’ç±»åˆ°ä¸åŒçš„binsä¸­ï¼Œä»¥ä¾¿è¿›ä¸€æ­¥ç ”ç©¶å®ƒä»¬çš„ç»„æˆå’ŒåŠŸèƒ½ã€‚è¿™ä¸ªè¿‡ç¨‹å¯ä»¥å°†ç±»ä¼¼çš„åºåˆ—ç»„åˆåœ¨ä¸€èµ·ï¼Œå½¢æˆä»£è¡¨ä¸åŒç‰©ç§æˆ–åŸºå› ç»„çš„binsï¼Œä»¥ä¾¿è¿›è¡Œåç»­åˆ†æï¼Œå¦‚ç‰©ç§æ³¨é‡Šã€åŸºå› ç»„ç»„è£…ç­‰ã€‚

ä»¥ä¸‹æ˜¯å¸¸ç”¨çš„å®åŸºå› ç»„binningæ–¹æ³•ï¼š

1.  åŸºäºèšç±»çš„æ–¹æ³•ï¼šè¯¥æ–¹æ³•ä½¿ç”¨åºåˆ—èšç±»å°†ç›¸ä¼¼åºåˆ—åˆ†åˆ°åŒä¸€ä¸ªbinä¸­ã€‚ä¸€èˆ¬æ¥è¯´ï¼Œèšç±»ç®—æ³•å¯åˆ†ä¸ºä¸¤ç±»ï¼šæ— ç›‘ç£èšç±»ï¼ˆå¦‚k-meansã€DBSCANç­‰ï¼‰å’Œæœ‰ç›‘ç£èšç±»ï¼ˆå¦‚CAMIã€MyCCç­‰ï¼‰ã€‚

2.  åŸºäºç»„è£…çš„æ–¹æ³•ï¼šè¯¥æ–¹æ³•ä½¿ç”¨de novoç»„è£…æ¥å°†ç›¸ä¼¼åºåˆ—ç»„è£…æˆè¿ç»­çš„åºåˆ—ï¼Œå†æ ¹æ®è¿™äº›åºåˆ—çš„åŸºå› ç»„ä¿¡æ¯æ¥å°†å…¶åˆ†ç±»åˆ°ä¸åŒçš„binsä¸­ã€‚è¿™ç§æ–¹æ³•çš„ä¼˜ç‚¹æ˜¯å¯ä»¥æ›´å¥½åœ°å¤„ç†é‡å¤åºåˆ—ï¼Œç¼ºç‚¹æ˜¯éœ€è¦å¤§é‡çš„è®¡ç®—èµ„æºå’Œæ—¶é—´ã€‚

3.  åŸºäºåˆ†ç±»å™¨çš„æ–¹æ³•ï¼šè¯¥æ–¹æ³•ä½¿ç”¨æœºå™¨å­¦ä¹ åˆ†ç±»å™¨æ¥å°†åºåˆ—åˆ†é…åˆ°ä¸åŒçš„binsä¸­ã€‚è¿™ç§æ–¹æ³•çš„ä¼˜ç‚¹æ˜¯å¯ä»¥è‡ªåŠ¨å­¦ä¹ ç‰¹å¾å¹¶åœ¨å¤„ç†å¤§è§„æ¨¡æ•°æ®æ—¶æ•ˆç‡é«˜ï¼Œç¼ºç‚¹æ˜¯éœ€è¦å…ˆå»ºç«‹ä¸€ä¸ªåˆ†ç±»å™¨å¹¶è¿›è¡Œè®­ç»ƒã€‚

åœ¨è¿›è¡Œå®åŸºå› ç»„binningæ—¶ï¼Œé€šå¸¸éœ€è¦ä½¿ç”¨å¤šä¸ªæ–¹æ³•è¿›è¡Œæ¯”è¾ƒï¼Œä»¥é€‰æ‹©æœ€é€‚åˆæ•°æ®é›†çš„æ–¹æ³•ã€‚å¯ä»¥ä½¿ç”¨ä¸€äº›æµè¡Œçš„å·¥å…·æ¥è¿›è¡Œbinningï¼Œå¦‚MetaBATã€MaxBinã€CONCOCTå’ŒMEGANç­‰ã€‚è¿™äº›å·¥å…·é€šå¸¸åŒ…å«å„ç§binningæ–¹æ³•ï¼Œå¯ä»¥æ ¹æ®æ•°æ®é›†å’Œåˆ†æç›®çš„é€‰æ‹©é€‚åˆçš„æ–¹æ³•ã€‚

ç¯‡å¹…é™åˆ¶ï¼Œå…·ä½“çš„æ–¹æ³•æ”¾åœ¨å¦ä¸€ç¯‡é‡Œé¢è®²è§£å§ã€‚

## Reference 
