---
title: å®åŸºå› ç»„åˆ†ææµç¨‹ï¼ˆMetagenomic workflowï¼‰202405ï½œæŒç»­æ›´æ–°
author: Peng Chen
date: '2024-05-28'
slug: metagenomic-workflow
categories:
  - metagenomic
tags:
  - microbiome
description: å®åŸºå› ç»„ï¼ˆMetagenomeï¼‰æ˜¯æŒ‡å¯¹ä¸€ä¸ªç”Ÿæ€ç³»ç»Ÿä¸­çš„æ‰€æœ‰å¾®ç”Ÿç‰©è¿›è¡ŒDNAåˆ†æçš„è¿‡ç¨‹ï¼Œå¯ä»¥å¸®åŠ©ç ”ç©¶äººå‘˜äº†è§£å¾®ç”Ÿç‰©çš„å¤šæ ·æ€§ã€åŠŸèƒ½å’Œäº’ä½œå…³ç³»ã€‚è¿™é‡Œä»‹ç»å¸¸ç”¨çš„ä¸Šæ¸¸åˆ†ææµç¨‹ã€‚
image: images/workflow.png
math: ~
license: ~
hidden: no
comments: yes
bibliography: [../../bib/My Library.bib]
link-citations: yes
csl: ../../bib/science.csl
---

## Logs

- å¢åŠ RåŒ…`pctax`å†…çš„ä¸€äº›å¸®åŠ©ä¸Šæ¸¸åˆ†æçš„å°è„šæœ¬ï¼ˆ2024.03.03ï¼‰
- å¢åŠ Mmseqs2ç”¨äºå»å†—ä½™ï¼ŒåŸºå› èšç±»çš„é€Ÿåº¦éå¸¸å¿«ï¼Œä¸”éšåºåˆ—é‡çº¿æ€§å¢é•¿ï¼ˆ2024.03.12ï¼‰
- æ›´æ–°å…¨æ–‡ç»†èŠ‚ï¼ˆ2024.05.29ï¼‰

**æ³¨æ„ï¼š**

- ç¯‡å¹…æœ‰é™ï¼Œæœ¬æ–‡ä¸€èˆ¬ä¸ä»‹ç»å…·ä½“çš„è½¯ä»¶å®‰è£…é…ç½®ï¼Œæ•°æ®åº“ä¸‹è½½ç­‰ï¼Œä¹Ÿä¸æä¾›ç¤ºä¾‹æ•°æ®ï¼Œä¸»è¦å¸®åŠ©å¿«é€Ÿäº†è§£æ•´ä¸ªå®åŸºå› ç»„çš„**ä¸Šæ¸¸åˆ†æ**æµç¨‹ã€‚æ‰€ä»¥è½¯ä»¶å®‰è£…è¯·å‚è€ƒè½¯ä»¶è‡ªèº«ä»‹ç»ï¼Œè¿è¡Œæ—¶é‡åˆ°å…·é—®é¢˜ä¹Ÿæœ€å¥½å…ˆGoogleæˆ–è€…å»github issueé‡Œé¢é—®é—®ã€‚

- å®åŸºå› ç»„åˆ†ææŠ€æœ¯è¿˜æ˜¯åœ¨ä¸åœçš„æ›´æ–°è¿­ä»£ï¼Œå¾ˆå¤šæ›´æ–°æ›´å¥½ç”¨çš„è½¯ä»¶åœ¨ä¸æ–­æ¨å‡ºï¼Œæ‰€ä»¥æˆ‘åœ¨è¿™ä¹Ÿä¼šæŒç»­æ›´æ–°ï¼ˆå¦‚æœæˆ‘è¿˜ä¸€ç›´åšè¿™ä¸ªæ–¹å‘çš„è¯ï¼‰ã€‚è¿™é‡Œä»‹ç»çš„ä¹Ÿåªæ˜¯æ¯”è¾ƒåŸºæœ¬çš„åˆ†ææµç¨‹ï¼Œä½†æ˜¯æŒæ¡äº†ä¹‹åï¼Œè‡ªå·±è¿›ä¸€æ­¥å»åšä¸ªæ€§åŒ–åˆ†æä¹Ÿä¼šé¡ºæ‰‹å¾ˆå¤šã€‚

- å®Œæˆä¸Šæ¸¸åˆ†æåæˆ‘ä»¬å¯ä»¥å¾—åˆ°å„ç§ç‰©ç§æˆ–è€…åŠŸèƒ½çš„profileï¼Œåç»­ä¸€èˆ¬ç”¨pythonï¼ŒRè¿›è¡Œæ•°æ®åˆ†æå’Œå¯è§†åŒ–ï¼Œå¯å‚è€ƒæˆ‘çš„å…¶ä»–åšæ–‡ã€‚

- ç»å¤§å¤šæ•°è¿™é‡Œä»‹ç»çš„è½¯ä»¶éƒ½æ˜¯ä»…æ”¯æŒlinuxå¹³å°çš„ï¼Œæˆ‘ä»¬åšæµ‹åºæ–‡ä»¶ä¸Šæ¸¸åˆ†æä¹Ÿè‚¯å®šæ˜¯åœ¨æœåŠ¡å™¨ä¸Šåšï¼Œä¸ªäººPCä¸€èˆ¬å¾ˆéš¾æ»¡è¶³éœ€æ±‚ï¼Œæ‰€ä»¥åœ¨åšè¿™äº›åˆ†æå‰å¿…é¡»å…ˆå­¦ä¹ linuxåŸºç¡€çŸ¥è¯†å¦‚æ–‡ä»¶ç³»ç»Ÿï¼Œshellè„šæœ¬ç¼–å†™ï¼Œè½¯ä»¶å®‰è£…ç­‰ã€‚å®‰è£…è½¯ä»¶å»ºè®®ä½¿ç”¨condaæˆ–mambaï¼ˆæ–°å»ºç¯å¢ƒå’Œç®¡ç†ï¼‰ï¼Œæœ‰å¾ˆå¤šå‚è€ƒæ–¹æ³•ã€‚

- æˆ‘ä½¿ç”¨çš„é›†ç¾¤æ˜¯slurmä½œä¸šç®¡ç†ç³»ç»Ÿï¼Œä¸‹é¢çš„ç¤ºä¾‹è„šæœ¬ä¹Ÿæ˜¯é€‚ç”¨ä¸slurmçš„ï¼Œå°½é‡å…ˆå­¦ä¹ ä¸€ä¸‹[slurmçš„ä½¿ç”¨](../slurm)å†å°è¯•æäº¤ä½œä¸šã€‚å¦‚æœä¸ç”¨slurmçš„è¯å¯ä»¥åªå‚è€ƒ`#############`ä¸­é—´çš„è½¯ä»¶éƒ¨åˆ†ã€‚

## Introduction

å®åŸºå› ç»„ï¼ˆMetagenomeï¼‰æ˜¯æŒ‡å¯¹ä¸€ä¸ªç”Ÿæ€ç³»ç»Ÿä¸­çš„æ‰€æœ‰å¾®ç”Ÿç‰©è¿›è¡ŒDNAåˆ†æçš„è¿‡ç¨‹ï¼Œå¯ä»¥å¸®åŠ©ç ”ç©¶äººå‘˜äº†è§£å¾®ç”Ÿç‰©çš„å¤šæ ·æ€§ã€åŠŸèƒ½å’Œäº’ä½œå…³ç³»ã€‚

å®åŸºå› ç»„çš„åº”ç”¨éå¸¸å¹¿æ³›ï¼ŒåŒ…æ‹¬ï¼š

- ç”Ÿç‰©å¤šæ ·æ€§ç ”ç©¶ï¼šé€šè¿‡å¯¹å®åŸºå› ç»„è¿›è¡Œåˆ†æï¼Œå¯ä»¥äº†è§£ä¸åŒç”Ÿæ€ç³»ç»Ÿä¸­å¾®ç”Ÿç‰©çš„å¤šæ ·æ€§å’Œåˆ†å¸ƒæƒ…å†µã€‚
- ç”Ÿæ€å­¦ç ”ç©¶ï¼šå®åŸºå› ç»„å¯ä»¥å¸®åŠ©ç ”ç©¶äººå‘˜äº†è§£å¾®ç”Ÿç‰©åœ¨ç”Ÿæ€ç³»ç»Ÿä¸­çš„åŠŸèƒ½ã€äº’ä½œå…³ç³»å’Œç”Ÿæ€ä½ç­‰ã€‚
- ç”Ÿç‰©æŠ€æœ¯ï¼šå®åŸºå› ç»„å¯ä»¥ç”¨äºç­›é€‰å…·æœ‰ç‰¹å®šåŠŸèƒ½çš„å¾®ç”Ÿç‰©ï¼Œä¾‹å¦‚ï¼Œå¯»æ‰¾èƒ½å¤Ÿé™è§£æœ‰å®³ç‰©è´¨çš„å¾®ç”Ÿç‰©ã€‚

å®åŸºå› ç»„çš„åˆ†æä¸€èˆ¬åŒ…æ‹¬ä»¥ä¸‹æ­¥éª¤ï¼š

1.  DNAæå–ä¸å»ºåº“ã€‚
2.  é«˜é€šé‡æµ‹åºï¼šä½¿ç”¨é«˜é€šé‡æµ‹åºæŠ€æœ¯å¯¹æ‰©å¢åçš„DNAè¿›è¡Œæµ‹åºï¼Œå¾—åˆ°åŸå§‹åºåˆ—æ•°æ®ã€‚
3.  æ•°æ®æ¸…æ´—å’Œç»„è£…ï¼šå¯¹åŸå§‹æ•°æ®è¿›è¡Œè´¨é‡æ§åˆ¶ã€å»é™¤ä½è´¨é‡åºåˆ—å’Œå†—ä½™åºåˆ—ï¼Œå°†åºåˆ—æ‹¼æ¥æˆè¾ƒé•¿çš„è¿ç»­åºåˆ—ï¼ˆcontigsï¼‰ã€‚
4.  åŸºå› æ³¨é‡Šï¼šå°†contigsä¸­çš„åŸºå› è¿›è¡Œæ³¨é‡Šï¼Œå¾—åˆ°åŸºå› åŠŸèƒ½ä¿¡æ¯ã€‚
5.  æ•°æ®åˆ†æï¼šäº†è§£å¾®ç”Ÿç‰©å¤šæ ·æ€§ã€ç¾¤è½ç»“æ„ã€åŠŸèƒ½ç‰¹å¾ç­‰ä¿¡æ¯ï¼ˆæ›´å¤šæ˜¯æŒ‡è·å–äº†ç‰©ç§ä¸°åº¦è¡¨æˆ–åŠŸèƒ½ä¸°åº¦è¡¨ä¹‹åçš„è¿›ä¸€æ­¥åˆ†æï¼‰ã€‚
6.  MAGs binningï¼Œ è¿›åŒ–åŠ¨æ€ç­‰è¿›ä¸€æ­¥åˆ†æ

ä¸‹å›¾æ˜¯æˆ‘å¸¸ç”¨çš„ä¸€å¥—ä¸Šæ¸¸åŸºæœ¬æµç¨‹ï¼Œå½“ç„¶åœ¨é¢å¯¹ä¸åŒé¡¹ç›®æ—¶åº”è¯¥æœ‰ä¸åŒçš„ä¾§é‡ç‚¹å’Œé€‚ç”¨çš„åˆ†ææ–¹æ³•ï¼Œå¯ä»¥åœ¨æ­¤åŸºç¡€ä¸Šæ·»åŠ æˆ–ä¿®æ”¹ã€‚

æœ€æ—©çš„æ—¶å€™ï¼Œè¿™æ–¹é¢çš„åˆ†ææˆ‘éƒ½æ˜¯å‚è€ƒåˆ˜æ°¸é‘«è€å¸ˆçš„[**EasyMetagenome**](https://github.com/YongxinLiu/EasyMetagenome)ï¼Œç°åœ¨è¿™å¥—æµç¨‹ä¹Ÿå‘æ–‡ç« äº† ([*1*](#ref-liuPracticalGuideAmplicon2021))ï¼Œå€¼å¾—å‚è€ƒï¼Œå¯¹ä¸Šæ‰‹16Sæµ‹åºæ•°æ®æˆ–å®åŸºå› ç»„æ•°æ®éƒ½å¾ˆæœ‰å¸®åŠ©ã€‚

<img src="images/workflow.png" title=""/>

æœ€è¿‘ï¼ˆ2024.3.3ï¼‰æˆ‘æ•´ç†äº†ä¸€ä¸‹æµç¨‹æ”¾åœ¨æˆ‘çš„RåŒ…`pctax`é‡Œäº†ï¼Œæ–¹ä¾¿æ•´ä¸ªå®åŸºå› ç»„çš„ä¸Šä¸‹æ¸¸æµç¨‹ç»“åˆï¼š

<img src="images/pctax_workflow.png" title=""/>

``` r
install.packages("pctax")
#ä½¿ç”¨micro_sbatchå¯ä»¥å¿«é€Ÿè·å¾—æ¯ä¸€æ­¥çš„slurmè„šæœ¬ï¼š
pctax::micro_sbatch(work_dir = "~/work/",step = "fastp")
```

## preprocess

ä¸€èˆ¬æŠŠæ‰€æœ‰æ ·æœ¬çš„æµ‹åºåŒç«¯æ–‡ä»¶æ”¾åœ¨ä¸€ä¸ªdataæ–‡ä»¶å¤¹ä¸‹ï¼Œç„¶åæŠŠæ‰€æœ‰æ ·æœ¬åå†™å…¥åˆ°ä¸€ä¸ª`samplelist`æ–‡ä»¶ä¸­ï¼Œæ–¹ä¾¿åç»­å„ç§è½¯ä»¶çš„æ‰¹é‡æ“ä½œã€‚

### è´¨æ§ï¼šfastp

æµ‹åºæ–‡ä»¶è´¨æ§æ˜¯æŒ‡åœ¨ä¸‹æ¸¸åˆ†æå‰å¯¹æµ‹åºæ•°æ®è¿›è¡Œè´¨é‡æ§åˆ¶å’Œæ¸…ç†ï¼Œä»¥ç¡®ä¿æ•°æ®çš„å‡†ç¡®æ€§å’Œå¯é æ€§ã€‚

fastpæ˜¯ä¸€ä¸ªéå¸¸å¿«é€Ÿã€å¤šåŠŸèƒ½çš„æµ‹åºæ•°æ®è´¨æ§å·¥å…·ï¼Œæ”¯æŒè‡ªåŠ¨åŒ–çš„è´¨æ§å¤„ç†ï¼ŒåŒ…æ‹¬å»é™¤ä½è´¨é‡åºåˆ—ã€å‰ªåˆ‡æ¥å¤´åºåˆ—å’Œç”Ÿæˆè´¨æ§æŠ¥å‘Šã€‚

``` bash
#!/bin/bash
#SBATCH --job-name=fastp
#SBATCH --output=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.out
#SBATCH --error=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.err
#SBATCH --array=1
#SBATCH --partition=cpu
#SBATCH --cpus-per-task=8
#SBATCH --mem-per-cpu=2G
samplelist=/share/home/jianglab/pengchen/work/asthma/samplelist

echo start: `date +"%Y-%m-%d %T"`
start=`date +%s`

####################
echo "SLURM_ARRAY_TASK_ID: " $SLURM_ARRAY_TASK_ID
START=$SLURM_ARRAY_TASK_ID

NUMLINES=40 #how many sample in one array

STOP=$((SLURM_ARRAY_TASK_ID*NUMLINES))
START="$(($STOP - $(($NUMLINES - 1))))"

echo "START=$START"
echo "STOP=$STOP"

####################
for (( N = $START; N <= $STOP; N++ ))
do
  sample=$(head -n "$N" $samplelist | tail -n 1)
  echo $sample
  mkdir -p data/fastp
  ~/miniconda3/envs/waste/bin/fastp -w 8 -i data/rawdata/${sample}_f1.fastq.gz -o data/fastp/${sample}_1.fq.gz \
    -I data/rawdata/${sample}_r2.fastq.gz -O data/fastp/${sample}_2.fq.gz -j data/fastp/${i}.json
done

##############
echo end: `date +"%Y-%m-%d %T"`
end=`date +%s`
echo TIME:`expr $end - $start`s
```

fastpçš„jsonæ–‡ä»¶ä¸­ç»Ÿè®¡äº†æµ‹åºæ•°æ®çš„åŸºæœ¬æŒ‡æ ‡ï¼Œæˆ‘ä»¬å¯ä»¥å°†å…¶æ•´ç†ä¸ºè¡¨æ ¼ï¼š
æŠŠæ‰€æœ‰çš„.jsonæ–‡ä»¶ç§»åˆ°ä¸€ä¸ªæ–‡ä»¶å¤¹é‡Œreport/ä¸‹ï¼Œä½¿ç”¨æˆ‘å†™çš„RåŒ…`pctax`è¯»å–jsonæ–‡ä»¶å¹¶æ•´ç†æˆè¡¨æ ¼ï¼š

``` r
pctax::pre_fastp("report/")
```

å¦å¤–ï¼ŒFastQCï¼ŒCutadaptï¼ŒTrimmomaticç­‰ä¹Ÿæ˜¯å¸¸ç”¨çš„æµ‹åºè´¨æ§å·¥å…·ã€‚

### å»å®¿ä¸»ï¼šbowtie2

å»å®¿ä¸»çš„è¿‡ç¨‹å…¶å®å°±æ˜¯å°†åºåˆ—æ¯”å¯¹åˆ°å®¿ä¸»åŸºå› ç»„ä¸Šï¼Œç„¶åæ²¡æœ‰æ¯”å¯¹åˆ°çš„åºåˆ—æ•´åˆæˆæ–°æ–‡ä»¶å°±æ˜¯å»å®¿ä¸»åçš„äº†ã€‚å®¿ä¸»åŸºå› ç»„éœ€è¦è‡ªå·±å…ˆä¸‹è½½å¥½å¹¶ç”¨ bowtie2-build å»ºç«‹ç´¢å¼•ï¼Œä»¥äººç±»ä¸ºä¾‹ï¼š

``` bash
#!/bin/bash
#SBATCH --job-name=rm_human
#SBATCH --output=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.out
#SBATCH --error=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.err
#SBATCH --array=1
#SBATCH --partition=cpu
#SBATCH --cpus-per-task=8
#SBATCH --mem-per-cpu=2G
samplelist=/share/home/jianglab/pengchen/work/asthma/samplelist

echo start: `date +"%Y-%m-%d %T"`
start=`date +%s`

####################
echo "SLURM_ARRAY_TASK_ID: " $SLURM_ARRAY_TASK_ID
START=$SLURM_ARRAY_TASK_ID

NUMLINES=40 #how many sample in one array

STOP=$((SLURM_ARRAY_TASK_ID*NUMLINES))
START="$(($STOP - $(($NUMLINES - 1))))"

echo "START=$START"
echo "STOP=$STOP"

####################
for (( N = $START; N <= $STOP; N++ ))
do
  sample=$(head -n "$N" $samplelist | tail -n 1)
  echo $sample
  mkdir -p data/rm_human
  ~/miniconda3/envs/waste/bin/bowtie2 -p 8 -x ~/db/humangenome/hg38 -1 data/fastp/${sample}_1.fq.gz \
    -2 data/fastp/${sample}_2.fq.gz -S data/rm_human/${sample}.sam \
    --un-conc data/rm_human/${sample}.fq --very-sensitive
  rm data/rm_human/${sample}.sam
done

##############
echo end: `date +"%Y-%m-%d %T"`
end=`date +%s`
echo TIME:`expr $end - $start`s
```

å¦å¤–ä¹Ÿå¯ä»¥ä½¿ç”¨bwaï¼Œkneaddataç­‰è½¯ä»¶å»å®¿ä¸»ã€‚

### fastqä¿¡æ¯ç»Ÿè®¡

å¯ä»¥ç”¨ä¸€äº›å°å·¥å…·å¦‚FastqCountï¼Œseqkitç­‰æ¥ç»Ÿè®¡readsæ•°ï¼Œç¢±åŸºæ•°ï¼ŒGCï¼ŒQ20ï¼ŒQ30ç­‰æŒ‡æ ‡ï¼š

``` bash
~/biosoft/FastqCount-master/FastqCount_v0.5 xx.fastq.gz

Total Reads     Total Bases     N Bases Q20     Q30     GC
11568822 (11.57 M)      1702829127 (1.70 G)     0.00%   98.00%  94.00%  54.00%
```

è¿˜å¯ä»¥æŠŠcleandataå†è·‘ä¸€ä¸‹fastpğŸ˜‚ï¼Œä½†æ˜¯åªçœ‹å¤„ç†å‰çš„ç»Ÿè®¡æŒ‡æ ‡ã€‚å› ä¸ºfastpç¡®å®éå¸¸å¿«ï¼Œçº¯ç²¹ç”¨æ¥ç»Ÿè®¡ä¹Ÿè¡Œã€‚

## reads-based

reads-basedå®åŸºå› ç»„åˆ†ææ˜¯æŒ‡ç›´æ¥åˆ©ç”¨æµ‹åºè¯»é•¿ï¼ˆreadsï¼‰è¿›è¡Œå®åŸºå› ç»„æ•°æ®åˆ†æçš„æ–¹æ³•ï¼Œè€Œä¸æ˜¯å…ˆè¿›è¡ŒåŸºå› ç»„ç»„è£…ã€‚è¯¥æ–¹æ³•é€šè¿‡å¯¹çŸ­è¯»é•¿è¿›è¡Œæ¯”å¯¹å’Œæ³¨é‡Šï¼Œä»ä¸­æå–åŠŸèƒ½ä¿¡æ¯å’Œç‰©ç§ç»„æˆï¼Œå¸¸ç”¨äºå¿«é€Ÿè¯„ä¼°ç¯å¢ƒæ ·æœ¬ä¸­çš„å¾®ç”Ÿç‰©ç¾¤è½ç»“æ„å’ŒåŠŸèƒ½æ½œåŠ›ã€‚

- **ä¼˜ç‚¹**ï¼š
  - **å¿«é€Ÿ**ï¼šæ— éœ€ç»„è£…ï¼Œå¤„ç†é€Ÿåº¦è¾ƒå¿«ã€‚
  - **å…¨é¢**ï¼šèƒ½å¤Ÿæ£€æµ‹åˆ°ä½ä¸°åº¦çš„å¾®ç”Ÿç‰©å’ŒåŠŸèƒ½åŸºå› ï¼Œé€‚åˆå¤§è§„æ¨¡æ ·æœ¬åˆ†æã€‚
- **å±€é™æ€§**ï¼š
  - **ç‰‡æ®µåŒ–**ï¼šç”±äºä¸è¿›è¡Œç»„è£…ï¼Œåˆ†æåŸºäºçŸ­è¯»é•¿ï¼Œå¯èƒ½ä¼šé”™è¿‡é•¿è·ç¦»çš„åŸºå› è”ç³»å’Œç»“æ„ä¿¡æ¯ã€‚
  - **ä¾èµ–æ•°æ®åº“**ï¼šåˆ†æç»“æœä¾èµ–äºå‚è€ƒæ•°æ®åº“çš„å…¨é¢æ€§å’Œå‡†ç¡®æ€§ã€‚

### ç‰©ç§æ³¨é‡Šï¼škraken2

Kraken2æ˜¯ä¸€ä¸ªç”¨äºå¯¹é«˜é€šé‡æµ‹åºæ•°æ®è¿›è¡Œåˆ†ç±»å’Œæ ‡è¯†ç‰©ç§çš„è½¯ä»¶ã€‚å®ƒä½¿ç”¨å‚è€ƒæ•°æ®åº“ä¸­çš„åŸºå› ç»„åºåˆ—æ¥è¿›è¡Œåˆ†ç±»ï¼Œå¹¶ä½¿ç”¨k-meræ–¹æ³•æ¥å®ç°å¿«é€Ÿå’Œå‡†ç¡®çš„åˆ†ç±»ã€‚

ä½¿ç”¨Kraken2è¿›è¡ŒåŸºæœ¬åˆ†ç±»çš„ç®€å•æ­¥éª¤ï¼š

1.  å‡†å¤‡å‚è€ƒæ•°æ®åº“ï¼šKraken2éœ€è¦ä¸€ä¸ªå‚è€ƒæ•°æ®åº“ï¼Œä»¥ä¾¿å¯¹æµ‹åºæ•°æ®è¿›è¡Œåˆ†ç±»ã€‚å¯ä»¥ç›´æ¥ä¸‹è½½å®˜æ–¹æ„å»ºçš„æ ‡å‡†åº“ï¼Œä¹Ÿå¯ä»¥ä»NCBIã€Ensemblæˆ–å…¶ä»–æ•°æ®åº“ä¸‹è½½ç›¸åº”çš„åŸºå› ç»„åºåˆ—ï¼Œå¹¶ä½¿ç”¨Kraken2å†…ç½®çš„å·¥å…·æ¥æ„å»ºæ•°æ®åº“ã€‚

`kraken2-build --standard --threads 24 --db ./`

`--standard`æ ‡å‡†æ¨¡å¼ä¸‹åªä¸‹è½½5ç§æ•°æ®åº“ï¼šå¤èŒarchaeaã€ç»†èŒbacteriaã€äººç±»humanã€è½½ä½“UniVec_Coreã€ç—…æ¯’viralã€‚

2.  å®‰è£…Kraken2ï¼šå¯ä»¥ä»Kraken2å®˜æ–¹ç½‘ç«™ä¸‹è½½å¹¶å®‰è£…Kraken2è½¯ä»¶ã€‚

3.  è¿è¡ŒKraken2ï¼šä½¿ç”¨Kraken2å¯¹æµ‹åºæ•°æ®è¿›è¡Œåˆ†ç±»éœ€è¦ä½¿ç”¨ä»¥ä¸‹å‘½ä»¤ï¼š

`kraken2 --db <path_to_database> <input_file> --output <output_file>`

è¿™é‡Œï¼Œ`<path_to_database>`æ˜¯å‚è€ƒæ•°æ®åº“çš„è·¯å¾„ï¼Œ`<input_file>`æ˜¯éœ€è¦è¿›è¡Œåˆ†ç±»çš„è¾“å…¥æ–‡ä»¶ï¼Œ`<output_file>`æ˜¯è¾“å‡ºæ–‡ä»¶çš„åç§°ã€‚Kraken2å°†è¾“å‡ºä¸€ä¸ªåˆ†ç±»æŠ¥å‘Šæ–‡ä»¶å’Œä¸€ä¸ªåºåˆ—æ–‡ä»¶ã€‚

éœ€è¦æ³¨æ„çš„æ˜¯krakenè¿è¡Œè‡³å°‘è¦æä¾›æ•°æ®åº“å¤§å°çš„å†…å­˜å¤§å°ï¼ˆè¿è¡Œå†…å­˜ï¼‰ï¼Œå› ä¸ºå®ƒä¼šæŠŠæ•´ä¸ªæ•°æ®åº“è½½å…¥å†…å­˜åè¿›è¡Œåºåˆ—çš„æ³¨é‡Šï¼Œæ‰€ä»¥å¦‚æœå‘ç°æ— æ³•è½½å…¥æ•°æ®åº“çš„æŠ¥é”™ï¼Œå¯ä»¥å°è¯•è°ƒå¤§å†…å­˜èµ„æºã€‚

``` bash
#!/bin/bash
#SBATCH --job-name=kraken2
#SBATCH --output=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.out
#SBATCH --error=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.err
#SBATCH --partition=cpu
#SBATCH --cpus-per-task=8
#SBATCH --mem-per-cpu=40G
samplelist=/share/home/jianglab/pengchen/work/asthma/samplelist

echo start: `date +"%Y-%m-%d %T"`
start=`date +%s`

####################
echo "SLURM_ARRAY_TASK_ID: " $SLURM_ARRAY_TASK_ID
START=$SLURM_ARRAY_TASK_ID

NUMLINES=40 #how many sample in one array

STOP=$((SLURM_ARRAY_TASK_ID*NUMLINES))
START="$(($STOP - $(($NUMLINES - 1))))"

####################
for (( N = $START; N <= $STOP; N++ ))
do
  sample=$(head -n "$N" $samplelist | tail -n 1)
  echo $sample
  mkdir -p result/kraken2
  ~/miniconda3/envs/waste/bin/kraken2 --threads 8 \
    --db ~/db/kraken2/stand_krakenDB \
    --confidence 0.05 \
    --output result/kraken2/${sample}.output \
    --report result/kraken2/${sample}.kreport \
    --paired \
    data/rm_human/${sample}.1.fq \
    data/rm_human/${sample}.2.fq
done

##############
echo end: `date +"%Y-%m-%d %T"`
end=`date +%s`
echo TIME:`expr $end - $start`s
```

å¦å¤–ï¼Œkrakenæ•°æ®åº“æ˜¯å¯ä»¥è‡ªå·±æ„å»ºçš„ï¼Œæ‰€ä»¥é€‚ç”¨äºå„ç§é¡¹ç›®çš„ç‰©ç§æ³¨é‡Šï¼Œæˆ‘åšçš„æ¯”è¾ƒå¤šçš„æ˜¯ç¯å¢ƒæ ·æœ¬çš„å®åŸºå› ç»„ï¼Œå°±å¯èƒ½éœ€è¦æ›´å…¨é¢çš„ç‰©ç§æ•°æ®åº“ï¼ˆç”šè‡³é™¤äº†å„ç§å¾®ç”Ÿç‰©ï¼Œè¿˜è¦åŠ¨æ¤ç‰©æ•°æ®ç­‰ï¼‰ï¼Œæˆ‘ä»¬å®éªŒå®¤çš„WXå¸ˆå§æ”¶é›†æ„å»ºäº†ä¸€ä¸ªè¶…å¤§çš„ç‰©ç§åº“ã€‚

krakenè½¯ä»¶è¿è¡Œæ—¶è½½å…¥æ•°æ®åº“æ˜¯ä¸€ä¸ªååˆ†è€—æ—¶çš„æ­¥éª¤ï¼Œè€Œæ¯æ¡åºåˆ—çš„é‰´å®šæ—¶é—´å·®ä¸å¤šï¼Œæ‰€ä»¥æˆ‘ä»¬å¯ä»¥å°†å¾ˆå¤šæ ·æœ¬çš„fastqæ–‡ä»¶åˆå¹¶æˆä¸€ä¸ªå¤§æ–‡ä»¶åè¾“å…¥krakenæ³¨é‡Šï¼Œä¹‹åå†æŒ‰ç…§åºåˆ—çš„æ•°é‡æ‹†åˆ†ç»“æœæ–‡ä»¶ï¼Œè¿™æ ·å¤šä¸ªæ ·æœ¬ä¹Ÿåªéœ€è¦è½½å…¥ä¸€æ¬¡æ•°æ®åº“ï¼ŒèŠ‚çœæ—¶é—´ï¼Œä»¥ä¸‹æ˜¯ç”¨è‡ªå®šä¹‰çš„kraken2M.pyè„šæœ¬å®ç°ã€‚

``` bash
#!/bin/bash
#SBATCH --job-name=kraken2M
#SBATCH --output=/share/home/jianglab/pengchen/work/asthma/kraken/%x_%a.out
#SBATCH --error=/share/home/jianglab/pengchen/work/asthma/kraken/%x_%a.err
#SBATCH --time=14-00:00:00
#SBATCH --partition=mem
#SBATCH --cpus-per-task=32
#SBATCH --mem-per-cpu=100G
echo start: `date +"%Y-%m-%d %T"`
start=`date +%s`

####################

  mkdir -p result/kraken
  python /share/home/jianglab/shared/krakenDB/K2ols/kraken2M.py -t 16 \
      -i data/rm_human \
      -c 0.05 \
      -s .1.fq,.2.fq \
      -o result/kraken \
      -d /share/home/jianglab/shared/krakenDB/mydb2 \
      -k ~/miniconda3/envs/waste/bin/kraken2 \
      -kt /share/home/jianglab/shared/krakenDB/K2ols/KrakenTools
  mkdir -p result/kraken/kreport
  mv result/kraken/*_kreport.txt result/kraken/kreport/
  python ~/db/script/format_kreports.py -i result/kraken/kreport \
      -kt /share/home/jianglab/shared/krakenDB/K2ols/KrakenTools --save-name2taxid

##############
echo end: `date +"%Y-%m-%d %T"`
end=`date +%s`
echo TIME:`expr $end - $start`s
```

#### Krakenè¾“å‡ºæ ¼å¼

æ ‡å‡†è¾“å‡ºæ ¼å¼outputæ–‡ä»¶(äº”åˆ—è¡¨):

- C/Uä»£è¡¨åˆ†ç±»classifiedæˆ–éåˆ†ç±»unclassifed
- åºåˆ—ID
- ç‰©ç§æ³¨é‡Š
- æ¯”åºåˆ—æ³¨é‡Šçš„åŒºåŸŸï¼Œå¦‚98\|94ä»£è¡¨å·¦ç«¯98bpï¼Œå³ç«¯94bpæ¯”å¯¹è‡³æ•°æ®åº“
- LCAæ¯”å¯¹ç»“æœï¼Œå¦‚â€562:13 561:4â€ä»£è¡¨13 k-meræ¯”å¯¹è‡³ç‰©ç§#562ï¼Œ4 k-meræ¯”å¯¹è‡³#561ç‰©ç§

æŠ¥å‘Šè¾“å‡ºæ ¼å¼reportæ–‡ä»¶(åŒ…æ‹¬6åˆ—ï¼Œæ–¹ä¾¿æ•´ç†ä¸‹æ¸¸åˆ†æ):

1.  ç™¾åˆ†æ¯”
2.  count
3.  countæœ€ä¼˜
4.  (U)nclassified, (R)oot, (D)omain, (K)ingdom, (P)hylum, (C)lass, (O)rder, (F)amily, (G)enus, or (S)pecies. â€œG2â€ä»£è¡¨ä½äºå±ä¸€ç§é—´
5.  NCBIç‰©ç§ID
6.  ç§‘å­¦ç‰©ç§å

å¸¸ç”¨çš„ç‰©ç§ä¸°åº¦è¡¨æ ¼å¼é™¤äº†kraken reportï¼Œè¿˜æœ‰mpaï¼Œspfï¼Œkronaç­‰æ ¼å¼ï¼Œå…³äºkrakenç»“æœçš„æ•´ç†ä»¥åŠæ ¼å¼è½¬æ¢æ–¹å¼ï¼Œæœ‰ä¸€äº›ç°æˆçš„è„šæœ¬æˆ–è€…è‡ªå·±å†™ã€‚

[KrakenTools (jhu.edu)](https://ccb.jhu.edu/software/krakentools/) å°±æ˜¯ä¸€å¥—å¾ˆå¥½ç”¨çš„krakenå·¥å…·åŒ…ï¼Œå…¶ä¸­å¸¸ç”¨çš„æœ‰ï¼š

1.  extract_kraken_reads.py
    æ­¤ç¨‹åºæå–è¯»å–åœ¨ä»»ä½•ç”¨æˆ·æŒ‡å®šçš„åˆ†ç±»idå¤„åˆ†ç±»çš„å†…å®¹ã€‚ç”¨æˆ·å¿…é¡»æŒ‡å®šKrakenè¾“å‡ºæ–‡ä»¶ã€åºåˆ—æ–‡ä»¶å’Œè‡³å°‘ä¸€ä¸ªåˆ†ç±»æ³•IDã€‚ä¸‹é¢æŒ‡å®šäº†å…¶ä»–é€‰é¡¹ã€‚æˆªè‡³2021å¹´4æœˆ19æ—¥ï¼Œæ­¤è„šæœ¬ä¸KrakenUniq/Kraken2UniqæŠ¥å‘Šå…¼å®¹ã€‚

2.  combine_kreports.py
    è¿™ä¸ªè„šæœ¬å°†å¤šä¸ªKrakenæŠ¥å‘Šåˆå¹¶åˆ°ä¸€ä¸ªåˆå¹¶çš„æŠ¥å‘Šæ–‡ä»¶ä¸­ã€‚

3.  kreport2krona.py
    è¿™ä¸ªç¨‹åºéœ€è¦ä¸€ä¸ªKrakenæŠ¥å‘Šæ–‡ä»¶ï¼Œå¹¶æ‰“å°å‡ºä¸€ä¸ªkronaå…¼å®¹çš„æ–‡æœ¬æ–‡ä»¶ï¼Œæ¢æˆkronaæ–‡ä»¶å¥½ç”»å›¾ã€‚

è½¬æ¢åçš„kronaä½¿ç”¨ `ktImportText MYSAMPLE.krona -o MYSAMPLE.krona.html`å¾—åˆ°ç½‘é¡µå¯è§†åŒ–ç»“æœã€‚

4.  kreport2mpa.py
    è¿™ä¸ªç¨‹åºéœ€è¦ä¸€ä¸ªKrakenæŠ¥å‘Šæ–‡ä»¶ï¼Œå¹¶æ‰“å°å‡ºä¸€ä¸ªmpa (MetaPhlAn)é£æ ¼çš„æ–‡æœ¬æ–‡ä»¶ã€‚

5.  combine_mpa.py
    è¿™ä¸ªç¨‹åºåˆå¹¶æ¥è‡ªkreport2mpa.pyçš„å¤šä¸ªè¾“å‡ºã€‚

### ç‰©ç§+åŠŸèƒ½ï¼šHUMAnN

HUMAnNï¼ˆThe HMP Unified Metabolic Analysis Networkï¼‰æ˜¯ä¸€æ¬¾ç”¨äºåˆ†æäººç±»å¾®ç”Ÿç‰©ç»„çš„åŠŸèƒ½å’Œä»£è°¢èƒ½åŠ›çš„å·¥å…·ã€‚

å®ƒé€šè¿‡å°†å®åŸºå› ç»„åºåˆ—ä¸å‚è€ƒåŸºå› ç»„æ•°æ®åº“æ¯”å¯¹ï¼Œåˆ©ç”¨MetaCycä»£è°¢é€šè·¯æ•°æ®åº“å’ŒUniRefè›‹ç™½è´¨åºåˆ—æ•°æ®åº“ï¼Œåˆ†æå¾®ç”Ÿç‰©ç»„åœ¨åŠŸèƒ½å’Œä»£è°¢é€šè·¯æ°´å¹³ä¸Šçš„ç»„æˆå’Œæ´»æ€§ã€‚HUMAnN2è¿˜æä¾›äº†å¤šæ ·æ€§åˆ†æã€å…³è”åˆ†æå’Œå¯è§†åŒ–å·¥å…·ï¼Œå¯ç”¨äºæ·±å…¥ç ”ç©¶äººç±»å¾®ç”Ÿç‰©ç»„å¯¹å®¿ä¸»å¥åº·çš„å½±å“å’Œæ²»ç–—ç­–ç•¥çš„åˆ¶å®šç­‰æ–¹é¢ã€‚

HUMAnNæ˜¯ç”±ç¾å›½å›½å®¶äººç±»å¾®ç”Ÿç‰©ç»„è®¡åˆ’ï¼ˆHMPï¼‰å¼€å‘çš„ï¼Œç›®å‰æœ€æ–°ç‰ˆæœ¬ä¸º[HUMAnN3](https://github.com/biobakery/humann)ï¼Œäº2020å¹´å‘å¸ƒã€‚ä¸HUMAnN2ç›¸æ¯”ï¼ŒHUMAnN3æ”¹è¿›äº†åŸºå› å®¶æ—æ³¨é‡Šçš„æ–¹æ³•ï¼Œæé«˜äº†æ³¨é‡Šç²¾åº¦å’Œé€Ÿåº¦ï¼Œå¹¶æä¾›äº†æ–°çš„åŠŸèƒ½å’Œå·¥å…·ï¼Œå¦‚åŠŸèƒ½éŸ§åº¦åˆ†æã€ä»£è°¢æŒ‡çº¹è¯†åˆ«å’Œå¤šæ ·æ€§åˆ†æç­‰ã€‚

ä½†æ˜¯HUMAnNçš„æ•°æ®åº“åŸºæœ¬éƒ½æ˜¯ä¸äººç›¸å…³çš„å¾®ç”Ÿç‰©ï¼Œæ¯”è¾ƒé€‚åˆåšå„ç§äººä½“å¾®ç”Ÿç‰©ç»„ï¼ˆè‚ é“ï¼Œè‚ºéƒ¨ï¼Œå£è…”ï¼Œçš®è‚¤ç­‰ç­‰ï¼‰ï¼Œå¯¹äºç¯å¢ƒæ ·æœ¬å¯èƒ½unclassifiedæ¯”è¾ƒå¤šã€‚

**HUMAnNè¦æ±‚åŒç«¯åºåˆ—åˆå¹¶çš„æ–‡ä»¶ä½œä¸ºè¾“å…¥**ï¼Œforå¾ªç¯æ ¹æ®å®éªŒè®¾è®¡æ ·æœ¬åæ‰¹é‡åŒç«¯åºåˆ—åˆå¹¶ã€‚

- **ç‰©ç§ç»„æˆè°ƒç”¨MetaPhlAn3, bowtie2æ¯”å¯¹è‡³æ ¸é…¸åºåˆ—åº“**ï¼Œè§£å†³æœ‰å“ªäº›å¾®ç”Ÿç‰©å­˜åœ¨çš„é—®é¢˜ï¼›
- **åŠŸèƒ½ç»„æˆä¸ºhumann3è°ƒç”¨diamondæ¯”å¯¹è‡³è›‹ç™½åº“**ï¼Œè§£å†³è¿™äº›å¾®ç”Ÿç‰©å‚ä¸å“ªäº›åŠŸèƒ½é€šè·¯çš„é—®é¢˜ï¼›

``` bash
cd data/rm_human
for i in `cat ~/work/asthma/samplelist`
do
echo $i
cat ${i}_f1.fastq ${i}_r2.fastq >${i}_paired.fastq
done
```

``` bash
#!/bin/bash
#SBATCH --job-name=humann
#SBATCH --output=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.out
#SBATCH --error=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.err
#SBATCH --array=1
#SBATCH --partition=cpu
#SBATCH --nodes=1
#SBATCH --tasks-per-node=1
#SBATCH --cpus-per-task=1
#SBATCH --mem-per-cpu=2G
samplelist=/share/home/jianglab/pengchen/work/asthma/samplelist

echo start: `date +"%Y-%m-%d %T"`
start=`date +%s`

####################
echo "SLURM_ARRAY_TASK_ID: " $SLURM_ARRAY_TASK_ID
START=$SLURM_ARRAY_TASK_ID

NUMLINES=40 #how many sample in one array

STOP=$((SLURM_ARRAY_TASK_ID*NUMLINES))
START="$(($STOP - $(($NUMLINES - 1))))"

#set the min and max
if [ $START -lt 1 ]
then
  START=1
fi
if [ $STOP -gt 40 ]
then
  STOP=40
fi

echo "START=$START"
echo "STOP=$STOP"

####################
for (( N = $START; N <= $STOP; N++ ))
do
  sample=$(head -n "$N" $samplelist | tail -n 1)
  echo $sample
  mkdir -p data/paired
  cat data/rm_human/${sample}.1.fq data/rm_human/${sample}.2.fq > data/paired/${sample}_paired.fq
  
  mkdir -p result/humann
  ~/miniconda3/envs/waste/bin/humann --input data/paired/${sample}_paired.fq \
    --output result/humann/ --threads 24
    
  ln result/humann/${sample}_paired_humann_temp/${sample}_paired_metaphlan_bugs_list.tsv result/humann/
  rm -rf result/humann/${sample}_paired_humann_temp
done

##############
echo end: `date +"%Y-%m-%d %T"`
end=`date +%s`
echo TIME:`expr $end - $start`s
```

``` bash
mkdir -p result/metaphlan2
## åˆå¹¶ã€ä¿®æ­£æ ·æœ¬å
merge_metaphlan_tables2.py \
  result/humann/*_metaphlan_bugs_list.tsv | \
  sed 's/_metaphlan_bugs_list//g' \
  > result/metaphlan2/taxonomy.tsv
```

## contigs-based

contigs-basedå®åŸºå› ç»„åˆ†ææ˜¯æŒ‡å…ˆå°†æµ‹åºè¯»é•¿ï¼ˆreadsï¼‰ç»„è£…æˆè¾ƒé•¿çš„åºåˆ—ç‰‡æ®µï¼ˆcontigsï¼‰ï¼Œç„¶åå¯¹è¿™äº›contigsè¿›è¡Œè¿›ä¸€æ­¥çš„åˆ†æã€‚è¿™ç§æ–¹æ³•å¯ä»¥æä¾›æ›´é•¿çš„åŸºå› ç»„ç‰‡æ®µï¼Œæœ‰åŠ©äºæ›´å‡†ç¡®åœ°è¿›è¡ŒåŸºå› æ³¨é‡Šå’Œå¾®ç”Ÿç‰©ç¾¤è½ç»“æ„åˆ†æã€‚

- **ä¼˜ç‚¹**ï¼š
  - **æ›´é«˜çš„è§£æåº¦**ï¼šé€šè¿‡ç»„è£…ç”Ÿæˆè¾ƒé•¿çš„åºåˆ—ç‰‡æ®µï¼Œå¯ä»¥æ›´å‡†ç¡®åœ°è¿›è¡ŒåŸºå› å’ŒåŸºå› ç»„æ³¨é‡Šã€‚
  - **ç»“æ„ä¿¡æ¯**ï¼šèƒ½å¤Ÿè·å¾—é•¿è·ç¦»çš„åŸºå› è”ç³»å’ŒåŸºå› ç»„ç»“æ„ä¿¡æ¯ï¼Œæœ‰åŠ©äºæ›´å…¨é¢çš„åŠŸèƒ½å’Œè¿›åŒ–åˆ†æã€‚
- **å±€é™æ€§**ï¼š
  - **è®¡ç®—èµ„æºéœ€æ±‚é«˜**ï¼šç»„è£…è¿‡ç¨‹éœ€è¦è¾ƒé«˜çš„è®¡ç®—èµ„æºå’Œæ—¶é—´ã€‚
  - **å¤æ‚æ€§**ï¼šç»„è£…å’Œbinningæ­¥éª¤å¢åŠ äº†åˆ†æçš„å¤æ‚æ€§ï¼Œéœ€è¦æ›´å¤šçš„æŠ€æœ¯ç»éªŒå’Œå·¥å…·æ”¯æŒã€‚

### ç»„è£…ï¼ˆassemblyï¼‰

ç»„è£…ï¼ˆassemblyï¼‰æ˜¯æŒ‡å°†æµ‹åºäº§ç”Ÿçš„çŸ­è¯»é•¿ï¼ˆreadsï¼‰æ‹¼æ¥æˆæ›´é•¿çš„è¿ç»­åºåˆ—ï¼ˆcontigsï¼‰çš„è¿‡ç¨‹ã€‚è¿™ä¸ªè¿‡ç¨‹åœ¨åŸºå› ç»„å­¦å’Œå®åŸºå› ç»„å­¦ç ”ç©¶ä¸­è‡³å…³é‡è¦ï¼Œå› ä¸ºå®ƒèƒ½å¤Ÿå°†ç‰‡æ®µåŒ–çš„ä¿¡æ¯æ•´åˆæˆæ›´å®Œæ•´çš„åŸºå› ç»„åºåˆ—ï¼Œä¾¿äºåç»­çš„åŠŸèƒ½å’Œåˆ†ç±»åˆ†æã€‚

æ³¨æ„assemblyæœ‰å•æ ·æœ¬æ‹¼è£…å’Œæ··æ‹¼ç­‰ç­–ç•¥ï¼Œå¯ä»¥è‡ªè¡Œæ ¹æ®è®¡ç®—èµ„æºå’Œç ”ç©¶ç›®çš„é€‰æ‹©ã€‚

- å•æ ·æœ¬æ‹¼è£…ï¼ˆSingle-sample Assemblyï¼‰

**å®šä¹‰**ï¼šå•æ ·æœ¬æ‹¼è£…æ˜¯æŒ‡å¯¹æ¥è‡ªå•ä¸€ç¯å¢ƒæˆ–æ¡ä»¶ä¸‹çš„æ ·æœ¬è¿›è¡Œç»„è£…ã€‚è¿™ç§ç­–ç•¥é€‚ç”¨äºç›¸å¯¹ç®€å•çš„æ ·æœ¬ï¼Œæˆ–è€…å¸Œæœ›å•ç‹¬åˆ†ææ¯ä¸ªæ ·æœ¬çš„åŸºå› ç»„ç»„æˆã€‚

**ä¼˜ç‚¹**ï¼š
1. **ç‹¬ç«‹åˆ†æ**ï¼šæ¯ä¸ªæ ·æœ¬å•ç‹¬ç»„è£…ï¼Œæœ‰åŠ©äºæ·±å…¥åˆ†ææ ·æœ¬ç‰¹å®šçš„åŸºå› ç»„ç‰¹å¾ã€‚
2. **æ•°æ®ç®€å•**ï¼šé€‚ç”¨äºæ•°æ®å¤æ‚åº¦è¾ƒä½çš„æ ·æœ¬ï¼Œç»„è£…ç»“æœè¾ƒä¸ºæ¸…æ™°ã€‚

**ç¼ºç‚¹**ï¼š
1. **èµ„æºéœ€æ±‚é«˜**ï¼šå¯¹æ¯ä¸ªæ ·æœ¬è¿›è¡Œç‹¬ç«‹ç»„è£…ï¼Œè®¡ç®—èµ„æºå’Œæ—¶é—´éœ€æ±‚è¾ƒé«˜ã€‚
2. **æœ‰é™çš„è¦†ç›–åº¦**ï¼šå¯¹äºä½ä¸°åº¦å¾®ç”Ÿç‰©ï¼Œå•æ ·æœ¬ç»„è£…å¯èƒ½æ— æ³•æä¾›è¶³å¤Ÿçš„è¦†ç›–åº¦å’Œç»„è£…è´¨é‡ã€‚

**é€‚ç”¨åœºæ™¯**ï¼š
- ç¯å¢ƒæ ·æœ¬è¾ƒä¸ºç®€å•çš„ç ”ç©¶ã€‚
- éœ€è¦å¯¹æ¯ä¸ªæ ·æœ¬è¿›è¡Œç‹¬ç«‹æ¯”è¾ƒå’Œåˆ†æçš„ç ”ç©¶ã€‚

- æ··æ‹¼ç­–ç•¥ï¼ˆCo-assemblyï¼‰

**å®šä¹‰**ï¼šæ··æ‹¼æ˜¯æŒ‡å°†å¤šä¸ªæ ·æœ¬çš„æ•°æ®åˆå¹¶åœ¨ä¸€èµ·è¿›è¡Œç»„è£…ã€‚è¿™ç§ç­–ç•¥é€‚ç”¨äºå¤æ‚çš„å®åŸºå› ç»„æ ·æœ¬ï¼Œé€šè¿‡æ•´åˆå¤šä¸ªæ ·æœ¬çš„æ•°æ®ï¼Œå¯ä»¥æé«˜ç»„è£…çš„è¦†ç›–åº¦å’Œè´¨é‡ã€‚

**ä¼˜ç‚¹**ï¼š
1. **æé«˜è¦†ç›–åº¦**ï¼šåˆå¹¶å¤šä¸ªæ ·æœ¬çš„æ•°æ®ï¼Œå¢åŠ äº†åºåˆ—çš„è¦†ç›–åº¦ï¼Œæœ‰åŠ©äºæ›´å®Œæ•´çš„ç»„è£…ã€‚
2. **å¤„ç†å¤æ‚æ ·æœ¬**ï¼šé€‚ç”¨äºå¤æ‚çš„ç¯å¢ƒæ ·æœ¬ï¼Œèƒ½å¤Ÿæ•æ‰åˆ°æ›´å¤šçš„ä½ä¸°åº¦å¾®ç”Ÿç‰©å’ŒåŸºå› ç»„ä¿¡æ¯ã€‚

**ç¼ºç‚¹**ï¼š
1. **å¤æ‚æ€§å¢åŠ **ï¼šæ··æ‹¼çš„æ•°æ®å¤æ‚åº¦é«˜ï¼Œç»„è£…å’Œåç»­åˆ†ææ›´ä¸ºå¤æ‚ã€‚
2. **æ ·æœ¬é—´å·®å¼‚æ¨¡ç³ŠåŒ–**ï¼šæ ·æœ¬é—´çš„ç‹¬ç‰¹ç‰¹å¾å¯èƒ½åœ¨æ··æ‹¼è¿‡ç¨‹ä¸­è¢«æ¨¡ç³ŠåŒ–ï¼Œå½±å“å¯¹ä¸ªä½“æ ·æœ¬çš„å…·ä½“åˆ†æã€‚

**é€‚ç”¨åœºæ™¯**ï¼š
- ç¯å¢ƒæ ·æœ¬å¤æ‚ï¼ŒåŒ…å«å¤šç§å¾®ç”Ÿç‰©ç¾¤è½ã€‚
- éœ€è¦æé«˜ä½ä¸°åº¦ç‰©ç§çš„æ£€æµ‹èƒ½åŠ›ã€‚
- ç›®æ ‡æ˜¯è·å–æ•´ä½“å¾®ç”Ÿç‰©ç¾¤è½çš„ç»¼åˆä¿¡æ¯ã€‚

#### megahit

Megahitæ˜¯ä¸€ä¸ªç”¨äºå¯¹é«˜é€šé‡æµ‹åºæ•°æ®è¿›è¡Œde novoç»„è£…çš„è½¯ä»¶ã€‚
å®ƒä½¿ç”¨äº†ä¸€ç§åŸºäºçŸ­è¯»æ¯”å¯¹å’Œå›¾å½¢æ„å»ºçš„ç®—æ³•æ¥ç»„è£…åŸºå› ç»„ï¼Œå¯ä»¥é«˜æ•ˆåœ°å¤„ç†å¤§è§„æ¨¡çš„æ•°æ®é›†ã€‚
ä»¥ä¸‹æ˜¯Megahitçš„ä¸€äº›ä¼˜ç‚¹å’Œé€‚ç”¨æƒ…å†µï¼š

1.  é€Ÿåº¦å¿«ï¼šMegahitçš„ç®—æ³•éå¸¸é«˜æ•ˆï¼Œå¯ä»¥å¤„ç†å¤§è§„æ¨¡çš„æ•°æ®é›†ï¼Œé€šå¸¸æ¯”å…¶ä»–de novoç»„è£…å·¥å…·æ›´å¿«ã€‚
2.  é«˜è´¨é‡çš„ç»„è£…ï¼šMegahitåœ¨ç»„è£…ç»“æœçš„è¿é€šæ€§å’Œå‡†ç¡®æ€§æ–¹é¢è¡¨ç°ä¼˜å¼‚ï¼Œå°¤å…¶åœ¨å¤„ç†é«˜GCå«é‡åŸºå› ç»„æ—¶æ•ˆæœæ˜¾è‘—ã€‚
3.  é€‚ç”¨äºä¸åŒç±»å‹çš„æµ‹åºæ•°æ®ï¼šMegahitæ”¯æŒå¤šç§ä¸åŒç±»å‹çš„æµ‹åºæ•°æ®ï¼ŒåŒ…æ‹¬ Illumina HiSeq/MiSeqã€IonTorrentå’ŒPacBioç­‰å¹³å°ã€‚
4.  æ˜“äºä½¿ç”¨ï¼šMegahitå…·æœ‰ç®€å•çš„å‘½ä»¤è¡Œè¯­æ³•ï¼Œæ–¹ä¾¿ç”¨æˆ·è¿›è¡Œç»„è£…æ“ä½œï¼Œä¸”å…·æœ‰ä¸­æ–­ç‚¹ï¼Œé¿å…å¤±è´¥åå…¨éƒ¨é‡è·‘ã€‚

``` bash
#!/bin/bash
#SBATCH --job-name=megahit
#SBATCH --output=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.out
#SBATCH --error=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.err
#SBATCH --array=1
#SBATCH --partition=cpu
#SBATCH --cpus-per-task=8
#SBATCH --mem-per-cpu=10G
samplelist=/share/home/jianglab/pengchen/work/asthma/samplelist

echo start: `date +"%Y-%m-%d %T"`
start=`date +%s`

####################
echo "SLURM_ARRAY_TASK_ID: " $SLURM_ARRAY_TASK_ID
START=$SLURM_ARRAY_TASK_ID

NUMLINES=40 #how many sample in one array

STOP=$((SLURM_ARRAY_TASK_ID*NUMLINES))
START="$(($STOP - $(($NUMLINES - 1))))"

echo "START=$START"
echo "STOP=$STOP"

####################
for (( N = $START; N <= $STOP; N++ ))
do
  sample=$(head -n "$N" $samplelist | tail -n 1)
  echo $sample
  #single sample
  mkdir -p result/megahit
  mkdir -p result/megahit/contigs
  ~/miniconda3/envs/waste/bin/megahit -t 8 -1 data/rm_human/${sample}.1.fq \
    -2 data/rm_human/${sample}.2.fq \
    -o result/megahit/${sample} --out-prefix ${sample}
  sed -i "/>/s/>/>${sample}_/" result/megahit/${sample}/${sample}.contigs.fa
  mv result/megahit/${sample}/${sample}.contigs.fa result/megahit/contigs/
done

##############
echo end: `date +"%Y-%m-%d %T"`
end=`date +%s`
echo TIME:`expr $end - $start`s
```

æ··æ‹¼çš„è¯ï¼š

``` bash
#!/bin/bash
#SBATCH --job-name=megahit2
#SBATCH --output=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.out
#SBATCH --error=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.err
#SBATCH --array=1
#SBATCH --partition=mem
#SBATCH --cpus-per-task=8
#SBATCH --mem-per-cpu=100G
samplelist=/share/home/jianglab/pengchen/work/asthma/samplelist

echo start: `date +"%Y-%m-%d %T"`
start=`date +%s`

####################

  #multi-sampleæ··æ‹¼
  #éœ€è¦å¤§å†…å­˜ï¼Œå»ºè®®3å€fq.gzå†…å­˜
  mkdir -p data/com_read
  mkdir -p result/megahit2

  for i in `cat samplelist`
  do
      echo ${i}
      cat data/rm_human/${i}.1.fq >> data/com_read/R1.fq
      cat data/rm_human/${i}.2.fq >> data/com_read/R2.fq
  done

  ~/miniconda3/envs/waste/bin/megahit -t 8 -1 data/com_read/R1.fq \
    -2 data/com_read/R2.fq \
    -o result/megahit2/ --out-prefix multi_sample

##############
echo end: `date +"%Y-%m-%d %T"`
end=`date +%s`
echo TIME:`expr $end - $start`s
```

#### metaSPAdes

SPAdes æ˜¯ä¸€æ¬¾å¤šåŠŸèƒ½å·¥å…·åŒ…ï¼Œä¸“ä¸ºæµ‹åºæ•°æ®çš„ç»„è£…å’Œåˆ†æè€Œè®¾è®¡ã€‚ SPAdes ä¸»è¦æ˜¯ä¸º Illumina æµ‹åºæ•°æ®è€Œå¼€å‘çš„ï¼Œä½†ä¹Ÿå¯ç”¨äº IonTorrentã€‚å¤§å¤šæ•° SPAdes ç®¡é“æ”¯æŒæ··åˆæ¨¡å¼ï¼Œå³å…è®¸ä½¿ç”¨é•¿è¯»å–ï¼ˆPacBio å’Œ Oxford Nanoporeï¼‰ä½œä¸ºè¡¥å……æ•°æ®ã€‚

è€ŒmetaSPAdesæ˜¯ç›®å‰å®åŸºå› ç»„é¢†åŸŸç»„è£…æŒ‡æ ‡è¾ƒå¥½çš„è½¯ä»¶ï¼Œå°¤å…¶åœ¨æ ªæ°´å¹³ç»„è£…ä¼˜åŠ¿æ˜æ˜¾ã€‚ç›¸æ¯”Megahitè¡¨ç°å‡ºæ›´å¥½çš„åŸºå› é•¿åº¦å’Œè¯»é•¿æ¯”è¾ƒç‡ï¼Œä½†æ˜¯æ›´å¥½çš„ç»„è£…è´¨é‡ä¹Ÿä¼´éšç€æ›´é•¿æ—¶é—´å’Œå†…å­˜æ¶ˆè€—ï¼ŒåŒæ—¶ä¹Ÿæœ‰é”™è¯¯ç»„è£…ä¸Šå‡çš„é£é™©ã€‚

``` bash
#!/bin/bash
#SBATCH --job-name=asthma_megahit
#SBATCH --output=/share/home/jianglab/pengchen/work/asthma/megahit/log/%x_%a.out
#SBATCH --error=/share/home/jianglab/pengchen/work/asthma/megahit/log/%x_%a.err
#SBATCH --array=1-33
#SBATCH --partition=cpu
#SBATCH --cpus-per-task=32

echo start: `date +'%Y-%m-%d %T'`
start=`date +%s`
echo "SLURM_ARRAY_TASK_ID: " $SLURM_ARRAY_TASK_ID
sample=$(head -n $SLURM_ARRAY_TASK_ID ~/work/asthma/data/namelist | tail -1)
#sample=$(head -n 1 namelist | tail -1)
echo handling: $sample	
####################
for (( N = $START; N <= $STOP; N++ ))
do
  sample=$(head -n "$N" $samplelist | tail -n 1)
  echo $sample
  #single sample
  mkdir -p result/metaspades
  mkdir -p result/metaspades/contigs
  ~/miniconda3/envs/metawrap/bin/metaspades.py â€“t 8 -k 21,33,55,77,99,127 --careful \
    -1 data/rm_human/${sample}.1.fq \
    -2 data/rm_human/${sample}.2.fq \
    -o result/metaspades/${sample}

  sed -i "/>/s/>/>${sample}_/" result/metaspades/${sample}/scaffolds.fasta
  mv result/metaspades/${sample}/scaffolds.fasta result/metaspades/contigs/
done
####################
echo end: `date +'%Y-%m-%d %T'`
end=`date +%s`
echo TIME:`expr $end - $start`s
```

### ç»„è£…è¯„ä¼°ï¼šQUAST

**QUAST**æ˜¯ä¸€ä¸ªè´¨é‡è¯„ä¼°å·¥å…·ã€‚ QUASTå¯ä»¥ä½¿ç”¨å‚è€ƒåŸºå› ç»„ä»¥åŠä¸ä½¿ç”¨å‚è€ƒåŸºå› ç»„æ¥è¯„ä¼°è£…é…ã€‚ QUASTç”Ÿæˆè¯¦ç»†çš„æŠ¥å‘Šï¼Œè¡¨æ ¼å’Œå›¾è§£ï¼Œä»¥æ˜¾ç¤ºè£…é…çš„ä¸åŒæ–¹é¢ã€‚

### åŸºå› é¢„æµ‹ï¼šProdigal

åŸºå› é¢„æµ‹æ˜¯æŒ‡åœ¨åŸºå› ç»„åºåˆ—ï¼ˆå¦‚contigsï¼‰ä¸­è¯†åˆ«å‡ºç¼–ç åŒºåºåˆ—ï¼ˆCDS, Coding DNA Sequencesï¼‰çš„è¿‡ç¨‹ã€‚è¿™ä¸ªè¿‡ç¨‹å¯¹äºç†è§£åŸºå› ç»„åŠŸèƒ½å’Œæ³¨é‡ŠåŸºå› ç»„ä¸­çš„åŸºå› æ˜¯è‡³å…³é‡è¦çš„ã€‚å¸¸ç”¨çš„è½¯ä»¶æœ‰Prodigalï¼ŒGeneMarkï¼ˆGeneMarkSç”¨äºç»†èŒï¼ŒGeneMark-ESç”¨äºçœŸæ ¸ç”Ÿç‰©ï¼‰ï¼ŒGlimmerç­‰ã€‚

è¿™é‡Œä»¥Prodigalä¸ºä¾‹ï¼š
è¾“å…¥æ–‡ä»¶ï¼šæ‹¼è£…å¥½çš„åºåˆ—æ–‡ä»¶ megahit/final.contigs.fa
è¾“å‡ºæ–‡ä»¶ï¼šprodigalé¢„æµ‹çš„åŸºå› åºåˆ— prodigal/gene.fa

prodigalä¸æ”¯æŒå¤šçº¿ç¨‹è¿è¡Œï¼Œæ‰€ä»¥æˆ‘ä»¬å¯ä»¥è‡ªè¡Œåˆ†å‰²åºåˆ—æ–‡ä»¶è°ƒç”¨å¤šä¸ªprodigalç¨‹åºåˆ†åˆ«è·‘å®ç°ä¼ªå¤šçº¿ç¨‹ã€‚

``` bash
#!/bin/bash
#SBATCH --job-name=prodigal
#SBATCH --output=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.out
#SBATCH --error=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.err
#SBATCH --array=1
#SBATCH --partition=cpu
#SBATCH --cpus-per-task=1
#SBATCH --mem-per-cpu=2G
samplelist=/share/home/jianglab/pengchen/work/asthma/samplelist

echo start: `date +"%Y-%m-%d %T"`
start=`date +%s`

####################
echo "SLURM_ARRAY_TASK_ID: " $SLURM_ARRAY_TASK_ID
START=$SLURM_ARRAY_TASK_ID

NUMLINES=40 #how many sample in one array

STOP=$((SLURM_ARRAY_TASK_ID*NUMLINES))
START="$(($STOP - $(($NUMLINES - 1))))"

echo "START=$START"
echo "STOP=$STOP"

####################
for (( N = $START; N <= $STOP; N++ ))
do
  sample=$(head -n "$N" $samplelist | tail -n 1)
  echo $sample
  mkdir -p result/prodigal
  mkdir -p result/prodigal/gene_fa
  mkdir -p result/prodigal/gene_gff
  ~/miniconda3/envs/waste/bin/prodigal -i result/megahit/contigs/${sample}.contigs.fa \
      -d result/prodigal/gene_fa/${sample}.gene.fa \
      -o result/prodigal/gene_gff/${sample}.gene.gff \
      -p meta -f gff

  mkdir -p result/prodigal/fullgene_fa
  ## æå–å®Œæ•´åŸºå› (å®Œæ•´ç‰‡æ®µè·å¾—çš„åŸºå› å…¨ä¸ºå®Œæ•´ï¼Œå¦‚æˆç¯çš„ç»†èŒåŸºå› ç»„)
  grep 'partial=00' result/prodigal/gene_fa/${sample}.gene.fa | cut -f1 -d ' '| sed 's/>//' > result/prodigal/gene_fa/${sample}.fullid
  seqkit grep -f result/prodigal/gene_fa/${sample}.fullid result/prodigal/gene_fa/${sample}.gene.fa > result/prodigal/fullgene_fa/${sample}.gene.fa

done

##############
echo end: `date +"%Y-%m-%d %T"`
end=`date +%s`
echo TIME:`expr $end - $start`s
```

### å»å†—ä½™

åŸºå› é›†å»å†—ä½™æ˜¯æŒ‡åœ¨é¢„æµ‹åˆ°çš„åŸºå› é›†ä¸­å»é™¤é‡å¤çš„åŸºå› æˆ–é«˜åº¦ç›¸ä¼¼çš„åŸºå› ï¼Œä»¥è·å¾—ä¸€ä¸ªéå†—ä½™çš„åŸºå› é›†åˆã€‚è¿™ä¸€æ­¥éª¤åœ¨å®åŸºå› ç»„åˆ†æä¸­éå¸¸é‡è¦ï¼Œå› ä¸ºå®åŸºå› ç»„æ ·æœ¬ä¸­é€šå¸¸åŒ…å«æ¥è‡ªå¤šç§å¾®ç”Ÿç‰©çš„é‡å¤æˆ–ç›¸ä¼¼åºåˆ—ï¼Œå»å†—ä½™å¯ä»¥ç®€åŒ–æ•°æ®é›†å¹¶æé«˜åç»­åˆ†æçš„æ•ˆç‡å’Œå‡†ç¡®æ€§ã€‚å¸¸ç”¨è½¯ä»¶æœ‰CD-HITï¼ŒMMseqs2ï¼ŒUCLUSTï¼ˆé›†æˆåœ¨QIIMEå’ŒUSEARCHä¸­çš„å·¥å…·ï¼‰ç­‰ã€‚

ä¸Šé¢äº§ç”Ÿäº†nä¸ªæ ·æœ¬çš„åŸºå› é¢„æµ‹ç»“æœæ–‡ä»¶ï¼Œgene.faæ–‡ä»¶è¦å…ˆæƒ³åŠæ³•æ•´åˆä¸ºä¸€ä¸ªæ–‡ä»¶å†å»å»å†—ä½™ã€‚

#### CD-HIT

ä¹‹å‰å¤§å®¶å¸¸ç”¨çš„æ˜¯CD-HITæ¥å»å†—ä½™ï¼š

``` bash
#!/bin/bash
#SBATCH --job-name=cluster
#SBATCH --output=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.out
#SBATCH --error=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.err
#SBATCH --array=1
#SBATCH --partition=cpu
#SBATCH --nodes=1
#SBATCH --tasks-per-node=1
#SBATCH --cpus-per-task=1
#SBATCH --mem-per-cpu=2G
samplelist=/share/home/jianglab/pengchen/work/asthma/samplelist

echo start: `date +"%Y-%m-%d %T"`
start=`date +%s`

####################
  echo "start merge"
  cat result/prodigal/gene_fa/*.gene.fa > result/prodigal/all.gene.fa
  cat result/prodigal/fullgene_fa/*.gene.fa > result/prodigal/all.fullgene.fa
  echo "done"
  
  ~/miniconda3/envs/waste/bin/cd-hit-est -i result/prodigal/all.gene.fa \
    -o result/NR/nucleotide.fa \
    -aS 0.9 -c 0.9 -G 0 -g 0 -T 0 -M 0

  mv result/NR/nucleotide_rep_seq.fasta result/NR/nucleotide.fa
  rm result/NR/nucleotide_all_seqs.fasta
  ~/miniconda3/envs/waste/bin/seqkit translate result/NR/nucleotide.fa > result/NR/protein.fa
  sed 's/\*//g' result/NR/protein.fa > result/NR/protein_no_end.fa
  rm result/NR/protein.fa

##############
echo end: `date +"%Y-%m-%d %T"`
end=`date +%s`
echo TIME:`expr $end - $start`s
```

#### Mmseqs2

æœ€è¿‘å‘ç°äº†Mmseqs2è¿™ä¸ªç¥å™¨ï¼Œè¿™ä¸ªèšç±»è¦æ¯”CD-HITå¿«éå¸¸å¤šï¼Œå…¶ä»–åº”ç”¨å‚è§[Mmseqs2çš„åŸºç¡€ä½¿ç”¨](../mmseqs2)ï¼š

``` bash
#!/bin/bash
#SBATCH --job-name=cluster
#SBATCH --output=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.out
#SBATCH --error=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.err
#SBATCH --array=1
#SBATCH --partition=cpu
#SBATCH --nodes=1
#SBATCH --tasks-per-node=1
#SBATCH --cpus-per-task=1
#SBATCH --mem-per-cpu=2G
samplelist=/share/home/jianglab/pengchen/work/asthma/samplelist

echo start: `date +"%Y-%m-%d %T"`
start=`date +%s`

####################
  echo "start merge"
  cat result/prodigal/gene_fa/*.gene.fa > result/prodigal/all.gene.fa
  cat result/prodigal/fullgene_fa/*.gene.fa > result/prodigal/all.fullgene.fa
  echo "done"
  
  mkdir -p result/NR
  mmseqs easy-linclust result/prodigal/all.gene.fa result/NR/nucleotide mmseqs_tmp \
    --min-seq-id 0.9 -c 0.9 --cov-mode 1  --threads 8

  mv result/NR/nucleotide_rep_seq.fasta result/NR/nucleotide.fa
  rm result/NR/nucleotide_all_seqs.fasta
  ~/miniconda3/envs/waste/bin/seqkit translate result/NR/nucleotide.fa > result/NR/protein.fa
  sed 's/\*//g' result/NR/protein.fa > result/NR/protein_no_end.fa
  rm result/NR/protein.fa

##############
echo end: `date +"%Y-%m-%d %T"`
end=`date +%s`
echo TIME:`expr $end - $start`s
```

### fastaä¿¡æ¯ç»Ÿè®¡

ä½¿ç”¨seqkitç»Ÿè®¡åˆšåˆšå¾—åˆ°çš„ä¸€äº›å…³é”®fastaæ–‡ä»¶çš„åŸºç¡€ä¿¡æ¯ã€‚

``` bash
test_file=`head -n1 $samplelist`
if [ -f result/megahit/contigs/${test_file}.contigs.fa ]; then
~/miniconda3/envs/waste/bin/seqkit stats result/megahit/contigs/*.contigs.fa >result/megahit/contig_stats
fi
if [ -f result/prodigal/gene_fa/${test_file}.gene.fa ]; then
~/miniconda3/envs/waste/bin/seqkit stats result/prodigal/gene_fa/*.gene.fa >result/prodigal/gene_fa_stats
fi
if [ -f result/prodigal/fullgene_fa/${test_file}.gene.fa ]; then
~/miniconda3/envs/waste/bin/seqkit stats result/prodigal/fullgene_fa/*.gene.fa >result/prodigal/fullgene_fa_stats
fi
if [ -f result/NR/nucleotide.fa ]; then
~/miniconda3/envs/waste/bin/seqkit stats result/NR/nucleotide.fa >result/NR/nucleotide_stat
fi
```

### åŸºå› å®šé‡ï¼šsalmon

1.  å»ºç«‹ç´¢å¼•

``` bash
#!/bin/bash
#SBATCH --job-name=salmon-index
#SBATCH --output=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.out
#SBATCH --error=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.err
#SBATCH --array=1
#SBATCH --partition=cpu
#SBATCH --cpus-per-task=4
#SBATCH --mem-per-cpu=10G
echo start: `date +"%Y-%m-%d %T"`
start=`date +%s`

####################

  ## å»ºç´¢å¼•, -tåºåˆ—, -i ç´¢å¼•
  # å¤§ç‚¹å†…å­˜
  mkdir -p result/salmon
  ~/miniconda3/envs/waste/share/salmon/bin/salmon index \
    -t result/NR/nucleotide.fa \
    -p 4 \
    -i result/salmon/index

##############
echo end: `date +"%Y-%m-%d %T"`
end=`date +%s`
echo TIME:`expr $end - $start`s
```

2.  å¯¹æ¯ä¸ªæ ·æœ¬å®šé‡

``` bash
#!/bin/bash
#SBATCH --job-name=salmon-quant
#SBATCH --output=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.out
#SBATCH --error=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.err
#SBATCH --array=1
#SBATCH --partition=cpu
#SBATCH --cpus-per-task=4
#SBATCH --mem-per-cpu=2G
samplelist=/share/home/jianglab/pengchen/work/asthma/samplelist

echo start: `date +"%Y-%m-%d %T"`
start=`date +%s`

####################
echo "SLURM_ARRAY_TASK_ID: " $SLURM_ARRAY_TASK_ID
START=$SLURM_ARRAY_TASK_ID

NUMLINES=40 #how many sample in one array

STOP=$((SLURM_ARRAY_TASK_ID*NUMLINES))
START="$(($STOP - $(($NUMLINES - 1))))"

echo "START=$START"
echo "STOP=$STOP"

####################
## è¾“å…¥æ–‡ä»¶ï¼šå»å†—ä½™åçš„åŸºå› å’Œè›‹ç™½åºåˆ—ï¼šNR/nucleotide.fa
## è¾“å‡ºæ–‡ä»¶ï¼šSalmonå®šé‡åçš„ç»“æœï¼šsalmon/gene.count; salmon/gene.TPM
## å®šé‡ï¼Œlæ–‡åº“ç±»å‹è‡ªåŠ¨é€‰æ‹©ï¼Œpçº¿ç¨‹ï¼Œ--metaå®åŸºå› ç»„æ¨¡å¼
for (( N = $START; N <= $STOP; N++ ))
do
  sample=$(head -n "$N" $samplelist | tail -n 1)
  echo $sample
  mkdir -p result/salmon/quant
  ~/miniconda3/envs/waste/share/salmon/bin/salmon quant \
      --validateMappings \
      -i result/salmon/index -l A -p 4 --meta \
      -1 data/rm_human/${sample}.1.fq \
      -2 data/rm_human/${sample}.2.fq \
      -o result/salmon/quant/${sample}.quant
done

##############
echo end: `date +"%Y-%m-%d %T"`
end=`date +%s`
echo TIME:`expr $end - $start`s
```

3.  åˆå¹¶å„æ ·æœ¬ç»“æœ

``` bash
#!/bin/bash
#SBATCH --job-name=salmon-merge
#SBATCH --output=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.out
#SBATCH --error=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.err
#SBATCH --array=1
#SBATCH --partition=cpu
#SBATCH --cpus-per-task=1
#SBATCH --mem-per-cpu=2G
samplelist=/share/home/jianglab/pengchen/work/asthma/samplelist

echo start: `date +"%Y-%m-%d %T"`
start=`date +%s`

####################

  ls result/salmon/quant/*.quant/quant.sf |awk -F'/' '{print $(NF-1)}' |sed 's/.quant//' >tmp_finished
  diff_rows -f samplelist -s tmp_finished >tmp_diff
  # è®¡ç®—ç»“æœçš„è¡Œæ•°
  line_count=$( cat tmp_diff| wc -l)

  # æ£€æŸ¥è¡Œæ•°æ˜¯å¦å¤§äº1ï¼Œå¦‚æœæ˜¯åˆ™ç»ˆæ­¢è„šæœ¬
  if [ "$line_count" -gt 1 ]; then
    echo 'sfæ–‡ä»¶å’Œsamplelistæ•°é‡ä¸ä¸€è‡´'
    exit 1
  fi

  ##mapping rates
  cat result/salmon/quant/*.quant/logs/salmon_quant.log |grep 'Mapping rate = '|awk '{print $NF}'> tmp
  paste samplelist tmp > result/salmon/mapping_rates

  ## combine
  ~/miniconda3/envs/waste/share/salmon/bin/salmon quantmerge \
      --quants result/salmon/quant/*.quant \
      -o result/salmon/gene.TPM
  ~/miniconda3/envs/waste/share/salmon/bin/salmon quantmerge \
      --quants result/salmon/quant/*.quant \
      --column NumReads -o result/salmon/gene.count
  sed -i '1 s/.quant//g' result/salmon/gene.*

##############
echo end: `date +"%Y-%m-%d %T"`
end=`date +%s`
echo TIME:`expr $end - $start`s
```

### åŠŸèƒ½åŸºå› æ³¨é‡Š

åŠŸèƒ½åŸºå› æ³¨é‡Šæ˜¯å¯¹éå†—ä½™åŸºå› é›†è¿›è¡Œåˆ†ç±»å’ŒåŠŸèƒ½æè¿°çš„è¿‡ç¨‹ï¼Œä»¥ä¾¿ç†è§£è¿™äº›åŸºå› åœ¨ç”Ÿç‰©ä½“å†…å¯èƒ½çš„è§’è‰²å’Œä½œç”¨ã€‚

ä¸Šä¸€æ­¥å·²ç»æœ‰äº†æ‰€æœ‰çš„åŸºå› å’Œæ¯ä¸ªæ ·æœ¬æ‰€æœ‰åŸºå› çš„read countå®šé‡ç»“æœï¼Œ
æˆ‘ä»¬åªéœ€è¦å¯¹ä¸Šä¸€æ­¥çš„åŸºå› åºåˆ—ï¼ˆæˆ–è›‹ç™½è´¨åºåˆ—ï¼‰è¿›è¡Œä¸åŒæ•°æ®åº“çš„æ³¨é‡Šï¼Œ
ç„¶ååˆå¹¶æ³¨é‡Šç»“æœå¾—åˆ°çš„å°±æ˜¯åŠŸèƒ½ä¸°åº¦è¡¨ã€‚

å¾ˆå¤šæ•°æ®åº“è‡ªå¸¦è½¯ä»¶éƒ½æ˜¯ç”¨diamondæ¯”å¯¹ï¼Œå¦‚æœæ²¡æœ‰ä¸“ç”¨è½¯ä»¶çš„æ•°æ®åº“æˆ‘ä»¬ä¹Ÿå¯ä»¥è‡ªå·±ç”¨diamondæ¯”å¯¹ã€‚

diamondé€‰æ‹©--outfmt 6çš„è¾“å‡ºç»“æœå’Œblastpæ ¼å¼ä¸€æ ·ã€‚

#### eggNOG (COG/KEGG/CAZy)

[EggNOGæ•°æ®åº“](http://eggnogdb.embl.de/)æ”¶é›†äº†COGï¼ˆClusters of Orthologous Groups of proteinsï¼Œç›´ç³»åŒæºè›‹ç™½ç°‡ï¼‰,æ„æˆæ¯ä¸ªCOGçš„è›‹ç™½éƒ½æ˜¯è¢«å‡å®šä¸ºæ¥è‡ªäºä¸€ä¸ªç¥–å…ˆè›‹ç™½ï¼Œå› æ­¤æ˜¯orthologsæˆ–è€…æ˜¯paralogsã€‚é€šè¿‡æŠŠæ‰€æœ‰å®Œæ•´åŸºå› ç»„çš„ç¼–ç è›‹ç™½ä¸€ä¸ªä¸€ä¸ªçš„äº’ç›¸æ¯”è¾ƒç¡®å®šçš„ã€‚åœ¨è€ƒè™‘æ¥è‡ªä¸€ä¸ªç»™å®šåŸºå› ç»„çš„è›‹ç™½æ—¶ï¼Œè¿™ç§æ¯”è¾ƒå°†ç»™å‡ºæ¯ä¸ªå…¶ä»–åŸºå› ç»„çš„ä¸€ä¸ªæœ€ç›¸ä¼¼çš„è›‹ç™½ï¼ˆå› æ­¤éœ€è¦ç”¨å®Œæ•´çš„åŸºå› ç»„æ¥å®šä¹‰COGï¼‰ï¼Œè¿™äº›åŸºå› çš„æ¯ä¸€ä¸ªéƒ½è½®ç•ªåœ°è¢«è€ƒè™‘ã€‚å¦‚æœåœ¨è¿™äº›è›‹ç™½ï¼ˆæˆ–å­é›†ï¼‰ä¹‹é—´ä¸€ä¸ªç›¸äº’çš„æœ€ä½³åŒ¹é…å…³ç³»è¢«å‘ç°ï¼Œé‚£ä¹ˆé‚£äº›ç›¸äº’çš„æœ€ä½³åŒ¹é…å°†å½¢æˆä¸€ä¸ªCOGã€‚è¿™æ ·ï¼Œä¸€ä¸ªCOGä¸­çš„æˆå‘˜å°†ä¸è¿™ä¸ªCOGä¸­çš„å…¶ä»–æˆå‘˜æ¯”èµ·è¢«æ¯”è¾ƒçš„åŸºå› ç»„ä¸­çš„å…¶ä»–è›‹ç™½æ›´ç›¸åƒã€‚

[EggNOG](http://eggnogdb.embl.de/)é‡Œé¢å·²ç»åŒ…å«äº†GOï¼ŒKEGGï¼ŒCAZyç­‰ã€‚

``` bash
## ä¸‹è½½å¸¸ç”¨æ•°æ®åº“ï¼Œæ³¨æ„è®¾ç½®ä¸‹è½½ä½ç½®
mkdir -p ${db}/eggnog5 && cd ${db}/eggnog5
## -yé»˜è®¤åŒæ„ï¼Œ-få¼ºåˆ¶ä¸‹è½½ï¼Œeggnog.db.gz 7.9G+4.9G
download_eggnog_data.py -y -f --data_dir ./

## ä¸‹è½½æ–¹å¼2(å¯é€‰)ï¼šé“¾æ¥ç›´æ¥ä¸‹è½½
wget -c http://eggnog5.embl.de/download/emapperdb-5.0.0/eggnog.db.gz ## 7.9G
wget -c http://eggnog5.embl.de/download/emapperdb-5.0.0/eggnog_proteins.dmnd.gz ## 4.9G
gunzip *.gz
```

``` bash
#!/bin/bash
#SBATCH --job-name=eggnog
#SBATCH --output=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.out
#SBATCH --error=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.err
#SBATCH --array=1
#SBATCH --partition=cpu
#SBATCH --cpus-per-task=8
#SBATCH --mem-per-cpu=10G
echo start: `date +"%Y-%m-%d %T"`
start=`date +%s`
####################
  ## å¤§ç‚¹å†…å­˜, æ•°æ®åº“æœ‰50Gå·¦å³
  source ~/miniconda3/etc/profile.d/conda.sh
  conda activate func
  ## diamondæ¯”å¯¹åŸºå› è‡³eggNOG 5.0æ•°æ®åº“, 1~9hï¼Œé»˜è®¤diamond 1e-3
  mkdir -p result/eggnog
  emapper.py --no_annot --no_file_comments --override \
    --data_dir ~/db/eggnog5 \
    -i result/NR/protein_no_end.fa \
    --cpu 8 -m diamond \
    -o result/eggnog/protein
  ## æ¯”å¯¹ç»“æœåŠŸèƒ½æ³¨é‡Š, 1h
  emapper.py \
    --annotate_hits_table result/eggnog/protein.emapper.seed_orthologs \
    --data_dir ~/db/eggnog5 \
    --cpu 8 --no_file_comments --override \
    -o result/eggnog/output

  ## æ·»è¡¨å¤´, 1åˆ—ä¸ºIDï¼Œ9åˆ—KOï¼Œ16åˆ—CAZyï¼Œ21åˆ—COGï¼Œ22åˆ—æè¿°
  sed '1 i Name\tortholog\tevalue\tscore\ttaxonomic\tprotein\tGO\tEC\tKO\tPathway\tModule\tReaction\trclass\tBRITE\tTC\tCAZy\tBiGG\ttax_scope\tOG\tbestOG\tCOG\tdescription' \
    result/eggnog/output.emapper.annotations \
    > result/eggnog/eggnog_anno_output

##############
echo end: `date +"%Y-%m-%d %T"`
end=`date +%s`
echo TIME:`expr $end - $start`s
```

#### dbCAN2ï¼ˆç¢³æ°´åŒ–åˆç‰©ï¼‰

``` bash
## dbCAN2 http://bcb.unl.edu/dbCAN2
## åˆ›å»ºæ•°æ®åº“å­˜æ”¾ç›®å½•å¹¶è¿›å…¥
mkdir -p ${db}/dbCAN2 && cd ${db}/dbCAN2
## ä¸‹è½½åºåˆ—å’Œæè¿°
wget -c http://bcb.unl.edu/dbCAN2/download/CAZyDB.07312020.fa
wget -c http://bcb.unl.edu/dbCAN2/download/Databases/CAZyDB.07302020.fam-activities.txt
## å¤‡ç”¨æ•°æ®åº“ä¸‹è½½åœ°å€å¹¶è§£å‹ 
#wget -c http://210.75.224.110/db/dbcan2/CAZyDB.07312020.fa.gz
#gunzip CAZyDB.07312020.fa.gz
## diamondå»ºç´¢å¼•
time diamond makedb \
  --in CAZyDB.07312020.fa \
  --db CAZyDB.07312020
```

``` bash
#!/bin/bash
#SBATCH --job-name=cazy
#SBATCH --output=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.out
#SBATCH --error=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.err
#SBATCH --array=1
#SBATCH --partition=cpu
#SBATCH --cpus-per-task=8
#SBATCH --mem-per-cpu=2G
echo start: `date +"%Y-%m-%d %T"`
start=`date +%s`

####################

  mkdir -p result/dbcan2
  diamond blastp   \
  	--db ~/db/dbcan2/CAZyDB.07312020  \
  	--query result/NR/protein_no_end.fa \
  	--threads 8 -e 1e-5 --outfmt 6 \
  	--max-target-seqs 1 --quiet \
  	--out result/dbcan2/gene_diamond.f6

##############
echo end: `date +"%Y-%m-%d %T"`
end=`date +%s`
echo TIME:`expr $end - $start`s
```

#### CARDï¼ˆæŠ—ç”Ÿç´ æŠ—æ€§åŸºå› ARGsï¼‰

RGIè½¯ä»¶Githubä¸»é¡µ: <https://github.com/arpcard/rgi>ï¼Œå¯ä»¥å‚è€ƒä¹‹å‰å†™çš„[Antibiotics resistance gene](../antibiotics-resistance)ã€‚

``` bash
#!/bin/bash
#SBATCH --job-name=rgi
#SBATCH --output=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.out
#SBATCH --error=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.err
#SBATCH --array=1
#SBATCH --partition=cpu
#SBATCH --cpus-per-task=8
#SBATCH --mem-per-cpu=2G

echo start: `date +"%Y-%m-%d %T"`
start=`date +%s`

####################

  source ~/miniconda3/etc/profile.d/conda.sh
  conda activate rgi
  mkdir -p result/card
  ~/miniconda3/envs/rgi/bin/rgi main \
    --input_sequence result/NR/protein_no_end.fa \
    --output_file result/card/protein \
    --input_type protein --num_threads 8 \
    --clean --alignment_tool DIAMOND # --low_quality #partial genes

##############
echo end: `date +"%Y-%m-%d %T"`
end=`date +%s`
echo TIME:`expr $end - $start`s
```

#### VFDBï¼ˆæ¯’åŠ›å› å­ï¼‰

å®˜ç½‘åœ°å€ï¼š<http://www.mgc.ac.cn/VFs/>

åœ¨å®˜ç½‘ä¸‹è½½æ•°æ®åº“æ—¶ï¼Œå¸¦æœ‰setA çš„åº“ä¸ºVFDBæ•°æ®åº“æ ¸å¿ƒåº“(set A)ï¼Œè€ŒsetBä¸ºå…¨åº“(setB), å…¶ä¸­setAä»…åŒ…å«ç»å®éªŒéªŒè¯è¿‡çš„æ¯’åŠ›åŸºå› ï¼Œè€ŒsetBåˆ™åœ¨setAçš„åŸºç¡€ä¸Šå¢åŠ äº†é¢„æµ‹çš„æ¯’åŠ›åŸºå› ï¼Œé€‰æ‹©å¥½æ•°æ®åº“åï¼Œç›´æ¥ç”¨blast/diamondå³å¯å®Œæˆæ³¨é‡Šã€‚

``` bash
#!/bin/bash
#SBATCH --job-name=vfdb
#SBATCH --output=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.out
#SBATCH --error=/share/home/jianglab/pengchen/work/asthma/log/%x_%A_%a.err
#SBATCH --array=1
#SBATCH --partition=cpu
#SBATCH --cpus-per-task=8
#SBATCH --mem-per-cpu=2G
echo start: `date +"%Y-%m-%d %T"`
start=`date +%s`

####################

  mkdir -p result/vfdb
  diamond blastp   \
  	--db ~/db/VFDB/VFDB_setB_pro \
  	--query result/NR/protein_no_end.fa \
  	--threads 8 -e 1e-5 --outfmt 6 \
  	--max-target-seqs 1 --quiet \
  	--out result/vfdb/gene_diamond.f6

##############
echo end: `date +"%Y-%m-%d %T"`
end=`date +%s`
echo TIME:`expr $end - $start`s
```

#### å…¶ä»–æ•°æ®åº“

- NCBI NRï¼ˆNon-redundantï¼‰:åŒ…å«å¹¿æ³›çš„å·²çŸ¥è›‹ç™½åºåˆ—ï¼Œæ˜¯å¸¸ç”¨çš„å‚è€ƒæ•°æ®åº“ã€‚
- Pfamï¼šåŒ…å«è›‹ç™½è´¨å®¶æ—å’ŒåŠŸèƒ½åŸŸçš„ä¿¡æ¯ï¼Œé€šè¿‡Hidden Markov Models (HMMs) æè¿°è›‹ç™½è´¨åŠŸèƒ½åŸŸã€‚
- å…ƒç´ å¾ªç¯æ•°æ®åº“ï¼šä¸“æ³¨äºç¯å¢ƒä¸­çš„å…ƒç´ ï¼ˆå¦‚ç¢³ã€æ°®ã€ç¡«ã€ç£·ç­‰ï¼‰çš„ç”Ÿç‰©åœ°çƒåŒ–å­¦å¾ªç¯ç›¸å…³åŸºå› å’Œé€”å¾„æ³¨é‡Šã€‚å¦‚NCycDBï¼ŒSCycDBï¼ŒPCycDBï¼ŒMCycDBç­‰

è¿™äº›æ•°æ®åº“æˆ‘ä»¬å¯ä»¥è‡ªè¡Œä¸‹è½½è›‹ç™½æ•°æ®åº“ï¼Œç„¶åç”¨blastå»ºåº“æ¯”å¯¹ã€‚

### åŠŸèƒ½æ³¨é‡Šåˆå¹¶

å†™ä¸€ä¸ªpythonè„šæœ¬ï¼Œå°†è¡¨1ï¼ˆåŸºå› -åŠŸèƒ½çš„å¯¹åº”è¡¨ï¼‰ä¸è¡¨2ï¼ˆåŸºå› ä¸°åº¦è¡¨ï¼‰åˆå¹¶ï¼Œå³ä¸åŒåŸºå› å¯èƒ½æ³¨é‡Šåˆ°ç›¸åŒåŠŸèƒ½ï¼ŒæŠŠå®ƒä»¬çš„ä¸°åº¦åŠ åœ¨ä¸€èµ·å¾—åˆ°æ–°è¡¨3ï¼ˆåŠŸèƒ½ä¸°åº¦è¡¨ï¼‰ã€‚

å¯ä»¥å‚è€ƒ<https://github.com/YongxinLiu/EasyMicrobiome/blob/master/script/summarizeAbundance.py>ã€‚

``` bash
mkdir -p result/summ_table
if [ -f result/eggnog/eggnog_anno_output ]; then
# æ±‡æ€»ï¼Œ9åˆ—KOï¼Œ16åˆ—CAZyæŒ‰é€—å·åˆ†éš”ï¼Œ21åˆ—COGæŒ‰å­—æ¯åˆ†éš”ï¼ŒåŸå§‹å€¼ç´¯åŠ 
~/db/script/summarizeAbundance.py \\
-i result/salmon/gene.count \\
-m result/eggnog/eggnog_anno_output \\
-c '9,16,21' -s ',+,+*' -n raw \\
-o result/summ_table/eggnog
sed -i 's/^ko://' summ_table/eggnog.KO.raw.txt
fi
```

## åˆ†ç®±ï¼ˆbinningï¼‰

å®åŸºå› ç»„binningæ˜¯æŒ‡å°†ä¸åŒçš„åºåˆ—é›†åˆï¼ˆå¦‚metagenomeåºåˆ—é›†åˆï¼‰æ ¹æ®å®ƒä»¬çš„ç‰©ç§å½’ç±»åˆ°ä¸åŒçš„binsä¸­ï¼Œä»¥ä¾¿è¿›ä¸€æ­¥ç ”ç©¶å®ƒä»¬çš„ç»„æˆå’ŒåŠŸèƒ½ã€‚è¿™ä¸ªè¿‡ç¨‹å¯ä»¥å°†ç±»ä¼¼çš„åºåˆ—ç»„åˆåœ¨ä¸€èµ·ï¼Œå½¢æˆä»£è¡¨ä¸åŒç‰©ç§æˆ–åŸºå› ç»„çš„binsï¼Œä»¥ä¾¿è¿›è¡Œåç»­åˆ†æï¼Œå¦‚ç‰©ç§æ³¨é‡Šã€åŸºå› ç»„ç»„è£…ç­‰ã€‚

ä»¥ä¸‹æ˜¯å¸¸ç”¨çš„å®åŸºå› ç»„binningæ–¹æ³•ï¼š

1.  åŸºäºèšç±»çš„æ–¹æ³•ï¼šè¯¥æ–¹æ³•ä½¿ç”¨åºåˆ—èšç±»å°†ç›¸ä¼¼åºåˆ—åˆ†åˆ°åŒä¸€ä¸ªbinä¸­ã€‚ä¸€èˆ¬æ¥è¯´ï¼Œèšç±»ç®—æ³•å¯åˆ†ä¸ºä¸¤ç±»ï¼šæ— ç›‘ç£èšç±»ï¼ˆå¦‚k-meansã€DBSCANç­‰ï¼‰å’Œæœ‰ç›‘ç£èšç±»ï¼ˆå¦‚CAMIã€MyCCç­‰ï¼‰ã€‚

2.  åŸºäºç»„è£…çš„æ–¹æ³•ï¼šè¯¥æ–¹æ³•ä½¿ç”¨de novoç»„è£…æ¥å°†ç›¸ä¼¼åºåˆ—ç»„è£…æˆè¿ç»­çš„åºåˆ—ï¼Œå†æ ¹æ®è¿™äº›åºåˆ—çš„åŸºå› ç»„ä¿¡æ¯æ¥å°†å…¶åˆ†ç±»åˆ°ä¸åŒçš„binsä¸­ã€‚è¿™ç§æ–¹æ³•çš„ä¼˜ç‚¹æ˜¯å¯ä»¥æ›´å¥½åœ°å¤„ç†é‡å¤åºåˆ—ï¼Œç¼ºç‚¹æ˜¯éœ€è¦å¤§é‡çš„è®¡ç®—èµ„æºå’Œæ—¶é—´ã€‚

3.  åŸºäºåˆ†ç±»å™¨çš„æ–¹æ³•ï¼šè¯¥æ–¹æ³•ä½¿ç”¨æœºå™¨å­¦ä¹ åˆ†ç±»å™¨æ¥å°†åºåˆ—åˆ†é…åˆ°ä¸åŒçš„binsä¸­ã€‚è¿™ç§æ–¹æ³•çš„ä¼˜ç‚¹æ˜¯å¯ä»¥è‡ªåŠ¨å­¦ä¹ ç‰¹å¾å¹¶åœ¨å¤„ç†å¤§è§„æ¨¡æ•°æ®æ—¶æ•ˆç‡é«˜ï¼Œç¼ºç‚¹æ˜¯éœ€è¦å…ˆå»ºç«‹ä¸€ä¸ªåˆ†ç±»å™¨å¹¶è¿›è¡Œè®­ç»ƒã€‚

åœ¨è¿›è¡Œå®åŸºå› ç»„binningæ—¶ï¼Œé€šå¸¸éœ€è¦ä½¿ç”¨å¤šä¸ªæ–¹æ³•è¿›è¡Œæ¯”è¾ƒï¼Œä»¥é€‰æ‹©æœ€é€‚åˆæ•°æ®é›†çš„æ–¹æ³•ã€‚å¯ä»¥ä½¿ç”¨ä¸€äº›æµè¡Œçš„å·¥å…·æ¥è¿›è¡Œbinningï¼Œå¦‚MetaBATã€MaxBinã€CONCOCTå’ŒMEGANç­‰ã€‚è¿™äº›å·¥å…·é€šå¸¸åŒ…å«å„ç§binningæ–¹æ³•ï¼Œå¯ä»¥æ ¹æ®æ•°æ®é›†å’Œåˆ†æç›®çš„é€‰æ‹©é€‚åˆçš„æ–¹æ³•ã€‚

ç¯‡å¹…é™åˆ¶ï¼Œå…·ä½“çš„æ–¹æ³•æ”¾åœ¨å¦ä¸€ç¯‡é‡Œé¢è®²è§£å§ï¼ŒæŸ¥çœ‹[å®åŸºå› ç»„åˆ†ç®±æµç¨‹](../binning)ã€‚

## Reference

<div id="refs" class="references csl-bib-body">

<div id="ref-liuPracticalGuideAmplicon2021" class="csl-entry">

<span class="csl-left-margin">1. </span><span class="csl-right-inline">Y.-X. Liu, Y. Qin, T. Chen, M. Lu, *et al.*, [A practical guide to amplicon and metagenomic analysis of microbiome data](https://doi.org/10.1007/s13238-020-00724-8). *Protein & Cell*. **12**, 315â€“330 (2021).</span>

</div>

</div>
